{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The autoreload extension is already loaded. To reload it, use:\n",
      "  %reload_ext autoreload\n"
     ]
    }
   ],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Comparison of mutational robustness of sRNA's vs. randomly generated RNA"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[CpuDevice(id=0)]"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import os\n",
    "from Bio import SeqIO\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "import pandas as pd\n",
    "import re\n",
    "import os\n",
    "from subprocess import Popen, PIPE\n",
    "from copy import deepcopy\n",
    "\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "from datetime import datetime\n",
    "from synbio_morpher.srv.io.manage.script_manager import script_preamble\n",
    "from synbio_morpher.srv.parameter_prediction.simulator import process_raw_stdout\n",
    "from synbio_morpher.srv.sequence_exploration.sequence_analysis import b_tabulate_mutation_info\n",
    "from synbio_morpher.utils.common.setup import expand_model_config, prepare_config, expand_config, compose_kwargs, instantiate_system\n",
    "from synbio_morpher.utils.evolution.evolver import Evolver\n",
    "from synbio_morpher.utils.evolution.mutation import apply_mutation_to_sequence, get_mutation_type_mapping, reverse_mut_mapping\n",
    "from synbio_morpher.utils.circuit.agnostic_circuits.circuit_manager import CircuitModeller\n",
    "from synbio_morpher.utils.misc.type_handling import flatten_listlike, flatten_nested_dict\n",
    "from synbio_morpher.utils.data.data_format_tools.common import load_multiple_as_list, load_json_as_dict\n",
    "\n",
    "\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import jax\n",
    "jax.config.update('jax_platform_name', 'cpu')\n",
    "\n",
    "jax.devices()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [],
   "source": [
    "fn = '../data/sRNA/merged_inter.csv'\n",
    "data = pd.read_csv(fn)\n",
    "try:\n",
    "    data = data.drop(columns=['Unnamed: 0'])\n",
    "except:\n",
    "    pass\n",
    "data = data[data['Sequence 1'].apply(lambda x: type(x) == str)]\n",
    "data = data[data['Sequence 2'].apply(lambda x: type(x) == str)]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Interactor 1</th>\n",
       "      <th>Interactor 2</th>\n",
       "      <th>ID 1</th>\n",
       "      <th>ID 2</th>\n",
       "      <th>Category 1</th>\n",
       "      <th>Category 2</th>\n",
       "      <th>Regulation</th>\n",
       "      <th>Binding position 1</th>\n",
       "      <th>Binding position 2</th>\n",
       "      <th>Sequence 1</th>\n",
       "      <th>Sequence 2</th>\n",
       "      <th>Source DB</th>\n",
       "      <th>Genome position 1</th>\n",
       "      <th>Genome position 2</th>\n",
       "      <th>Strand 1</th>\n",
       "      <th>Strand 2</th>\n",
       "      <th>Sequence before 1</th>\n",
       "      <th>Sequence before 2</th>\n",
       "      <th>Sequence after 1</th>\n",
       "      <th>Sequence after 2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>dsrA</td>\n",
       "      <td>dsrB</td>\n",
       "      <td>NCBI:946470</td>\n",
       "      <td>NCBI:946468</td>\n",
       "      <td>sRNA</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AACACATCAGATTTCCTGGTGTAACGAATTTTTTAAGTGCTTCTTG...</td>\n",
       "      <td>TAAATGNAAACAAGTTTTATTTTTATCATCAGAACATTCTAATAAA...</td>\n",
       "      <td>RNAInter</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ttcatcaccttatccgcaatttttttcgctgacaagaaaatattcg...</td>\n",
       "      <td>atcgttttcctcctgtggctttgtgccagtgtagaacaatttcgtt...</td>\n",
       "      <td>cgtgaggtcagcgtcgatgtcgatgccctgctggcggcgatcaatg...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>dsrA</td>\n",
       "      <td>ilvH</td>\n",
       "      <td>NCBI:946470</td>\n",
       "      <td>NCBI:947267</td>\n",
       "      <td>sRNA</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AACACATCAGATTTCCTGGTGTAACGAATTTTTTAAGTGCTTCTTG...</td>\n",
       "      <td>ATTCGCCACAACAAGGACAGGAAGGTTCGGCGCAAGGAGCCCAAGA...</td>\n",
       "      <td>RNAInter</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ttcatcaccttatccgcaatttttttcgctgacaagaaaatattcg...</td>\n",
       "      <td>ATTCGCCACAACAAGGACAGGAAGGTTCGGCGCAAGGAGCCCAAGA...</td>\n",
       "      <td>cgtgaggtcagcgtcgatgtcgatgccctgctggcggcgatcaatg...</td>\n",
       "      <td>ATTCGCCACAACAAGGACAGGAAGGTTCGGCGCAAGGAGCCCAAGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>oxyS</td>\n",
       "      <td>fhlA</td>\n",
       "      <td>NCBI:2847701</td>\n",
       "      <td>NCBI:947181</td>\n",
       "      <td>sRNA</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GAAACGGAGCGGCACCTCTTTTAACCCTTGAAGTCACTGCCCGTTT...</td>\n",
       "      <td>CACCAGATTGCCAAGCTGATGGTCGAGCTGTCCAAATCCCAGGATG...</td>\n",
       "      <td>RNAInter</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TTTGCGAACTTATCGGTTTCGTATTGAGGGCAATAAACTCTCGAGG...</td>\n",
       "      <td>CACCAGATTGCCAAGCTGATGGTCGAGCTGTCCAAATCCCAGGATG...</td>\n",
       "      <td>TTTGCGAACTTATCGGTTTCGTATTGAGGGCAATAAACTCTCGAGG...</td>\n",
       "      <td>CACCAGATTGCCAAGCTGATGGTCGAGCTGTCCAAATCCCAGGATG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>sgrS</td>\n",
       "      <td>hfq</td>\n",
       "      <td>NCBI:4056038</td>\n",
       "      <td>NCBI:948689</td>\n",
       "      <td>sRNA</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GATGAAGCAAGGGGGTGCCCCATGCGTCAGTTTTATCAGCACTATT...</td>\n",
       "      <td>ATGGCTAAGGGGCAATCTTTACAAGATCCGTTCCTGAACGCACTGC...</td>\n",
       "      <td>RNAInter</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ttgagcagggtgcgcatatgacgacgcgagcagctcaataacgctg...</td>\n",
       "      <td>cgaagcggcagataacctggctgcgtggttgggaaggggttcactg...</td>\n",
       "      <td>ttattctcgccgcgctaaaaagggaacgtatgatctggataatgac...</td>\n",
       "      <td>aaggtttcgggctgtttttttacacggggagccagcgatcctgcgt...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>oxyS</td>\n",
       "      <td>nusG</td>\n",
       "      <td>NCBI:2847701</td>\n",
       "      <td>NCBI:948485</td>\n",
       "      <td>sRNA</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GAAACGGAGCGGCACCTCTTTTAACCCTTGAAGTCACTGCCCGTTT...</td>\n",
       "      <td>ATGTCTGAAGCTCCTAAAAAGCGCTGGTACGTCGTTCAGGCGTTTT...</td>\n",
       "      <td>RNAInter</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TTTGCGAACTTATCGGTTTCGTATTGAGGGCAATAAACTCTCGAGG...</td>\n",
       "      <td>cgacaaaaggtaaagctaccgttgcttttgcccgtgaagcgcgtac...</td>\n",
       "      <td>TTTGCGAACTTATCGGTTTCGTATTGAGGGCAATAAACTCTCGAGG...</td>\n",
       "      <td>aacccagcgatcaaaaaagcggcgatttaatcgttgcacaaggcgt...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>373</th>\n",
       "      <td>ryhB</td>\n",
       "      <td>msrA</td>\n",
       "      <td>chromosome:NC_000913.3, Gene ID:2847761</td>\n",
       "      <td>chromosome:NC_000913.3, Gene ID:948734</td>\n",
       "      <td>sRNA</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>No Interaction</td>\n",
       "      <td>['NA']</td>\n",
       "      <td>['NA']</td>\n",
       "      <td>GCGATCAGGAAGACCCTCGCGGAGAACCTGAAAGCACGACATTGCT...</td>\n",
       "      <td>ATGAGTTTATTTGATAAAAAGCATCTGGTTTCCCCCGCCGATGCCC...</td>\n",
       "      <td>sRNATarBase</td>\n",
       "      <td>3580927..3581016</td>\n",
       "      <td>4441538..4442176</td>\n",
       "      <td>reverse</td>\n",
       "      <td>reverse</td>\n",
       "      <td>TTTGAGAGCGTCTCTGTCCCTCGTTTTGCGGTTAAGCCGCATCCAT...</td>\n",
       "      <td>atggtgtcgctctcccgaaaatcggtcattggggttaaggttgtaa...</td>\n",
       "      <td>TTTGAGAGCGTCTCTGTCCCTCGTTTTGCGGTTAAGCCGCATCCAT...</td>\n",
       "      <td>aggtactgtaaagccgccagagggttaaaattcagacagctgaaaa...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>374</th>\n",
       "      <td>gcvB</td>\n",
       "      <td>hdeA</td>\n",
       "      <td>chromosome:NC_000913.3, Gene ID:2847720</td>\n",
       "      <td>chromosome:NC_000913.3, Gene ID:948025</td>\n",
       "      <td>trans-encoded antisense RNA</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>Induction</td>\n",
       "      <td>['NA']</td>\n",
       "      <td>['NA']</td>\n",
       "      <td>ACTTCCTGAGCCGGAACGAAAAGTTTTATCGGAATGCGTGTTCTGG...</td>\n",
       "      <td>ATGAAAAAAGTATTAGGCGTTATTCTTGGTGGTCTGCTTCTTCTGC...</td>\n",
       "      <td>sRNATarBase</td>\n",
       "      <td>2942696..2942901</td>\n",
       "      <td>3656408..3656740</td>\n",
       "      <td>forward</td>\n",
       "      <td>reverse</td>\n",
       "      <td>GGTAGAGCGCCATTGCATTCTTTAGCGCCTCGATCCCTTTATTTTT...</td>\n",
       "      <td>atcgtaatatcctcaactataaagtgaaagagccgtcacgaatcaa...</td>\n",
       "      <td>GGTAGAGCGCCATTGCATTCTTTAGCGCCTCGATCCCTTTATTTTT...</td>\n",
       "      <td>gccaacgcagattgtgcgttcaccagtgacaaagccgctacagcgc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>375</th>\n",
       "      <td>gcvB</td>\n",
       "      <td>hdeB</td>\n",
       "      <td>chromosome:NC_000913.3, Gene ID:2847720</td>\n",
       "      <td>chromosome:NC_000913.3, Gene ID:948026</td>\n",
       "      <td>trans-encoded antisense RNA</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>Induction</td>\n",
       "      <td>['NA']</td>\n",
       "      <td>['NA']</td>\n",
       "      <td>ACTTCCTGAGCCGGAACGAAAAGTTTTATCGGAATGCGTGTTCTGG...</td>\n",
       "      <td>ATGAATATTTCATCTCTCCGTAAAGCGTTTATTTTTATGGGCGCTG...</td>\n",
       "      <td>sRNATarBase</td>\n",
       "      <td>2942696..2942901</td>\n",
       "      <td>3655966..3656292</td>\n",
       "      <td>forward</td>\n",
       "      <td>reverse</td>\n",
       "      <td>GGTAGAGCGCCATTGCATTCTTTAGCGCCTCGATCCCTTTATTTTT...</td>\n",
       "      <td>attttgtaacccattcaatatagagattatatatgaacaggaggaa...</td>\n",
       "      <td>GGTAGAGCGCCATTGCATTCTTTAGCGCCTCGATCCCTTTATTTTT...</td>\n",
       "      <td>cataaacagggcgcttcccatgccaattaatacatgtgtgcgtaac...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>376</th>\n",
       "      <td>Esr41</td>\n",
       "      <td>fliC</td>\n",
       "      <td>chromosome:NC_002695.1, Gene ID:</td>\n",
       "      <td>chromosome:NC_002695.1, Gene ID:</td>\n",
       "      <td>trans-encoded antisense RNA</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>Induction</td>\n",
       "      <td>['NA']</td>\n",
       "      <td>['NA']</td>\n",
       "      <td>GATGCTCTAGGCATCACATTTTCTCCATGGGGTATTCCCTCCGCCG...</td>\n",
       "      <td>ATGGCACAAGTCATTAATACCAACAGCCTCTCGCTGATCACTCAAA...</td>\n",
       "      <td>sRNATarBase</td>\n",
       "      <td>1422406..1422479</td>\n",
       "      <td>2624379..2626136</td>\n",
       "      <td>forward</td>\n",
       "      <td>reverse</td>\n",
       "      <td>gatgctctaggcatcacattttctccatggggtattccctccgccg...</td>\n",
       "      <td>ttaaccctgcagcagagacagaacctgctgcggtacctggttagct...</td>\n",
       "      <td>gatgctctaggcatcacattttctccatggggtattccctccgccg...</td>\n",
       "      <td>ttaaccctgcagcagagacagaacctgctgcggtacctggttagct...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>377</th>\n",
       "      <td>dsrA</td>\n",
       "      <td>mreB</td>\n",
       "      <td>chromosome:NC_000913.3, Gene ID:946470</td>\n",
       "      <td>chromosome:NC_000913.3, Gene ID:948588</td>\n",
       "      <td>trans-encoded antisense RNA</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>Repression</td>\n",
       "      <td>['24..41']</td>\n",
       "      <td>['-3..18']</td>\n",
       "      <td>AACACATCAGATTTCCTGGTGTAACGAATTTTTTAAGTGCTTCTTG...</td>\n",
       "      <td>ATGTTGAAAAAATTTCGTGGCATGTTTTCCAATGACTTGTCCATTG...</td>\n",
       "      <td>sRNATarBase</td>\n",
       "      <td>2025227..2025313</td>\n",
       "      <td>3400044..3401087</td>\n",
       "      <td>reverse</td>\n",
       "      <td>reverse</td>\n",
       "      <td>ttcatcaccttatccgcaatttttttcgctgacaagaaaatattcg...</td>\n",
       "      <td>atactaagggataatcctgaaagctggggcggaaaagaaaatccgc...</td>\n",
       "      <td>cgtgaggtcagcgtcgatgtcgatgccctgctggcggcgatcaatg...</td>\n",
       "      <td>cggcggtatccatataagtacggatttgactgaacgtccccaggcg...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>378 rows × 20 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Interactor 1 Interactor 2                                     ID 1  \\\n",
       "0           dsrA         dsrB                              NCBI:946470   \n",
       "1           dsrA         ilvH                              NCBI:946470   \n",
       "2           oxyS         fhlA                             NCBI:2847701   \n",
       "3           sgrS          hfq                             NCBI:4056038   \n",
       "4           oxyS         nusG                             NCBI:2847701   \n",
       "..           ...          ...                                      ...   \n",
       "373         ryhB         msrA  chromosome:NC_000913.3, Gene ID:2847761   \n",
       "374         gcvB         hdeA  chromosome:NC_000913.3, Gene ID:2847720   \n",
       "375         gcvB         hdeB  chromosome:NC_000913.3, Gene ID:2847720   \n",
       "376        Esr41         fliC         chromosome:NC_002695.1, Gene ID:   \n",
       "377         dsrA         mreB   chromosome:NC_000913.3, Gene ID:946470   \n",
       "\n",
       "                                       ID 2                   Category 1  \\\n",
       "0                               NCBI:946468                         sRNA   \n",
       "1                               NCBI:947267                         sRNA   \n",
       "2                               NCBI:947181                         sRNA   \n",
       "3                               NCBI:948689                         sRNA   \n",
       "4                               NCBI:948485                         sRNA   \n",
       "..                                      ...                          ...   \n",
       "373  chromosome:NC_000913.3, Gene ID:948734                         sRNA   \n",
       "374  chromosome:NC_000913.3, Gene ID:948025  trans-encoded antisense RNA   \n",
       "375  chromosome:NC_000913.3, Gene ID:948026  trans-encoded antisense RNA   \n",
       "376        chromosome:NC_002695.1, Gene ID:  trans-encoded antisense RNA   \n",
       "377  chromosome:NC_000913.3, Gene ID:948588  trans-encoded antisense RNA   \n",
       "\n",
       "    Category 2      Regulation Binding position 1 Binding position 2  \\\n",
       "0         mRNA             NaN                NaN                NaN   \n",
       "1         mRNA             NaN                NaN                NaN   \n",
       "2         mRNA             NaN                NaN                NaN   \n",
       "3         mRNA             NaN                NaN                NaN   \n",
       "4         mRNA             NaN                NaN                NaN   \n",
       "..         ...             ...                ...                ...   \n",
       "373       mRNA  No Interaction             ['NA']             ['NA']   \n",
       "374       mRNA       Induction             ['NA']             ['NA']   \n",
       "375       mRNA       Induction             ['NA']             ['NA']   \n",
       "376       mRNA       Induction             ['NA']             ['NA']   \n",
       "377       mRNA      Repression         ['24..41']         ['-3..18']   \n",
       "\n",
       "                                            Sequence 1  \\\n",
       "0    AACACATCAGATTTCCTGGTGTAACGAATTTTTTAAGTGCTTCTTG...   \n",
       "1    AACACATCAGATTTCCTGGTGTAACGAATTTTTTAAGTGCTTCTTG...   \n",
       "2    GAAACGGAGCGGCACCTCTTTTAACCCTTGAAGTCACTGCCCGTTT...   \n",
       "3    GATGAAGCAAGGGGGTGCCCCATGCGTCAGTTTTATCAGCACTATT...   \n",
       "4    GAAACGGAGCGGCACCTCTTTTAACCCTTGAAGTCACTGCCCGTTT...   \n",
       "..                                                 ...   \n",
       "373  GCGATCAGGAAGACCCTCGCGGAGAACCTGAAAGCACGACATTGCT...   \n",
       "374  ACTTCCTGAGCCGGAACGAAAAGTTTTATCGGAATGCGTGTTCTGG...   \n",
       "375  ACTTCCTGAGCCGGAACGAAAAGTTTTATCGGAATGCGTGTTCTGG...   \n",
       "376  GATGCTCTAGGCATCACATTTTCTCCATGGGGTATTCCCTCCGCCG...   \n",
       "377  AACACATCAGATTTCCTGGTGTAACGAATTTTTTAAGTGCTTCTTG...   \n",
       "\n",
       "                                            Sequence 2    Source DB  \\\n",
       "0    TAAATGNAAACAAGTTTTATTTTTATCATCAGAACATTCTAATAAA...     RNAInter   \n",
       "1    ATTCGCCACAACAAGGACAGGAAGGTTCGGCGCAAGGAGCCCAAGA...     RNAInter   \n",
       "2    CACCAGATTGCCAAGCTGATGGTCGAGCTGTCCAAATCCCAGGATG...     RNAInter   \n",
       "3    ATGGCTAAGGGGCAATCTTTACAAGATCCGTTCCTGAACGCACTGC...     RNAInter   \n",
       "4    ATGTCTGAAGCTCCTAAAAAGCGCTGGTACGTCGTTCAGGCGTTTT...     RNAInter   \n",
       "..                                                 ...          ...   \n",
       "373  ATGAGTTTATTTGATAAAAAGCATCTGGTTTCCCCCGCCGATGCCC...  sRNATarBase   \n",
       "374  ATGAAAAAAGTATTAGGCGTTATTCTTGGTGGTCTGCTTCTTCTGC...  sRNATarBase   \n",
       "375  ATGAATATTTCATCTCTCCGTAAAGCGTTTATTTTTATGGGCGCTG...  sRNATarBase   \n",
       "376  ATGGCACAAGTCATTAATACCAACAGCCTCTCGCTGATCACTCAAA...  sRNATarBase   \n",
       "377  ATGTTGAAAAAATTTCGTGGCATGTTTTCCAATGACTTGTCCATTG...  sRNATarBase   \n",
       "\n",
       "    Genome position 1 Genome position 2 Strand 1 Strand 2  \\\n",
       "0                 NaN               NaN      NaN      NaN   \n",
       "1                 NaN               NaN      NaN      NaN   \n",
       "2                 NaN               NaN      NaN      NaN   \n",
       "3                 NaN               NaN      NaN      NaN   \n",
       "4                 NaN               NaN      NaN      NaN   \n",
       "..                ...               ...      ...      ...   \n",
       "373  3580927..3581016  4441538..4442176  reverse  reverse   \n",
       "374  2942696..2942901  3656408..3656740  forward  reverse   \n",
       "375  2942696..2942901  3655966..3656292  forward  reverse   \n",
       "376  1422406..1422479  2624379..2626136  forward  reverse   \n",
       "377  2025227..2025313  3400044..3401087  reverse  reverse   \n",
       "\n",
       "                                     Sequence before 1  \\\n",
       "0    ttcatcaccttatccgcaatttttttcgctgacaagaaaatattcg...   \n",
       "1    ttcatcaccttatccgcaatttttttcgctgacaagaaaatattcg...   \n",
       "2    TTTGCGAACTTATCGGTTTCGTATTGAGGGCAATAAACTCTCGAGG...   \n",
       "3    ttgagcagggtgcgcatatgacgacgcgagcagctcaataacgctg...   \n",
       "4    TTTGCGAACTTATCGGTTTCGTATTGAGGGCAATAAACTCTCGAGG...   \n",
       "..                                                 ...   \n",
       "373  TTTGAGAGCGTCTCTGTCCCTCGTTTTGCGGTTAAGCCGCATCCAT...   \n",
       "374  GGTAGAGCGCCATTGCATTCTTTAGCGCCTCGATCCCTTTATTTTT...   \n",
       "375  GGTAGAGCGCCATTGCATTCTTTAGCGCCTCGATCCCTTTATTTTT...   \n",
       "376  gatgctctaggcatcacattttctccatggggtattccctccgccg...   \n",
       "377  ttcatcaccttatccgcaatttttttcgctgacaagaaaatattcg...   \n",
       "\n",
       "                                     Sequence before 2  \\\n",
       "0    atcgttttcctcctgtggctttgtgccagtgtagaacaatttcgtt...   \n",
       "1    ATTCGCCACAACAAGGACAGGAAGGTTCGGCGCAAGGAGCCCAAGA...   \n",
       "2    CACCAGATTGCCAAGCTGATGGTCGAGCTGTCCAAATCCCAGGATG...   \n",
       "3    cgaagcggcagataacctggctgcgtggttgggaaggggttcactg...   \n",
       "4    cgacaaaaggtaaagctaccgttgcttttgcccgtgaagcgcgtac...   \n",
       "..                                                 ...   \n",
       "373  atggtgtcgctctcccgaaaatcggtcattggggttaaggttgtaa...   \n",
       "374  atcgtaatatcctcaactataaagtgaaagagccgtcacgaatcaa...   \n",
       "375  attttgtaacccattcaatatagagattatatatgaacaggaggaa...   \n",
       "376  ttaaccctgcagcagagacagaacctgctgcggtacctggttagct...   \n",
       "377  atactaagggataatcctgaaagctggggcggaaaagaaaatccgc...   \n",
       "\n",
       "                                      Sequence after 1  \\\n",
       "0    cgtgaggtcagcgtcgatgtcgatgccctgctggcggcgatcaatg...   \n",
       "1    cgtgaggtcagcgtcgatgtcgatgccctgctggcggcgatcaatg...   \n",
       "2    TTTGCGAACTTATCGGTTTCGTATTGAGGGCAATAAACTCTCGAGG...   \n",
       "3    ttattctcgccgcgctaaaaagggaacgtatgatctggataatgac...   \n",
       "4    TTTGCGAACTTATCGGTTTCGTATTGAGGGCAATAAACTCTCGAGG...   \n",
       "..                                                 ...   \n",
       "373  TTTGAGAGCGTCTCTGTCCCTCGTTTTGCGGTTAAGCCGCATCCAT...   \n",
       "374  GGTAGAGCGCCATTGCATTCTTTAGCGCCTCGATCCCTTTATTTTT...   \n",
       "375  GGTAGAGCGCCATTGCATTCTTTAGCGCCTCGATCCCTTTATTTTT...   \n",
       "376  gatgctctaggcatcacattttctccatggggtattccctccgccg...   \n",
       "377  cgtgaggtcagcgtcgatgtcgatgccctgctggcggcgatcaatg...   \n",
       "\n",
       "                                      Sequence after 2  \n",
       "0                                                  NaN  \n",
       "1    ATTCGCCACAACAAGGACAGGAAGGTTCGGCGCAAGGAGCCCAAGA...  \n",
       "2    CACCAGATTGCCAAGCTGATGGTCGAGCTGTCCAAATCCCAGGATG...  \n",
       "3    aaggtttcgggctgtttttttacacggggagccagcgatcctgcgt...  \n",
       "4    aacccagcgatcaaaaaagcggcgatttaatcgttgcacaaggcgt...  \n",
       "..                                                 ...  \n",
       "373  aggtactgtaaagccgccagagggttaaaattcagacagctgaaaa...  \n",
       "374  gccaacgcagattgtgcgttcaccagtgacaaagccgctacagcgc...  \n",
       "375  cataaacagggcgcttcccatgccaattaatacatgtgtgcgtaac...  \n",
       "376  ttaaccctgcagcagagacagaacctgctgcggtacctggttagct...  \n",
       "377  cggcggtatccatataagtacggatttgactgaacgtccccaggcg...  \n",
       "\n",
       "[378 rows x 20 columns]"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['dsrA', 'oxyS', 'sgrS', 'ryhB', 'yzgL', 'ygaY', 'yagJ', 'sdsR',\n",
       "       'micF', 'micC', 'istR', 'omrA', 'omrB', 'rprA', 'rseX', 'rybB',\n",
       "       'Spot_42', 'cyaR', 'dicF', 'gadY', 'symR', 'sokB', 'sokC', 'gcvB',\n",
       "       'rdlD', 'smpB', 'fnrS', 'sibC', 'arcZ', 'chiX', 'mgrR', 'sibB',\n",
       "       'mcaS', 'ohsC', 'OrzO', 'sibD', 'sibE', 'sibA', 'IpeX', 'rdlA',\n",
       "       'rdlB', 'rdlC', 'arrS', 'mntS', 'AfaR', 'Esr41'], dtype=object)"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data['Interactor 1'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Interactor 1</th>\n",
       "      <th>Interactor 2</th>\n",
       "      <th>ID 1</th>\n",
       "      <th>ID 2</th>\n",
       "      <th>Category 1</th>\n",
       "      <th>Category 2</th>\n",
       "      <th>Regulation</th>\n",
       "      <th>Binding position 1</th>\n",
       "      <th>Binding position 2</th>\n",
       "      <th>Sequence 1</th>\n",
       "      <th>Sequence 2</th>\n",
       "      <th>Source DB</th>\n",
       "      <th>Genome position 1</th>\n",
       "      <th>Genome position 2</th>\n",
       "      <th>Strand 1</th>\n",
       "      <th>Strand 2</th>\n",
       "      <th>Sequence before 1</th>\n",
       "      <th>Sequence before 2</th>\n",
       "      <th>Sequence after 1</th>\n",
       "      <th>Sequence after 2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>234</th>\n",
       "      <td>chiX</td>\n",
       "      <td>rpoS</td>\n",
       "      <td>chromosome:NC_000913.3, Gene ID:5061500</td>\n",
       "      <td>chromosome:NC_000913.3, Gene ID:947210</td>\n",
       "      <td>sRNA</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>No Interaction</td>\n",
       "      <td>['NA']</td>\n",
       "      <td>['NA']</td>\n",
       "      <td>ACACCGTCGCTTAAAGTGACGGCATAATAATAAAAAAATGAAATTC...</td>\n",
       "      <td>ATGAGTCAGAATACGCTGAAAGTTCATGATTTAAATGAAGATGCGG...</td>\n",
       "      <td>sRNATarBase</td>\n",
       "      <td>507204..507287</td>\n",
       "      <td>2866559..2867551</td>\n",
       "      <td>forward</td>\n",
       "      <td>reverse</td>\n",
       "      <td>gtttcagccggatcgtgctcgtaggtatggatttgaaacgaaatct...</td>\n",
       "      <td>ATCCTCGGGTCTTGCAGGCCACACAGGACACCCTGAACCGTCATGG...</td>\n",
       "      <td>ttagcgcaacatttgcggcaaattcccttctccatacaggtgtagt...</td>\n",
       "      <td>ATCCTCGGGTCTTGCAGGCCACACAGGACACCCTGAACCGTCATGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>290</th>\n",
       "      <td>chiX</td>\n",
       "      <td>citA</td>\n",
       "      <td>chromosome:NC_000913.3, Gene ID:5061500</td>\n",
       "      <td>chromosome:NC_000913.3, Gene ID:945233</td>\n",
       "      <td>trans-encoded antisense RNA</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>Repression</td>\n",
       "      <td>['46..57']</td>\n",
       "      <td>['-37..-26']</td>\n",
       "      <td>ACACCGTCGCTTAAAGTGACGGCATAATAATAAAAAAATGAAATTC...</td>\n",
       "      <td>ATGTTGCAGCTTAACGAGAATAAACAGTTTGCATTTTTCCAAAGAC...</td>\n",
       "      <td>sRNATarBase</td>\n",
       "      <td>507204..507287</td>\n",
       "      <td>652235..653893</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>gtttcagccggatcgtgctcgtaggtatggatttgaaacgaaatct...</td>\n",
       "      <td>GCCTGTGACCGACGTACTGGGGGATACGCAGATAAACTCAT</td>\n",
       "      <td>ttagcgcaacatttgcggcaaattcccttctccatacaggtgtagt...</td>\n",
       "      <td>GCCTGTGACCGACGTACTGGGGGATACGCAGATAAACTCAT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>308</th>\n",
       "      <td>chiX</td>\n",
       "      <td>chiP</td>\n",
       "      <td>chromosome:NC_000913.3, Gene ID:5061500</td>\n",
       "      <td>chromosome:NC_000913.3, Gene ID:945296</td>\n",
       "      <td>trans-encoded antisense RNA</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>Repression</td>\n",
       "      <td>['81..92']</td>\n",
       "      <td>['45..56']</td>\n",
       "      <td>ACACCGTCGCTTAAAGTGACGGCATAATAATAAAAAAATGAAATTC...</td>\n",
       "      <td>ATGCGTACGTTTAGTGGCAAACGTAGTACGCTGGCGCTGGCTATCG...</td>\n",
       "      <td>sRNATarBase</td>\n",
       "      <td>507204..507287</td>\n",
       "      <td>708334..709740</td>\n",
       "      <td>forward</td>\n",
       "      <td>forward</td>\n",
       "      <td>gtttcagccggatcgtgctcgtaggtatggatttgaaacgaaatct...</td>\n",
       "      <td>GCCTGTGACGGGGTCGGTTTGTACACCGATAAACTCAT</td>\n",
       "      <td>ttagcgcaacatttgcggcaaattcccttctccatacaggtgtagt...</td>\n",
       "      <td>GCCTGTGACGGGGTCGGTTTGTACACCGATAAACTCAT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>341</th>\n",
       "      <td>chiX</td>\n",
       "      <td>chbC</td>\n",
       "      <td>chromosome:NC_000913.3, Gene ID:5061500</td>\n",
       "      <td>chromosome:NC_000913.3, Gene ID:945982</td>\n",
       "      <td>trans-encoded antisense RNA</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>Repression</td>\n",
       "      <td>['NA']</td>\n",
       "      <td>['NA']</td>\n",
       "      <td>ACACCGTCGCTTAAAGTGACGGCATAATAATAAAAAAATGAAATTC...</td>\n",
       "      <td>ATGAGTAATGTTATTGCATCGCTTGAAAAGGTACTCCTCCCTTTTG...</td>\n",
       "      <td>sRNATarBase</td>\n",
       "      <td>507204..507287</td>\n",
       "      <td>1819856..1821214</td>\n",
       "      <td>forward</td>\n",
       "      <td>reverse</td>\n",
       "      <td>gtttcagccggatcgtgctcgtaggtatggatttgaaacgaaatct...</td>\n",
       "      <td>atagaaaaatacccttaaaaaccgcaatttaaatattgcggtattg...</td>\n",
       "      <td>ttagcgcaacatttgcggcaaattcccttctccatacaggtgtagt...</td>\n",
       "      <td>ttgctgcggcaaaatcgccctgcttcgcctgtttcagtgccgcata...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Interactor 1 Interactor 2                                     ID 1  \\\n",
       "234         chiX         rpoS  chromosome:NC_000913.3, Gene ID:5061500   \n",
       "290         chiX         citA  chromosome:NC_000913.3, Gene ID:5061500   \n",
       "308         chiX         chiP  chromosome:NC_000913.3, Gene ID:5061500   \n",
       "341         chiX         chbC  chromosome:NC_000913.3, Gene ID:5061500   \n",
       "\n",
       "                                       ID 2                   Category 1  \\\n",
       "234  chromosome:NC_000913.3, Gene ID:947210                         sRNA   \n",
       "290  chromosome:NC_000913.3, Gene ID:945233  trans-encoded antisense RNA   \n",
       "308  chromosome:NC_000913.3, Gene ID:945296  trans-encoded antisense RNA   \n",
       "341  chromosome:NC_000913.3, Gene ID:945982  trans-encoded antisense RNA   \n",
       "\n",
       "    Category 2      Regulation Binding position 1 Binding position 2  \\\n",
       "234       mRNA  No Interaction             ['NA']             ['NA']   \n",
       "290       mRNA      Repression         ['46..57']       ['-37..-26']   \n",
       "308       mRNA      Repression         ['81..92']         ['45..56']   \n",
       "341       mRNA      Repression             ['NA']             ['NA']   \n",
       "\n",
       "                                            Sequence 1  \\\n",
       "234  ACACCGTCGCTTAAAGTGACGGCATAATAATAAAAAAATGAAATTC...   \n",
       "290  ACACCGTCGCTTAAAGTGACGGCATAATAATAAAAAAATGAAATTC...   \n",
       "308  ACACCGTCGCTTAAAGTGACGGCATAATAATAAAAAAATGAAATTC...   \n",
       "341  ACACCGTCGCTTAAAGTGACGGCATAATAATAAAAAAATGAAATTC...   \n",
       "\n",
       "                                            Sequence 2    Source DB  \\\n",
       "234  ATGAGTCAGAATACGCTGAAAGTTCATGATTTAAATGAAGATGCGG...  sRNATarBase   \n",
       "290  ATGTTGCAGCTTAACGAGAATAAACAGTTTGCATTTTTCCAAAGAC...  sRNATarBase   \n",
       "308  ATGCGTACGTTTAGTGGCAAACGTAGTACGCTGGCGCTGGCTATCG...  sRNATarBase   \n",
       "341  ATGAGTAATGTTATTGCATCGCTTGAAAAGGTACTCCTCCCTTTTG...  sRNATarBase   \n",
       "\n",
       "    Genome position 1 Genome position 2 Strand 1 Strand 2  \\\n",
       "234    507204..507287  2866559..2867551  forward  reverse   \n",
       "290    507204..507287    652235..653893  forward  forward   \n",
       "308    507204..507287    708334..709740  forward  forward   \n",
       "341    507204..507287  1819856..1821214  forward  reverse   \n",
       "\n",
       "                                     Sequence before 1  \\\n",
       "234  gtttcagccggatcgtgctcgtaggtatggatttgaaacgaaatct...   \n",
       "290  gtttcagccggatcgtgctcgtaggtatggatttgaaacgaaatct...   \n",
       "308  gtttcagccggatcgtgctcgtaggtatggatttgaaacgaaatct...   \n",
       "341  gtttcagccggatcgtgctcgtaggtatggatttgaaacgaaatct...   \n",
       "\n",
       "                                     Sequence before 2  \\\n",
       "234  ATCCTCGGGTCTTGCAGGCCACACAGGACACCCTGAACCGTCATGG...   \n",
       "290          GCCTGTGACCGACGTACTGGGGGATACGCAGATAAACTCAT   \n",
       "308             GCCTGTGACGGGGTCGGTTTGTACACCGATAAACTCAT   \n",
       "341  atagaaaaatacccttaaaaaccgcaatttaaatattgcggtattg...   \n",
       "\n",
       "                                      Sequence after 1  \\\n",
       "234  ttagcgcaacatttgcggcaaattcccttctccatacaggtgtagt...   \n",
       "290  ttagcgcaacatttgcggcaaattcccttctccatacaggtgtagt...   \n",
       "308  ttagcgcaacatttgcggcaaattcccttctccatacaggtgtagt...   \n",
       "341  ttagcgcaacatttgcggcaaattcccttctccatacaggtgtagt...   \n",
       "\n",
       "                                      Sequence after 2  \n",
       "234  ATCCTCGGGTCTTGCAGGCCACACAGGACACCCTGAACCGTCATGG...  \n",
       "290          GCCTGTGACCGACGTACTGGGGGATACGCAGATAAACTCAT  \n",
       "308             GCCTGTGACGGGGTCGGTTTGTACACCGATAAACTCAT  \n",
       "341  ttgctgcggcaaaatcgccctgcttcgcctgtttcagtgccgcata...  "
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data[data['Interactor 1'] == 'chiX']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "378"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(data)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Cut out the RNA's that are longer than 300bp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "375"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data = data[data['Sequence 1'].apply(len) <= 300]\n",
    "len(data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "43"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(data['Interactor 1'].unique())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Config"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [],
   "source": [
    "config = {\n",
    "    \"experiment\": {\n",
    "        \"purpose\": \"tests\",\n",
    "        \"no_visualisations\": False,\n",
    "        \"no_numerical\": False,\n",
    "        \"debug_mode\": False\n",
    "    },\n",
    "    \"system_type\": \"RNA\",\n",
    "    \"include_prod_deg\": False,\n",
    "    \"interaction_simulator\": {\n",
    "        \"name\": \"IntaRNA\",\n",
    "        \"postprocess\": True,\n",
    "        \"multithread\": 0,\n",
    "        \"simulator_kwargs\": {\n",
    "            \"outcsvcols\": \"id1,id2,E,E_norm,subseqDB,Pu1,Pu2\"\n",
    "        }\n",
    "    },\n",
    "    \"mutations_args\": {\n",
    "        \"algorithm\": \"all\",\n",
    "        \"mutation_counts\": 0,\n",
    "        \"mutation_nums_within_sequence\": [1],\n",
    "        \"mutation_nums_per_position\": 1,\n",
    "        \"concurrent_species_to_mutate\": \"single_species_at_a_time\",\n",
    "        \"seed\": 0\n",
    "    },\n",
    "    \"signal\": {\n",
    "        \"inputs\": [data['Interactor 1'].iloc[0]],\n",
    "        \"outputs\": [],\n",
    "        \"function_name\": \"step_function\",\n",
    "        \"function_kwargs\": {\n",
    "            \"impulse_center\": 1,\n",
    "            \"impulse_halfwidth\": 5,\n",
    "            \"target\": 2\n",
    "        }\n",
    "    },\n",
    "    \"simulation\": {\n",
    "        \"dt0\": 0.1,\n",
    "        \"t0\": 0,\n",
    "        \"t1\": 100,\n",
    "        \"tmax\": 2000,\n",
    "        \"solver\": \"diffrax\",\n",
    "        \"use_batch_mutations\": True,\n",
    "        \"interaction_factor\": 1,\n",
    "        \"batch_size\": 200,\n",
    "        \"max_circuits\": 2000,\n",
    "        \"device\": \"cpu\",\n",
    "        \"threshold_steady_states\": 0.05,\n",
    "        \"use_rate_scaling\": True\n",
    "    },\n",
    "    \"molecular_params_factor\": 1,\n",
    "    \"molecular_params\": {\n",
    "        \"avg_mRNA_per_cell\": 100,\n",
    "        \"cell_doubling_time\": 1200,\n",
    "        \"creation_rate\": 2.35,\n",
    "        \"starting_copynumbers\": 200,\n",
    "        \"degradation_rate\": 0.01175,\n",
    "        \"association_binding_rate\": 1000000\n",
    "    }\n",
    "}\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Simulate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [],
   "source": [
    "# data_writer = None\n",
    "# config, data_writer = script_preamble(config, data_writer)\n",
    "# config_file = expand_config(config=config)\n",
    "# config_file = prepare_config(config_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_3392914/599768138.py:14: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  data[f'Binding position {i}'] = list(map(lambda raw: flatten_listlike(\n",
      "/tmp/ipykernel_3392914/599768138.py:14: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  data[f'Binding position {i}'] = list(map(lambda raw: flatten_listlike(\n"
     ]
    }
   ],
   "source": [
    "# Load previously simulated IntaRNA script\n",
    "fn = './data/05_orthogonality_mRNA_official/2024_01_04_105924/inter_data_raw.json'\n",
    "sim_data_l = load_json_as_dict(fn)\n",
    "default_vals = {\n",
    "    'id1': '', 'id2': '', 'E': 0.0, 'E_norm': 0.0, 'subseqDB': '', 'bpList': '', 'hybridDPfull': '', 'Pu1': 0.0, 'Pu2': 0.0, 'seedPu1': '', 'seedPu2': '', 'seedStart1': '', 'seedStart2': '', 'seedEnd1': '', 'seedEnd2': ''\n",
    "}\n",
    "for s, v in sim_data_l.items():\n",
    "    for s2 in v:\n",
    "        if v[s2] == {}:\n",
    "            v[s2].update(default_vals)\n",
    "\n",
    "pattern = r'\\s*\\([^)]*\\)'\n",
    "for i in [1, 2]:\n",
    "    data[f'Binding position {i}'] = list(map(lambda raw: flatten_listlike(\n",
    "        [[[int(re.sub(pattern, '', bp)) for bp in bps.split('..')]\n",
    "          for bps in re.sub(pattern, '', par).split('; ')] for par in raw if par.split('..')[0].strip('-').isnumeric()]), data[f'Binding position {i}'].apply(str).str.split(\"'\")))\n",
    "\n",
    "\n",
    "def get_srnas(data):\n",
    "    srnas, mrnas = [], []\n",
    "    filt_srna1 = data['Category 1'].isin(['sRNA', 'trans-encoded antisense RNA', 'protein binding small RNA'])\n",
    "    filt_srna2 = data['Category 2'].isin(['sRNA', 'trans-encoded antisense RNA', 'protein binding small RNA']) & (~data['Category 1'].isin(['Protein']))\n",
    "    for i in [1, 2]:\n",
    "        filt = filt_srna1 if i == 1 else filt_srna2\n",
    "        for s in data[f'Interactor {i}'].unique():\n",
    "            if s in data[data[f'Category {i}'] == 'mRNA'][f'Interactor {i}'].unique():\n",
    "                mrnas.append(s)\n",
    "            elif s in data[filt][f'Interactor {i}'].unique():\n",
    "                srnas.append(s)\n",
    "    return srnas, mrnas\n",
    "            \n",
    "srnas, mrnas = get_srnas(data)\n",
    "sim_data_l = {k: v for k, v in sim_data_l.items() if k in srnas}\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Mutate\n",
    "\n",
    "Only mutate the parts of the sequence that are already binding, otherwise the whole simulation will take way too long.\n",
    "\n",
    "First preprocess the binding sites curated from the interaction databases. If there are no documented binding "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'id1': 'ecnB',\n",
       " 'id2': 'dsrA',\n",
       " 'E': '-2.56',\n",
       " 'E_norm': '-0.270718',\n",
       " 'subseqDB': '39GGUGCUU&35AAGUGCU',\n",
       " 'Pu1': '0.0835363',\n",
       " 'Pu2': '0.0745675'}"
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "k1 = list(sim_data_l.keys())[0]\n",
    "k2 = list(sim_data_l[k1].keys())[0]\n",
    "sim_data_l[k1][k2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(42, 267)"
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(sim_data_l), len(sim_data_l[k1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sim_data_l.keys() == (set(sim_data_l.keys()) - set(sim_data_l[k1].keys()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sim_data_l[k1].keys() == (set(sim_data_l[k1].keys()) - set(sim_data_l.keys()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [],
   "source": [
    "def int_from_string_nums(text):\n",
    "    return int(''.join(re.findall(r'\\d', text)))\n",
    "\n",
    "def str_nonum(text):\n",
    "    return re.sub(r'\\d', '', text)\n",
    "\n",
    "k1s = []\n",
    "k2s = []\n",
    "for k1, v in sim_data_l.items():\n",
    "    if k1 not in data[['Interactor 1', 'Interactor 2']].melt()['value'].to_list():\n",
    "        k1s.append(k1)\n",
    "    for k2, vv in v.items():\n",
    "        if k2 not in data[['Interactor 1', 'Interactor 2']].melt()['value'].to_list():\n",
    "            k2s.append(k2)\n",
    "        for i, vvv in enumerate(vv['subseqDB'].split('&')):\n",
    "            dna_start = int_from_string_nums(vvv)\n",
    "            sim_data_l[k1][k2][f'binding_range_id{i + 1}'] = (dna_start, dna_start + len(str_nonum(vvv)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [],
   "source": [
    "for k1 in k1s:\n",
    "    del sim_data_l[k1]\n",
    "\n",
    "for k1 in sim_data_l:\n",
    "    for k2 in k2s:\n",
    "        if k2 in sim_data_l[k1]:\n",
    "            del sim_data_l[k1][k2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def get_labels(vsim, data):\n",
    "    assert vsim['id1'] in data['Interactor 1'].to_list() + data['Interactor 2'].to_list(), 'id1 is not in data'\n",
    "    assert vsim['id2'] in data['Interactor 1'].to_list() + data['Interactor 2'].to_list(), 'id2 is not in data'\n",
    "    i = 1 if (vsim['id1'] in data['Interactor 1'].to_list()) else 2\n",
    "    j = 2 if (vsim['id2'] in data['Interactor 2'].to_list()) else 1\n",
    "    if i == j:\n",
    "        # print(f'Interactions {i} == {j} for species {vsim[\"id1\"]} and {vsim[\"id2\"]}')\n",
    "        pass\n",
    "    ri = data[(data[f'Interactor {i}'] == vsim['id1'])]\n",
    "    rj = data[(data[f'Interactor {j}'] == vsim['id2'])]\n",
    "    r = data[(data[f'Interactor {i}'] == vsim['id1']) & (data[f'Interactor {j}'] == vsim['id2'])]\n",
    "    id_srna = 1 if (ri[f'Category {i}'] != 'mRNA').any() else 2\n",
    "    id_mrna = 2 if (rj[f'Category {j}'] == 'mRNA').any() else 1\n",
    "    # assert r[f'Category {j}'] != 'mRNA', 'Label the sRNA better'\n",
    "    if id_srna == id_mrna:\n",
    "        # print(f'The sRNA and mRNA ids are the same: {i} == {j} for species {vsim[\"id1\"]} and {vsim[\"id2\"]}')\n",
    "        pass\n",
    "    return i, j, r, ri, rj, id_srna, id_mrna\n",
    "\n",
    "\n",
    "def make_seqs(i, j, r, ri, rj):\n",
    "    rs = [ri, rj]\n",
    "    seqs = [''] * 2\n",
    "    for idx, p in enumerate([i, j]):\n",
    "        seqs[idx] = rs[idx][f'Sequence {p}'].iloc[0]\n",
    "        if r[f'Binding position {p}'].any():\n",
    "            mins = r[f'Binding position {p}'].apply(lambda x: np.min(x) if x else 1)\n",
    "            if mins.min() < 0:\n",
    "                seqs[idx] = r[f'Sequence before {p}'].iloc[np.argmin(mins)] + seqs[idx]\n",
    "        seqs[idx] = seqs[idx].upper().replace('T', 'U')\n",
    "    return seqs\n",
    "\n",
    "\n",
    "def make_mutations(circuits, seqs, vsim, id_srna, id_mrna):\n",
    "    mutations = {k: [] for k in ['mutation_name', 'mutation_types', 'positions']}\n",
    "    for i, mutation_pos in enumerate(np.arange(vsim[f'binding_range_id{id_srna}'][0], vsim[f'binding_range_id{id_srna}'][1])):\n",
    "        i_mut = 0\n",
    "        for nuc in ['A', 'C', 'G', 'U']:\n",
    "            if seqs[id_srna - 1][mutation_pos] == nuc:\n",
    "                continue\n",
    "            mutation_name = 'mutation_' + str(i) + '-' + str(i_mut)\n",
    "            circuits[vsim[f'id{id_srna}']][vsim[f'id{id_mrna}']][mutation_name] = {\n",
    "                vsim[f'id{id_srna}']: seqs[id_srna - 1][:mutation_pos] + nuc + seqs[id_srna - 1][mutation_pos+1:],\n",
    "                vsim[f'id{id_mrna}']: seqs[id_mrna - 1]\n",
    "            }\n",
    "            i_mut += 1\n",
    "            mutations['mutation_name'].append(mutation_name)\n",
    "            mutations['mutation_types'].append([get_mutation_type_mapping('RNA')[seqs[id_srna - 1][mutation_pos]][nuc]])\n",
    "            mutations['positions'].append([mutation_pos])\n",
    "    mutations = pd.DataFrame.from_dict(mutations)\n",
    "    mutations['count'] = 1\n",
    "    mutations['template_file'] = ''\n",
    "    return circuits, mutations\n",
    "\n",
    "circuits_all = {}\n",
    "mutations_all = {}\n",
    "for k1, v in sim_data_l.items():\n",
    "    for k2, vsim in v.items():\n",
    "        i, j, r, ri, rj, id_srna, id_mrna = get_labels(vsim, data)\n",
    "        if id_srna == id_mrna:\n",
    "            continue\n",
    "        circuits_all.setdefault(vsim[f'id{id_srna}'], {})\n",
    "        \n",
    "        seqs = make_seqs(i, j, r, ri, rj)\n",
    "        mutation_name = 'ref_circuit'\n",
    "        circuits_all[vsim[f'id{id_srna}']].setdefault(vsim[f'id{id_mrna}'], {})\n",
    "        circuits_all[vsim[f'id{id_srna}']][vsim[f'id{id_mrna}']][mutation_name] = {\n",
    "            vsim[f'id{id_srna}']: seqs[id_srna - 1],\n",
    "            vsim[f'id{id_mrna}']: seqs[id_mrna - 1]\n",
    "        }\n",
    "        circuits_all, mutations = make_mutations(circuits_all, seqs, vsim, id_srna, id_mrna)\n",
    "        mutations_all.setdefault(vsim[f'id{id_srna}'], {})\n",
    "        mutations_all[vsim[f'id{id_srna}']].setdefault(vsim[f'id{id_mrna}'], {})\n",
    "        mutations_all[vsim[f'id{id_srna}']][vsim[f'id{id_mrna}']] = mutations\n",
    "        break\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys(['ref_circuit', 'mutation_0-0', 'mutation_0-1', 'mutation_0-2', 'mutation_1-0', 'mutation_1-1', 'mutation_1-2', 'mutation_2-0', 'mutation_2-1', 'mutation_2-2', 'mutation_3-0', 'mutation_3-1', 'mutation_3-2', 'mutation_4-0', 'mutation_4-1', 'mutation_4-2', 'mutation_5-0', 'mutation_5-1', 'mutation_5-2', 'mutation_6-0', 'mutation_6-1', 'mutation_6-2', 'mutation_7-0', 'mutation_7-1', 'mutation_7-2'])"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "circuits_all['AfaR']['dsrB'].keys()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Make circuits"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "5"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "circuits = []\n",
    "for srna, d_mrna in circuits_all.items():\n",
    "    for mrna, muts in d_mrna.items():\n",
    "        circuit_name = srna + '_' + mrna\n",
    "        mutations = mutations_all[srna][mrna]\n",
    "        mutations['circuit_name'] = circuit_name\n",
    "        data_writer.output(data=mutations, out_name='mutations',\n",
    "                           out_type='csv', subfolder=circuit_name)\n",
    "\n",
    "        for mutation_name, d_mutation in muts.items():\n",
    "\n",
    "            config_file['data'] = d_mutation\n",
    "            config_file['name'] = mutation_name\n",
    "            config_file = expand_model_config(\n",
    "                config_file, config_file, sample_names=[srna, mrna])\n",
    "\n",
    "            kwargs = compose_kwargs(prev_configs=None, config=config_file)\n",
    "            kwargs['name'] = circuit_name\n",
    "            circuit = instantiate_system(kwargs)\n",
    "            circuit.subname = mutation_name\n",
    "            circuits.append(circuit)\n",
    "\n",
    "len(circuit.model.species)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total circuits:  32794\n"
     ]
    }
   ],
   "source": [
    "# for circuit in circuits:\n",
    "# num_subcircuits = [len(flatten_nested_dict(\n",
    "#     c.mutations)) + 1 for c in circuits_all]\n",
    "# tot_subcircuits = sum(num_subcircuits)\n",
    "tot_subcircuits = len(circuits)\n",
    "\n",
    "print('Total circuits: ', tot_subcircuits)\n",
    "# print(num_subcircuits)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of different interactions: 43\n",
      "Estimated time:  11.386805555555556 days\n",
      "Estimated time with multi:  0.4744502314814815 days\n"
     ]
    }
   ],
   "source": [
    "# A simulation where RNAs have lengths [1071, 828] (aroF, YagJ) takes 20s\n",
    "# A simulation where RNAs have lengths [1122, 828] (aroF, YagJ) takes 28s\n",
    "# A simulation where RNAs have lengths [1758, 828] (fliC, YagJ) takes 29s\n",
    "\n",
    "n_segs = 24\n",
    "ave_sim_per_inter = 30 #s\n",
    "\n",
    "print('Number of different interactions:', len(mutations_all))\n",
    "print('Estimated time: ', tot_subcircuits * ave_sim_per_inter / 60 / 60 / 24, 'days')\n",
    "print('Estimated time with multi: ', tot_subcircuits * ave_sim_per_inter / 60 / 60 / 24 / n_segs, 'days')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'name': 'Esr41',\n",
       " 'physical_data': 'GAUGCUCUAGGCAUCACAUUUUCUCCAUGGGGUAUUCCCUCCGCCGGCACGAUGUGUUGCUGGCGUUUUUUUAU'}"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "i = 0\n",
    "# circuits = circuits[i*int(len(circuits)/n_segs):(i+1)*int(len(circuits)/n_segs)]\n",
    "# circuits = circuits[:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # {c.name: (len(c.get_input_species()[0].physical_data), len(c.get_input_species()[1].physical_data)) for c in circuits}\n",
    "# data[data['Interactor 2'] == 'fliC']['Sequence 2'].iloc[0]\n",
    "# {c.name: list(len(s.physical_data) for s in c.get_input_species()) for c in circuits}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# from synbio_morpher.utils.common.setup import compose_kwargs\n",
    "# modeller = CircuitModeller(result_writer=data_writer, config=config_file)\n",
    "# curr_subcircuits = []\n",
    "# for i, circuit in enumerate(circuits[:3]):\n",
    "#     curr_subcircuits.append(modeller.load_mutations(circuit))\n",
    "# curr_subcircuits = flatten_listlike(curr_subcircuits)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['ref_circuit', 'mutation_0-0', 'mutation_0-1', 'mutation_0-2']"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[c.subname for c in circuits[:4]]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Run sim"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def simulate_IntaRNA_local(query: dict,\n",
    "                           fn_targets: str,\n",
    "                           sim_kwargs={}):\n",
    "    fn1 = data_writer.output(data=query, out_type='fasta', out_name='query', byseq=True, return_path=True)\n",
    "    sim_kwargs['query'] = fn1\n",
    "    sim_kwargs['target'] = fn_targets\n",
    "    \n",
    "    def run(query: str, target: str, qidxpos0: int, tidxpos0: int, outcsvcols: str, threads: int, n: int = 1,\n",
    "            param_file: str = '', extra_params: list = [], raw_stdout: bool = False):\n",
    "        p = Popen(['IntaRNA', '-q', query, '-t', target,\n",
    "                    '--outMode=C', f'--outcsvcols={outcsvcols}',\n",
    "                    f'--qIdxPos0={qidxpos0}',\n",
    "                    f'--tIdxPos0={tidxpos0}',\n",
    "                    f'--outNumber={n}',\n",
    "                    f'--threads={threads}', param_file]\n",
    "                    + extra_params, stdout=PIPE, stderr=PIPE, universal_newlines=True)\n",
    "        stdout, stderr = p.communicate()\n",
    "        return process_raw_stdout(stdout)\n",
    "    \n",
    "    return run(**sim_kwargs)\n",
    "\n",
    "\n",
    "fn_targets = data_writer.output(data=mrnas, out_type='fasta', out_name='target', byseq=True, return_path=True)\n",
    "sim_data = {}\n",
    "for s, sseq in srnas.items():\n",
    "    sim_data[s] = {}\n",
    "    t = datetime.now()\n",
    "\n",
    "    sim_data[s] = simulate_IntaRNA_local(query={s: sseq},\n",
    "                                         fn_targets=fn_targets,\n",
    "                                         sim_kwargs=config['interaction_simulator']['simulator_kwargs'])\n",
    "    logging.warning('Finished ', s, ' in ', (datetime.now() - t).total_seconds(), ' s')\n",
    "\n",
    "    data_writer.output(data=sim_data, out_type='json',\n",
    "                       out_name='inter_data_raw', overwrite=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "circuit_manager.py:batch_circuits():692: \tFrom 4 circuits, a total of 4 mutated circuits will be simulated. WARNING\n",
      "circuit_manager.py:batch_circuits():699: \t\tStarting new round of viable circuits (0 - 4 / 4) WARNING\n",
      "circuit_manager.py:batch_circuits():717: \t\tMaking subcircuits 0 - 4 took 4.2e-05s WARNING\n",
      "circuit_manager.py:batch_circuits():723: \tBatching 0 - 200 circuits (out of 0 - 4 (total: 4)) (Circuits: 0 - 4 of 4) WARNING\n",
      "circuit_manager.py:run_batch():755: \t\tRunning 4 Subcircuits - dsrA_fhlA: compute_interactions WARNING\n"
     ]
    }
   ],
   "source": [
    "# modeller = CircuitModeller(result_writer=data_writer, config=config_file)\n",
    "# circuits_sim = modeller.batch_circuits(\n",
    "#     circuits=circuits[4:8],\n",
    "#     write_to_subsystem=True, batch_size=config_file['simulation'].get('batch_size', 100),\n",
    "#     methods={\n",
    "#         \"compute_interactions\": {},\n",
    "#         # \"compute_interactions_batch\": {'batch': True},\n",
    "#         # \"write_interactions\": {},\n",
    "#         # \"init_circuits\": {'batch': True},\n",
    "#         # \"simulate_signal_batch\": {'ref_circuit': None,\n",
    "#         #                           'batch': True},\n",
    "#         \"write_results\": {'no_visualisations': config_file['experiment'].get('no_visualisations', True),\n",
    "#                           'no_numerical': config_file['experiment'].get('no_numerical', False)}\n",
    "#     })"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Save"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mutation_name</th>\n",
       "      <th>mutation_types</th>\n",
       "      <th>positions</th>\n",
       "      <th>count</th>\n",
       "      <th>template_file</th>\n",
       "      <th>circuit_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>mutation_0-0</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[39]</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>dsrA_fhlA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>mutation_0-1</td>\n",
       "      <td>[4]</td>\n",
       "      <td>[39]</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>dsrA_fhlA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>mutation_0-2</td>\n",
       "      <td>[5]</td>\n",
       "      <td>[39]</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>dsrA_fhlA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>mutation_1-0</td>\n",
       "      <td>[9]</td>\n",
       "      <td>[40]</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>dsrA_fhlA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>mutation_1-1</td>\n",
       "      <td>[10]</td>\n",
       "      <td>[40]</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>dsrA_fhlA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>mutation_1-2</td>\n",
       "      <td>[11]</td>\n",
       "      <td>[40]</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>dsrA_fhlA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>mutation_2-0</td>\n",
       "      <td>[9]</td>\n",
       "      <td>[41]</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>dsrA_fhlA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>mutation_2-1</td>\n",
       "      <td>[10]</td>\n",
       "      <td>[41]</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>dsrA_fhlA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>mutation_2-2</td>\n",
       "      <td>[11]</td>\n",
       "      <td>[41]</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>dsrA_fhlA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>mutation_3-0</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[42]</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>dsrA_fhlA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>mutation_3-1</td>\n",
       "      <td>[4]</td>\n",
       "      <td>[42]</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>dsrA_fhlA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>mutation_3-2</td>\n",
       "      <td>[5]</td>\n",
       "      <td>[42]</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>dsrA_fhlA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>mutation_4-0</td>\n",
       "      <td>[9]</td>\n",
       "      <td>[43]</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>dsrA_fhlA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>mutation_4-1</td>\n",
       "      <td>[10]</td>\n",
       "      <td>[43]</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>dsrA_fhlA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>mutation_4-2</td>\n",
       "      <td>[11]</td>\n",
       "      <td>[43]</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>dsrA_fhlA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>mutation_5-0</td>\n",
       "      <td>[9]</td>\n",
       "      <td>[44]</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>dsrA_fhlA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>mutation_5-1</td>\n",
       "      <td>[10]</td>\n",
       "      <td>[44]</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>dsrA_fhlA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>mutation_5-2</td>\n",
       "      <td>[11]</td>\n",
       "      <td>[44]</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>dsrA_fhlA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>mutation_6-0</td>\n",
       "      <td>[6]</td>\n",
       "      <td>[45]</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>dsrA_fhlA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>mutation_6-1</td>\n",
       "      <td>[7]</td>\n",
       "      <td>[45]</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>dsrA_fhlA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>mutation_6-2</td>\n",
       "      <td>[8]</td>\n",
       "      <td>[45]</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>dsrA_fhlA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   mutation_name mutation_types positions  count  template_file circuit_name\n",
       "0   mutation_0-0            [3]      [39]      1            NaN    dsrA_fhlA\n",
       "1   mutation_0-1            [4]      [39]      1            NaN    dsrA_fhlA\n",
       "2   mutation_0-2            [5]      [39]      1            NaN    dsrA_fhlA\n",
       "3   mutation_1-0            [9]      [40]      1            NaN    dsrA_fhlA\n",
       "4   mutation_1-1           [10]      [40]      1            NaN    dsrA_fhlA\n",
       "5   mutation_1-2           [11]      [40]      1            NaN    dsrA_fhlA\n",
       "6   mutation_2-0            [9]      [41]      1            NaN    dsrA_fhlA\n",
       "7   mutation_2-1           [10]      [41]      1            NaN    dsrA_fhlA\n",
       "8   mutation_2-2           [11]      [41]      1            NaN    dsrA_fhlA\n",
       "9   mutation_3-0            [3]      [42]      1            NaN    dsrA_fhlA\n",
       "10  mutation_3-1            [4]      [42]      1            NaN    dsrA_fhlA\n",
       "11  mutation_3-2            [5]      [42]      1            NaN    dsrA_fhlA\n",
       "12  mutation_4-0            [9]      [43]      1            NaN    dsrA_fhlA\n",
       "13  mutation_4-1           [10]      [43]      1            NaN    dsrA_fhlA\n",
       "14  mutation_4-2           [11]      [43]      1            NaN    dsrA_fhlA\n",
       "15  mutation_5-0            [9]      [44]      1            NaN    dsrA_fhlA\n",
       "16  mutation_5-1           [10]      [44]      1            NaN    dsrA_fhlA\n",
       "17  mutation_5-2           [11]      [44]      1            NaN    dsrA_fhlA\n",
       "18  mutation_6-0            [6]      [45]      1            NaN    dsrA_fhlA\n",
       "19  mutation_6-1            [7]      [45]      1            NaN    dsrA_fhlA\n",
       "20  mutation_6-2            [8]      [45]      1            NaN    dsrA_fhlA"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fn = 'data/tests/2024_01_20_041436/dsrA_fhlA/mutations.csv'\n",
    "m = pd.read_csv(fn)\n",
    "m"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "array length 349 does not match index length 357",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[16], line 3\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;66;03m# source_dirs = [data_writer.top_write_dir]\u001b[39;00m\n\u001b[1;32m      2\u001b[0m source_dirs \u001b[38;5;241m=\u001b[39m [\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mdata/tests/2024_01_18_175210/\u001b[39m\u001b[38;5;124m'\u001b[39m]\n\u001b[0;32m----> 3\u001b[0m tables \u001b[38;5;241m=\u001b[39m \u001b[43mload_multiple_as_list\u001b[49m\u001b[43m(\u001b[49m\u001b[43minputs_list\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43msource_dirs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mload_func\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mb_tabulate_mutation_info\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\n\u001b[1;32m      4\u001b[0m \u001b[43m            \u001b[49m\u001b[43mdata_writer\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mdata_writer\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m      5\u001b[0m saves \u001b[38;5;241m=\u001b[39m pd\u001b[38;5;241m.\u001b[39mconcat(tables, axis\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m0\u001b[39m)\n\u001b[1;32m      6\u001b[0m saves\n",
      "File \u001b[0;32m~/Olive/env_evo/src/synbio-morpher/synbio_morpher/utils/data/data_format_tools/common.py:100\u001b[0m, in \u001b[0;36mload_multiple_as_list\u001b[0;34m(inputs_list, load_func, **kwargs)\u001b[0m\n\u001b[1;32m     97\u001b[0m \u001b[38;5;28;01massert\u001b[39;00m \u001b[38;5;28mtype\u001b[39m(\n\u001b[1;32m     98\u001b[0m     inputs_list) \u001b[38;5;241m==\u001b[39m \u001b[38;5;28mlist\u001b[39m, \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mInput type of \u001b[39m\u001b[38;5;132;01m{\u001b[39;00minputs_list\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m should be list and not \u001b[39m\u001b[38;5;132;01m{\u001b[39;00m\u001b[38;5;28mtype\u001b[39m(inputs_list)\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m'\u001b[39m\n\u001b[1;32m     99\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m inp \u001b[38;5;129;01min\u001b[39;00m inputs_list:\n\u001b[0;32m--> 100\u001b[0m     collection_list\u001b[38;5;241m.\u001b[39mappend(\u001b[43mload_func\u001b[49m\u001b[43m(\u001b[49m\u001b[43minp\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m)\n\u001b[1;32m    101\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m collection_list\n",
      "File \u001b[0;32m~/Olive/env_evo/src/synbio-morpher/synbio_morpher/srv/sequence_exploration/sequence_analysis.py:314\u001b[0m, in \u001b[0;36mb_tabulate_mutation_info\u001b[0;34m(source_dir, data_writer, experiment_config)\u001b[0m\n\u001b[1;32m    310\u001b[0m     interaction_stats \u001b[38;5;241m=\u001b[39m make_interaction_stats_and_sample_names(\n\u001b[1;32m    311\u001b[0m         mutation_dirs, experiment_config\u001b[38;5;241m=\u001b[39mexperiment_config)[\u001b[38;5;241m0\u001b[39m]\n\u001b[1;32m    313\u001b[0m     \u001b[38;5;66;03m# Mutated circuits\u001b[39;00m\n\u001b[0;32m--> 314\u001b[0m     mutation_table \u001b[38;5;241m=\u001b[39m \u001b[43mpd\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mDataFrame\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mfrom_dict\u001b[49m\u001b[43m(\u001b[49m\u001b[43m{\u001b[49m\n\u001b[1;32m    315\u001b[0m \u001b[43m        \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mcircuit_name\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[43m[\u001b[49m\u001b[43mcircuit_name\u001b[49m\u001b[43m]\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;28;43mlen\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43mmutation_dirs\u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    316\u001b[0m \u001b[43m        \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mmutation_name\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[43mmutations\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mmutation_name\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    317\u001b[0m \u001b[43m        \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mmutation_num\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[43mmutations\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mcount\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    318\u001b[0m \u001b[43m        \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mmutation_type\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[43mcast_astype\u001b[49m\u001b[43m(\u001b[49m\u001b[43mmutations\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mmutation_types\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mint\u001b[39;49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    319\u001b[0m \u001b[43m        \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mmutation_positions\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[43mcast_astype\u001b[49m\u001b[43m(\u001b[49m\u001b[43mmutations\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mpositions\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mint\u001b[39;49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    320\u001b[0m \u001b[43m        \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mpath_to_template_circuit\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[43mmutations\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mtemplate_file\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    321\u001b[0m \u001b[43m        \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43msample_num\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mlen\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43mall_sample_names\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    322\u001b[0m \u001b[43m    \u001b[49m\u001b[43m}\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    323\u001b[0m     current_og_table[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mpath_to_template_circuit\u001b[39m\u001b[38;5;124m'\u001b[39m] \u001b[38;5;241m=\u001b[39m mutations[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mtemplate_file\u001b[39m\u001b[38;5;124m'\u001b[39m]\u001b[38;5;241m.\u001b[39miloc[\u001b[38;5;241m0\u001b[39m]\n\u001b[1;32m    325\u001b[0m \u001b[38;5;66;03m# Expand the interaction keys in the table\u001b[39;00m\n",
      "File \u001b[0;32m~/Olive/env_evo/lib/python3.8/site-packages/pandas/core/frame.py:1763\u001b[0m, in \u001b[0;36mDataFrame.from_dict\u001b[0;34m(cls, data, orient, dtype, columns)\u001b[0m\n\u001b[1;32m   1757\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\n\u001b[1;32m   1758\u001b[0m         \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mExpected \u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mindex\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m, \u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mcolumns\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m or \u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mtight\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m for orient parameter. \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m   1759\u001b[0m         \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mGot \u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;132;01m{\u001b[39;00morient\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m instead\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m   1760\u001b[0m     )\n\u001b[1;32m   1762\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m orient \u001b[38;5;241m!=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mtight\u001b[39m\u001b[38;5;124m\"\u001b[39m:\n\u001b[0;32m-> 1763\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mcls\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43mdata\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mindex\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mindex\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcolumns\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mcolumns\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mdtype\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mdtype\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1764\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m   1765\u001b[0m     realdata \u001b[38;5;241m=\u001b[39m data[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mdata\u001b[39m\u001b[38;5;124m\"\u001b[39m]\n",
      "File \u001b[0;32m~/Olive/env_evo/lib/python3.8/site-packages/pandas/core/frame.py:663\u001b[0m, in \u001b[0;36mDataFrame.__init__\u001b[0;34m(self, data, index, columns, dtype, copy)\u001b[0m\n\u001b[1;32m    657\u001b[0m     mgr \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_init_mgr(\n\u001b[1;32m    658\u001b[0m         data, axes\u001b[38;5;241m=\u001b[39m{\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mindex\u001b[39m\u001b[38;5;124m\"\u001b[39m: index, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mcolumns\u001b[39m\u001b[38;5;124m\"\u001b[39m: columns}, dtype\u001b[38;5;241m=\u001b[39mdtype, copy\u001b[38;5;241m=\u001b[39mcopy\n\u001b[1;32m    659\u001b[0m     )\n\u001b[1;32m    661\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(data, \u001b[38;5;28mdict\u001b[39m):\n\u001b[1;32m    662\u001b[0m     \u001b[38;5;66;03m# GH#38939 de facto copy defaults to False only in non-dict cases\u001b[39;00m\n\u001b[0;32m--> 663\u001b[0m     mgr \u001b[38;5;241m=\u001b[39m \u001b[43mdict_to_mgr\u001b[49m\u001b[43m(\u001b[49m\u001b[43mdata\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mindex\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcolumns\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mdtype\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mdtype\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcopy\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mcopy\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mtyp\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mmanager\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    664\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(data, ma\u001b[38;5;241m.\u001b[39mMaskedArray):\n\u001b[1;32m    665\u001b[0m     \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01mnumpy\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mma\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mmrecords\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m \u001b[38;5;21;01mmrecords\u001b[39;00m\n",
      "File \u001b[0;32m~/Olive/env_evo/lib/python3.8/site-packages/pandas/core/internals/construction.py:493\u001b[0m, in \u001b[0;36mdict_to_mgr\u001b[0;34m(data, index, columns, dtype, typ, copy)\u001b[0m\n\u001b[1;32m    489\u001b[0m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    490\u001b[0m         \u001b[38;5;66;03m# dtype check to exclude e.g. range objects, scalars\u001b[39;00m\n\u001b[1;32m    491\u001b[0m         arrays \u001b[38;5;241m=\u001b[39m [x\u001b[38;5;241m.\u001b[39mcopy() \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mhasattr\u001b[39m(x, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mdtype\u001b[39m\u001b[38;5;124m\"\u001b[39m) \u001b[38;5;28;01melse\u001b[39;00m x \u001b[38;5;28;01mfor\u001b[39;00m x \u001b[38;5;129;01min\u001b[39;00m arrays]\n\u001b[0;32m--> 493\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43marrays_to_mgr\u001b[49m\u001b[43m(\u001b[49m\u001b[43marrays\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcolumns\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mindex\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mdtype\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mdtype\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mtyp\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mtyp\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mconsolidate\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mcopy\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/Olive/env_evo/lib/python3.8/site-packages/pandas/core/internals/construction.py:118\u001b[0m, in \u001b[0;36marrays_to_mgr\u001b[0;34m(arrays, columns, index, dtype, verify_integrity, typ, consolidate)\u001b[0m\n\u001b[1;32m    115\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m verify_integrity:\n\u001b[1;32m    116\u001b[0m     \u001b[38;5;66;03m# figure out the index, if necessary\u001b[39;00m\n\u001b[1;32m    117\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m index \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[0;32m--> 118\u001b[0m         index \u001b[38;5;241m=\u001b[39m \u001b[43m_extract_index\u001b[49m\u001b[43m(\u001b[49m\u001b[43marrays\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    119\u001b[0m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    120\u001b[0m         index \u001b[38;5;241m=\u001b[39m ensure_index(index)\n",
      "File \u001b[0;32m~/Olive/env_evo/lib/python3.8/site-packages/pandas/core/internals/construction.py:680\u001b[0m, in \u001b[0;36m_extract_index\u001b[0;34m(data)\u001b[0m\n\u001b[1;32m    675\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m lengths[\u001b[38;5;241m0\u001b[39m] \u001b[38;5;241m!=\u001b[39m \u001b[38;5;28mlen\u001b[39m(index):\n\u001b[1;32m    676\u001b[0m         msg \u001b[38;5;241m=\u001b[39m (\n\u001b[1;32m    677\u001b[0m             \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124marray length \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mlengths[\u001b[38;5;241m0\u001b[39m]\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m does not match index \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    678\u001b[0m             \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mlength \u001b[39m\u001b[38;5;132;01m{\u001b[39;00m\u001b[38;5;28mlen\u001b[39m(index)\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    679\u001b[0m         )\n\u001b[0;32m--> 680\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(msg)\n\u001b[1;32m    681\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    682\u001b[0m     index \u001b[38;5;241m=\u001b[39m default_index(lengths[\u001b[38;5;241m0\u001b[39m])\n",
      "\u001b[0;31mValueError\u001b[0m: array length 349 does not match index length 357"
     ]
    }
   ],
   "source": [
    "# source_dirs = [data_writer.top_write_dir]\n",
    "source_dirs = ['data/tests/2024_01_18_175210/']\n",
    "tables = load_multiple_as_list(inputs_list=source_dirs, load_func=b_tabulate_mutation_info, \n",
    "            data_writer=data_writer)\n",
    "saves = pd.concat(tables, axis=0)\n",
    "saves"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Compare"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number simulated: 5\n"
     ]
    }
   ],
   "source": [
    "# print(f'Number simulated: {len(saves)}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "env_evo",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
