{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Comparison of mutational robustness of sRNA's vs. randomly generated RNA"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "xla_bridge.py:backends():603: Unable to initialize backend 'rocm': NOT_FOUND: Could not find registered platform with name: \"rocm\". Available platform names are: Interpreter CUDA INFO\n",
      "xla_bridge.py:backends():603: Unable to initialize backend 'tpu': module 'jaxlib.xla_extension' has no attribute 'get_tpu_client' INFO\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[CpuDevice(id=0)]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import os\n",
    "from Bio import SeqIO\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "import pandas as pd\n",
    "\n",
    "from synbio_morpher.srv.io.manage.script_manager import script_preamble\n",
    "from synbio_morpher.srv.sequence_exploration.sequence_analysis import b_tabulate_mutation_info\n",
    "from synbio_morpher.utils.common.setup import construct_circuit_from_cfg, prepare_config, expand_config\n",
    "from synbio_morpher.utils.evolution.evolver import Evolver\n",
    "from synbio_morpher.utils.circuit.agnostic_circuits.circuit_manager import CircuitModeller\n",
    "from synbio_morpher.utils.misc.type_handling import flatten_listlike\n",
    "from synbio_morpher.utils.data.data_format_tools.common import load_multiple_as_list\n",
    "from synbio_morpher.utils.misc.type_handling import flatten_nested_dict\n",
    "\n",
    "\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import jax\n",
    "jax.config.update('jax_platform_name', 'cpu')\n",
    "\n",
    "jax.devices()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "fn = '../data/sRNA/merged_inter.csv'\n",
    "data = pd.read_csv(fn)\n",
    "try:\n",
    "    data = data.drop(columns=['Unnamed: 0'])\n",
    "except:\n",
    "    pass\n",
    "data = data[data['Sequence 1'].apply(lambda x: type(x) == str)]\n",
    "data = data[data['Sequence 2'].apply(lambda x: type(x) == str)]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Interactor 1</th>\n",
       "      <th>Interactor 2</th>\n",
       "      <th>ID 1</th>\n",
       "      <th>ID 2</th>\n",
       "      <th>Category 1</th>\n",
       "      <th>Category 2</th>\n",
       "      <th>Regulation</th>\n",
       "      <th>Binding position 1</th>\n",
       "      <th>Binding position 2</th>\n",
       "      <th>Sequence 1</th>\n",
       "      <th>Sequence 2</th>\n",
       "      <th>Source DB</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>micA</td>\n",
       "      <td>dppA</td>\n",
       "      <td>NCBI:2847697</td>\n",
       "      <td>NCBI:948062</td>\n",
       "      <td>sRNA</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CAAAAATGGTGTCGCTCGTATTATTGTCGAGAAACCTTTCGGCAAG...</td>\n",
       "      <td>ATGCGTATTTCCTTGAAAAAGTCAGGGATGCTGAAGCTTGGTCTCA...</td>\n",
       "      <td>RNAInter</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>dsrA</td>\n",
       "      <td>dsrB</td>\n",
       "      <td>NCBI:946470</td>\n",
       "      <td>NCBI:946468</td>\n",
       "      <td>sRNA</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AACACATCAGATTTCCTGGTGTAACGAATTTTTTAAGTGCTTCTTG...</td>\n",
       "      <td>TAAATGNAAACAAGTTTTATTTTTATCATCAGAACATTCTAATAAA...</td>\n",
       "      <td>RNAInter</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>dsrA</td>\n",
       "      <td>ilvH</td>\n",
       "      <td>NCBI:946470</td>\n",
       "      <td>NCBI:947267</td>\n",
       "      <td>sRNA</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AACACATCAGATTTCCTGGTGTAACGAATTTTTTAAGTGCTTCTTG...</td>\n",
       "      <td>ATTCGCCACAACAAGGACAGGAAGGTTCGGCGCAAGGAGCCCAAGA...</td>\n",
       "      <td>RNAInter</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>micA</td>\n",
       "      <td>ecnB</td>\n",
       "      <td>NCBI:2847697</td>\n",
       "      <td>NCBI:2847737</td>\n",
       "      <td>sRNA</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CAAAAATGGTGTCGCTCGTATTATTGTCGAGAAACCTTTCGGCAAG...</td>\n",
       "      <td>ATGGTGAAGAAGACAATTGCAGCGATCTTTTCTGTTCTGGTGCTTT...</td>\n",
       "      <td>RNAInter</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>oxyS</td>\n",
       "      <td>fhlA</td>\n",
       "      <td>NCBI:2847701</td>\n",
       "      <td>NCBI:947181</td>\n",
       "      <td>sRNA</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GAAACGGAGCGGCACCTCTTTTAACCCTTGAAGTCACTGCCCGTTT...</td>\n",
       "      <td>CACCAGATTGCCAAGCTGATGGTCGAGCTGTCCAAATCCCAGGATG...</td>\n",
       "      <td>RNAInter</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>444</th>\n",
       "      <td>ryhB</td>\n",
       "      <td>msrA</td>\n",
       "      <td>chromosome:NC_000913.3, Gene ID:2847761</td>\n",
       "      <td>chromosome:NC_000913.3, Gene ID:948734</td>\n",
       "      <td>NaN</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>No Interaction</td>\n",
       "      <td>['NA']</td>\n",
       "      <td>['NA']</td>\n",
       "      <td>GCGATCAGGAAGACCCTCGCGGAGAACCTGAAAGCACGACATTGCT...</td>\n",
       "      <td>ATGAGTTTATTTGATAAAAAGCATCTGGTTTCCCCCGCCGATGCCC...</td>\n",
       "      <td>sRNATarBase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>445</th>\n",
       "      <td>gcvB</td>\n",
       "      <td>hdeA</td>\n",
       "      <td>chromosome:NC_000913.3, Gene ID:2847720</td>\n",
       "      <td>chromosome:NC_000913.3, Gene ID:948025</td>\n",
       "      <td>trans-encoded antisense RNA</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>Induction</td>\n",
       "      <td>['NA']</td>\n",
       "      <td>['NA']</td>\n",
       "      <td>ACTTCCTGAGCCGGAACGAAAAGTTTTATCGGAATGCGTGTTCTGG...</td>\n",
       "      <td>ATGAAAAAAGTATTAGGCGTTATTCTTGGTGGTCTGCTTCTTCTGC...</td>\n",
       "      <td>sRNATarBase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>446</th>\n",
       "      <td>gcvB</td>\n",
       "      <td>hdeB</td>\n",
       "      <td>chromosome:NC_000913.3, Gene ID:2847720</td>\n",
       "      <td>chromosome:NC_000913.3, Gene ID:948026</td>\n",
       "      <td>trans-encoded antisense RNA</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>Induction</td>\n",
       "      <td>['NA']</td>\n",
       "      <td>['NA']</td>\n",
       "      <td>ACTTCCTGAGCCGGAACGAAAAGTTTTATCGGAATGCGTGTTCTGG...</td>\n",
       "      <td>ATGAATATTTCATCTCTCCGTAAAGCGTTTATTTTTATGGGCGCTG...</td>\n",
       "      <td>sRNATarBase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>447</th>\n",
       "      <td>Esr41</td>\n",
       "      <td>fliC</td>\n",
       "      <td>chromosome:NC_002695.1, Gene ID:</td>\n",
       "      <td>chromosome:NC_002695.1, Gene ID:</td>\n",
       "      <td>trans-encoded antisense RNA</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>Induction</td>\n",
       "      <td>['NA']</td>\n",
       "      <td>['NA']</td>\n",
       "      <td>GATGCTCTAGGCATCACATTTTCTCCATGGGGTATTCCCTCCGCCG...</td>\n",
       "      <td>ATGGCACAAGTCATTAATACCAACAGCCTCTCGCTGATCACTCAAA...</td>\n",
       "      <td>sRNATarBase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>448</th>\n",
       "      <td>dsrA</td>\n",
       "      <td>mreB</td>\n",
       "      <td>chromosome:NC_000913.3, Gene ID:946470</td>\n",
       "      <td>chromosome:NC_000913.3, Gene ID:948588</td>\n",
       "      <td>trans-encoded antisense RNA</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>Repression</td>\n",
       "      <td>['24..41']</td>\n",
       "      <td>['-3..18']</td>\n",
       "      <td>AACACATCAGATTTCCTGGTGTAACGAATTTTTTAAGTGCTTCTTG...</td>\n",
       "      <td>ATGTTGAAAAAATTTCGTGGCATGTTTTCCAATGACTTGTCCATTG...</td>\n",
       "      <td>sRNATarBase</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>446 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Interactor 1 Interactor 2                                     ID 1  \\\n",
       "0           micA         dppA                             NCBI:2847697   \n",
       "1           dsrA         dsrB                              NCBI:946470   \n",
       "2           dsrA         ilvH                              NCBI:946470   \n",
       "3           micA         ecnB                             NCBI:2847697   \n",
       "4           oxyS         fhlA                             NCBI:2847701   \n",
       "..           ...          ...                                      ...   \n",
       "444         ryhB         msrA  chromosome:NC_000913.3, Gene ID:2847761   \n",
       "445         gcvB         hdeA  chromosome:NC_000913.3, Gene ID:2847720   \n",
       "446         gcvB         hdeB  chromosome:NC_000913.3, Gene ID:2847720   \n",
       "447        Esr41         fliC         chromosome:NC_002695.1, Gene ID:   \n",
       "448         dsrA         mreB   chromosome:NC_000913.3, Gene ID:946470   \n",
       "\n",
       "                                       ID 2                   Category 1  \\\n",
       "0                               NCBI:948062                         sRNA   \n",
       "1                               NCBI:946468                         sRNA   \n",
       "2                               NCBI:947267                         sRNA   \n",
       "3                              NCBI:2847737                         sRNA   \n",
       "4                               NCBI:947181                         sRNA   \n",
       "..                                      ...                          ...   \n",
       "444  chromosome:NC_000913.3, Gene ID:948734                          NaN   \n",
       "445  chromosome:NC_000913.3, Gene ID:948025  trans-encoded antisense RNA   \n",
       "446  chromosome:NC_000913.3, Gene ID:948026  trans-encoded antisense RNA   \n",
       "447        chromosome:NC_002695.1, Gene ID:  trans-encoded antisense RNA   \n",
       "448  chromosome:NC_000913.3, Gene ID:948588  trans-encoded antisense RNA   \n",
       "\n",
       "    Category 2      Regulation Binding position 1 Binding position 2  \\\n",
       "0         mRNA             NaN                NaN                NaN   \n",
       "1         mRNA             NaN                NaN                NaN   \n",
       "2         mRNA             NaN                NaN                NaN   \n",
       "3         mRNA             NaN                NaN                NaN   \n",
       "4         mRNA             NaN                NaN                NaN   \n",
       "..         ...             ...                ...                ...   \n",
       "444       mRNA  No Interaction             ['NA']             ['NA']   \n",
       "445       mRNA       Induction             ['NA']             ['NA']   \n",
       "446       mRNA       Induction             ['NA']             ['NA']   \n",
       "447       mRNA       Induction             ['NA']             ['NA']   \n",
       "448       mRNA      Repression         ['24..41']         ['-3..18']   \n",
       "\n",
       "                                            Sequence 1  \\\n",
       "0    CAAAAATGGTGTCGCTCGTATTATTGTCGAGAAACCTTTCGGCAAG...   \n",
       "1    AACACATCAGATTTCCTGGTGTAACGAATTTTTTAAGTGCTTCTTG...   \n",
       "2    AACACATCAGATTTCCTGGTGTAACGAATTTTTTAAGTGCTTCTTG...   \n",
       "3    CAAAAATGGTGTCGCTCGTATTATTGTCGAGAAACCTTTCGGCAAG...   \n",
       "4    GAAACGGAGCGGCACCTCTTTTAACCCTTGAAGTCACTGCCCGTTT...   \n",
       "..                                                 ...   \n",
       "444  GCGATCAGGAAGACCCTCGCGGAGAACCTGAAAGCACGACATTGCT...   \n",
       "445  ACTTCCTGAGCCGGAACGAAAAGTTTTATCGGAATGCGTGTTCTGG...   \n",
       "446  ACTTCCTGAGCCGGAACGAAAAGTTTTATCGGAATGCGTGTTCTGG...   \n",
       "447  GATGCTCTAGGCATCACATTTTCTCCATGGGGTATTCCCTCCGCCG...   \n",
       "448  AACACATCAGATTTCCTGGTGTAACGAATTTTTTAAGTGCTTCTTG...   \n",
       "\n",
       "                                            Sequence 2    Source DB  \n",
       "0    ATGCGTATTTCCTTGAAAAAGTCAGGGATGCTGAAGCTTGGTCTCA...     RNAInter  \n",
       "1    TAAATGNAAACAAGTTTTATTTTTATCATCAGAACATTCTAATAAA...     RNAInter  \n",
       "2    ATTCGCCACAACAAGGACAGGAAGGTTCGGCGCAAGGAGCCCAAGA...     RNAInter  \n",
       "3    ATGGTGAAGAAGACAATTGCAGCGATCTTTTCTGTTCTGGTGCTTT...     RNAInter  \n",
       "4    CACCAGATTGCCAAGCTGATGGTCGAGCTGTCCAAATCCCAGGATG...     RNAInter  \n",
       "..                                                 ...          ...  \n",
       "444  ATGAGTTTATTTGATAAAAAGCATCTGGTTTCCCCCGCCGATGCCC...  sRNATarBase  \n",
       "445  ATGAAAAAAGTATTAGGCGTTATTCTTGGTGGTCTGCTTCTTCTGC...  sRNATarBase  \n",
       "446  ATGAATATTTCATCTCTCCGTAAAGCGTTTATTTTTATGGGCGCTG...  sRNATarBase  \n",
       "447  ATGGCACAAGTCATTAATACCAACAGCCTCTCGCTGATCACTCAAA...  sRNATarBase  \n",
       "448  ATGTTGAAAAAATTTCGTGGCATGTTTTCCAATGACTTGTCCATTG...  sRNATarBase  \n",
       "\n",
       "[446 rows x 12 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "446"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(data)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Config"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "config = {\n",
    "    \"experiment\": {\n",
    "        \"purpose\": \"tests\",\n",
    "        \"no_visualisations\": False,\n",
    "        \"no_numerical\": False,\n",
    "        \"debug_mode\": False\n",
    "    },\n",
    "    \"system_type\": \"RNA\",\n",
    "    \"include_prod_deg\": False,\n",
    "    \"interaction_simulator\": {\n",
    "        \"name\": \"IntaRNA\",\n",
    "        \"postprocess\": True,\n",
    "        \"simulator_kwargs\": {\n",
    "            \"outcsvcols\": \"id1, id2, E, E_norm, seedPu1, seedPu2, seedStart1, seedStart2, seedEnd1, seedEnd2\",\n",
    "            \"threads\": 12,\n",
    "            \"n\": 1,\n",
    "            \"raw_stdout\": True\n",
    "        }\n",
    "    },\n",
    "    \"mutations_args\": {\n",
    "        \"algorithm\": \"all\",\n",
    "        \"mutation_counts\": 0,\n",
    "        \"mutation_nums_within_sequence\": [1],\n",
    "        \"mutation_nums_per_position\": 1,\n",
    "        \"concurrent_species_to_mutate\": \"single_species_at_a_time\",\n",
    "        \"seed\": 0\n",
    "    },\n",
    "    \"signal\": {\n",
    "        \"inputs\": [data['Interactor 1'].iloc[0]],\n",
    "        \"outputs\": [],\n",
    "        \"function_name\": \"step_function\",\n",
    "        \"function_kwargs\": {\n",
    "            \"impulse_center\": 1,\n",
    "            \"impulse_halfwidth\": 5,\n",
    "            \"target\": 2\n",
    "        }\n",
    "    },\n",
    "    \"simulation\": {\n",
    "        \"dt0\": 0.1,\n",
    "        \"t0\": 0,\n",
    "        \"t1\": 100,\n",
    "        \"tmax\": 2000,\n",
    "        \"solver\": \"diffrax\",\n",
    "        \"use_batch_mutations\": True,\n",
    "        \"interaction_factor\": 1,\n",
    "        \"batch_size\": 2000,\n",
    "        \"max_circuits\": 9000,\n",
    "        \"device\": \"cpu\",\n",
    "        \"threshold_steady_states\": 0.05,\n",
    "        \"use_rate_scaling\": True\n",
    "    },\n",
    "    \"molecular_params_factor\": 1,\n",
    "    \"molecular_params\": {\n",
    "        \"avg_mRNA_per_cell\": 100,\n",
    "        \"cell_doubling_time\": 1200,\n",
    "        \"creation_rate\": 2.35,\n",
    "        \"starting_copynumbers\": 200,\n",
    "        \"degradation_rate\": 0.01175,\n",
    "        \"association_binding_rate\": 1000000\n",
    "    }\n",
    "}\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Simulate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_writer = None\n",
    "config, data_writer = script_preamble(config, data_writer)\n",
    "config_file = expand_config(config=config)\n",
    "config_file = prepare_config(config_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "''"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Evolver(data_writer=data_writer, sequence_type=config_file.get('system_type'),\n",
    "                          seed=config_file.get('mutations_args', {}).get('seed', 0)).concurrent_species_to_mutate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "5"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "circuits = []\n",
    "for sp in data.iterrows():\n",
    "    sp = sp[1]\n",
    "    config_file['mutations_args']['concurrent_species_to_mutate'] = sp['Interactor 1']\n",
    "    # config_file['data_path'] = 'data/circuit.fasta'\n",
    "    config_file['data'] = {\n",
    "        sp['Interactor 1']: sp['Sequence 1'].replace('T', 'U'),\n",
    "        sp['Interactor 2']: sp['Sequence 2'].replace('T', 'U')\n",
    "    }\n",
    "    circuit = construct_circuit_from_cfg(\n",
    "        prev_configs=None, config_file=config_file)\n",
    "    circuit = Evolver(data_writer=data_writer, sequence_type=config_file.get('system_type'),\n",
    "                      seed=config_file['mutations_args'].get('seed', 0),\n",
    "                      concurrent_species_to_mutate=config_file['mutations_args']['concurrent_species_to_mutate']).mutate(\n",
    "        circuit,\n",
    "        write_to_subsystem=True, algorithm=config_file.get('mutations_args', {}).get('algorithm', 'random'))\n",
    "    circuits.append(circuit)\n",
    "\n",
    "len(circuit.model.species)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[433]\n"
     ]
    }
   ],
   "source": [
    "# for circuit in circuits:\n",
    "#     num_subcircuits = [len(flatten_nested_dict(\n",
    "#         c.mutations)) + 1 for c in [circuit]]\n",
    "#     tot_subcircuits = sum(num_subcircuits)\n",
    "\n",
    "# print(num_subcircuits)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# from synbio_morpher.utils.common.setup import compose_kwargs\n",
    "# modeller = CircuitModeller(result_writer=data_writer, config=config_file)\n",
    "# curr_subcircuits = []\n",
    "# for i, circuit in enumerate(circuits[:3]):\n",
    "#     curr_subcircuits.append(modeller.load_mutations(circuit))\n",
    "# curr_subcircuits = flatten_listlike(curr_subcircuits)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "circuit_manager.py:batch_circuits():600: \tFrom 69 circuits, a total of 28890 mutated circuits will be simulated. WARNING\n",
      "circuit_manager.py:batch_circuits():607: \t\tStarting new round of viable circuits (0 - 7 / 69) WARNING\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "circuit_manager.py:batch_circuits():625: \t\tMaking subcircuits 0 - 2029 took 216.816958s WARNING\n",
      "circuit_manager.py:batch_circuits():631: \tBatching 0 - 2000 circuits (out of 0 - 2029 (total: 28890)) (Circuits: 0 - 7 of 69) WARNING\n",
      "circuit_manager.py:run_batch():662: \t\tRunning 2000 Subcircuits - merged: compute_interactions WARNING\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[1;32m/workdir/notebooks_sRNA/05_distribution_mutational_robustness.ipynb Cell 18\u001b[0m line \u001b[0;36m2\n\u001b[1;32m      <a href='vscode-notebook-cell://attached-container%2B7b22636f6e7461696e65724e616d65223a222f676367227d@ssh-remote%2B7b22686f73744e616d65223a224d75666173612d5453227d/workdir/notebooks_sRNA/05_distribution_mutational_robustness.ipynb#X34sdnNjb2RlLXJlbW90ZQ%3D%3D?line=0'>1</a>\u001b[0m modeller \u001b[39m=\u001b[39m CircuitModeller(result_writer\u001b[39m=\u001b[39mdata_writer, config\u001b[39m=\u001b[39mconfig_file)\n\u001b[0;32m----> <a href='vscode-notebook-cell://attached-container%2B7b22636f6e7461696e65724e616d65223a222f676367227d@ssh-remote%2B7b22686f73744e616d65223a224d75666173612d5453227d/workdir/notebooks_sRNA/05_distribution_mutational_robustness.ipynb#X34sdnNjb2RlLXJlbW90ZQ%3D%3D?line=1'>2</a>\u001b[0m circuits \u001b[39m=\u001b[39m modeller\u001b[39m.\u001b[39;49mbatch_circuits(\n\u001b[1;32m      <a href='vscode-notebook-cell://attached-container%2B7b22636f6e7461696e65724e616d65223a222f676367227d@ssh-remote%2B7b22686f73744e616d65223a224d75666173612d5453227d/workdir/notebooks_sRNA/05_distribution_mutational_robustness.ipynb#X34sdnNjb2RlLXJlbW90ZQ%3D%3D?line=2'>3</a>\u001b[0m     circuits\u001b[39m=\u001b[39;49mcircuits,\n\u001b[1;32m      <a href='vscode-notebook-cell://attached-container%2B7b22636f6e7461696e65724e616d65223a222f676367227d@ssh-remote%2B7b22686f73744e616d65223a224d75666173612d5453227d/workdir/notebooks_sRNA/05_distribution_mutational_robustness.ipynb#X34sdnNjb2RlLXJlbW90ZQ%3D%3D?line=3'>4</a>\u001b[0m     write_to_subsystem\u001b[39m=\u001b[39;49m\u001b[39mTrue\u001b[39;49;00m, batch_size\u001b[39m=\u001b[39;49mconfig_file[\u001b[39m'\u001b[39;49m\u001b[39msimulation\u001b[39;49m\u001b[39m'\u001b[39;49m]\u001b[39m.\u001b[39;49mget(\u001b[39m'\u001b[39;49m\u001b[39mbatch_size\u001b[39;49m\u001b[39m'\u001b[39;49m, \u001b[39m100\u001b[39;49m),\n\u001b[1;32m      <a href='vscode-notebook-cell://attached-container%2B7b22636f6e7461696e65724e616d65223a222f676367227d@ssh-remote%2B7b22686f73744e616d65223a224d75666173612d5453227d/workdir/notebooks_sRNA/05_distribution_mutational_robustness.ipynb#X34sdnNjb2RlLXJlbW90ZQ%3D%3D?line=4'>5</a>\u001b[0m     methods\u001b[39m=\u001b[39;49m{\n\u001b[1;32m      <a href='vscode-notebook-cell://attached-container%2B7b22636f6e7461696e65724e616d65223a222f676367227d@ssh-remote%2B7b22686f73744e616d65223a224d75666173612d5453227d/workdir/notebooks_sRNA/05_distribution_mutational_robustness.ipynb#X34sdnNjb2RlLXJlbW90ZQ%3D%3D?line=5'>6</a>\u001b[0m         \u001b[39m\"\u001b[39;49m\u001b[39mcompute_interactions\u001b[39;49m\u001b[39m\"\u001b[39;49m: {},\n\u001b[1;32m      <a href='vscode-notebook-cell://attached-container%2B7b22636f6e7461696e65724e616d65223a222f676367227d@ssh-remote%2B7b22686f73744e616d65223a224d75666173612d5453227d/workdir/notebooks_sRNA/05_distribution_mutational_robustness.ipynb#X34sdnNjb2RlLXJlbW90ZQ%3D%3D?line=6'>7</a>\u001b[0m         \u001b[39m# \"init_circuits\": {'batch': True},\u001b[39;49;00m\n\u001b[1;32m      <a href='vscode-notebook-cell://attached-container%2B7b22636f6e7461696e65724e616d65223a222f676367227d@ssh-remote%2B7b22686f73744e616d65223a224d75666173612d5453227d/workdir/notebooks_sRNA/05_distribution_mutational_robustness.ipynb#X34sdnNjb2RlLXJlbW90ZQ%3D%3D?line=7'>8</a>\u001b[0m         \u001b[39m# \"simulate_signal_batch\": {'ref_circuit': None,\u001b[39;49;00m\n\u001b[1;32m      <a href='vscode-notebook-cell://attached-container%2B7b22636f6e7461696e65724e616d65223a222f676367227d@ssh-remote%2B7b22686f73744e616d65223a224d75666173612d5453227d/workdir/notebooks_sRNA/05_distribution_mutational_robustness.ipynb#X34sdnNjb2RlLXJlbW90ZQ%3D%3D?line=8'>9</a>\u001b[0m         \u001b[39m#                           'batch': True},\u001b[39;49;00m\n\u001b[1;32m     <a href='vscode-notebook-cell://attached-container%2B7b22636f6e7461696e65724e616d65223a222f676367227d@ssh-remote%2B7b22686f73744e616d65223a224d75666173612d5453227d/workdir/notebooks_sRNA/05_distribution_mutational_robustness.ipynb#X34sdnNjb2RlLXJlbW90ZQ%3D%3D?line=9'>10</a>\u001b[0m         \u001b[39m\"\u001b[39;49m\u001b[39mwrite_results\u001b[39;49m\u001b[39m\"\u001b[39;49m: {\u001b[39m'\u001b[39;49m\u001b[39mno_visualisations\u001b[39;49m\u001b[39m'\u001b[39;49m: config_file[\u001b[39m'\u001b[39;49m\u001b[39mexperiment\u001b[39;49m\u001b[39m'\u001b[39;49m]\u001b[39m.\u001b[39;49mget(\u001b[39m'\u001b[39;49m\u001b[39mno_visualisations\u001b[39;49m\u001b[39m'\u001b[39;49m, \u001b[39mTrue\u001b[39;49;00m),\n\u001b[1;32m     <a href='vscode-notebook-cell://attached-container%2B7b22636f6e7461696e65724e616d65223a222f676367227d@ssh-remote%2B7b22686f73744e616d65223a224d75666173612d5453227d/workdir/notebooks_sRNA/05_distribution_mutational_robustness.ipynb#X34sdnNjb2RlLXJlbW90ZQ%3D%3D?line=10'>11</a>\u001b[0m                           \u001b[39m'\u001b[39;49m\u001b[39mno_numerical\u001b[39;49m\u001b[39m'\u001b[39;49m: config_file[\u001b[39m'\u001b[39;49m\u001b[39mexperiment\u001b[39;49m\u001b[39m'\u001b[39;49m]\u001b[39m.\u001b[39;49mget(\u001b[39m'\u001b[39;49m\u001b[39mno_numerical\u001b[39;49m\u001b[39m'\u001b[39;49m, \u001b[39mFalse\u001b[39;49;00m)}\n\u001b[1;32m     <a href='vscode-notebook-cell://attached-container%2B7b22636f6e7461696e65724e616d65223a222f676367227d@ssh-remote%2B7b22686f73744e616d65223a224d75666173612d5453227d/workdir/notebooks_sRNA/05_distribution_mutational_robustness.ipynb#X34sdnNjb2RlLXJlbW90ZQ%3D%3D?line=11'>12</a>\u001b[0m     })\n",
      "File \u001b[0;32m/usr/local/lib/python3.10/dist-packages/synbio_morpher/utils/circuit/agnostic_circuits/circuit_manager.py:639\u001b[0m, in \u001b[0;36mCircuitModeller.batch_circuits\u001b[0;34m(self, circuits, methods, batch_size, include_normal_run, write_to_subsystem)\u001b[0m\n\u001b[1;32m    637\u001b[0m     \u001b[39mif\u001b[39;00m \u001b[39mnot\u001b[39;00m b_circuits:\n\u001b[1;32m    638\u001b[0m         \u001b[39mcontinue\u001b[39;00m\n\u001b[0;32m--> 639\u001b[0m     ref_circuit \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mrun_batch(\n\u001b[1;32m    640\u001b[0m         b_circuits, methods, leading_ref_circuit\u001b[39m=\u001b[39;49mref_circuit,\n\u001b[1;32m    641\u001b[0m         include_normal_run\u001b[39m=\u001b[39;49minclude_normal_run,\n\u001b[1;32m    642\u001b[0m         write_to_subsystem\u001b[39m=\u001b[39;49mwrite_to_subsystem)\n\u001b[1;32m    644\u001b[0m single_batch_time \u001b[39m=\u001b[39m datetime\u001b[39m.\u001b[39mnow() \u001b[39m-\u001b[39m single_batch_time\n\u001b[1;32m    645\u001b[0m logging\u001b[39m.\u001b[39mwarning(\n\u001b[1;32m    646\u001b[0m     \u001b[39mf\u001b[39m\u001b[39m'\u001b[39m\u001b[39mSingle batch: \u001b[39m\u001b[39m{\u001b[39;00msingle_batch_time\u001b[39m}\u001b[39;00m\u001b[39m \u001b[39m\u001b[39m\\n\u001b[39;00m\u001b[39mProjected time: \u001b[39m\u001b[39m{\u001b[39;00msingle_batch_time\u001b[39m.\u001b[39mtotal_seconds()\u001b[39m \u001b[39m\u001b[39m*\u001b[39m\u001b[39m \u001b[39m\u001b[39mlen\u001b[39m(subcircuits)\u001b[39m/\u001b[39mtot_subcircuits\u001b[39m}\u001b[39;00m\u001b[39ms \u001b[39m\u001b[39m\\n\u001b[39;00m\u001b[39mTotal time: \u001b[39m\u001b[39m{\u001b[39;00m\u001b[39mstr\u001b[39m(datetime\u001b[39m.\u001b[39mnow()\u001b[39m \u001b[39m\u001b[39m-\u001b[39m\u001b[39m \u001b[39mstart_time)\u001b[39m}\u001b[39;00m\u001b[39m'\u001b[39m)\n",
      "File \u001b[0;32m/usr/local/lib/python3.10/dist-packages/synbio_morpher/utils/circuit/agnostic_circuits/circuit_manager.py:691\u001b[0m, in \u001b[0;36mCircuitModeller.run_batch\u001b[0;34m(self, subcircuits, methods, leading_ref_circuit, include_normal_run, write_to_subsystem)\u001b[0m\n\u001b[1;32m    688\u001b[0m         \u001b[39mcontinue\u001b[39;00m\n\u001b[1;32m    690\u001b[0m a_time \u001b[39m=\u001b[39m datetime\u001b[39m.\u001b[39mnow()\n\u001b[0;32m--> 691\u001b[0m subcircuit \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mapply_to_circuit(\n\u001b[1;32m    692\u001b[0m     subcircuit, {method: kwargs})\n\u001b[1;32m    693\u001b[0m a_time \u001b[39m=\u001b[39m datetime\u001b[39m.\u001b[39mnow() \u001b[39m-\u001b[39m a_time\n\u001b[1;32m    694\u001b[0m subcircuits[i] \u001b[39m=\u001b[39m subcircuit\n",
      "File \u001b[0;32m/usr/local/lib/python3.10/dist-packages/synbio_morpher/utils/circuit/agnostic_circuits/circuit_manager.py:710\u001b[0m, in \u001b[0;36mCircuitModeller.apply_to_circuit\u001b[0;34m(self, circuit, methods)\u001b[0m\n\u001b[1;32m    708\u001b[0m \u001b[39mfor\u001b[39;00m method, kwargs \u001b[39min\u001b[39;00m methods\u001b[39m.\u001b[39mitems():\n\u001b[1;32m    709\u001b[0m     \u001b[39mif\u001b[39;00m \u001b[39mhasattr\u001b[39m(\u001b[39mself\u001b[39m, method):\n\u001b[0;32m--> 710\u001b[0m         circuit \u001b[39m=\u001b[39m \u001b[39mgetattr\u001b[39;49m(\u001b[39mself\u001b[39;49m, method)(circuit, \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49mkwargs)\n\u001b[1;32m    711\u001b[0m     \u001b[39melse\u001b[39;00m:\n\u001b[1;32m    712\u001b[0m         logging\u001b[39m.\u001b[39mwarning(\n\u001b[1;32m    713\u001b[0m             \u001b[39mf\u001b[39m\u001b[39m'\u001b[39m\u001b[39mCould not find method @\u001b[39m\u001b[39m{\u001b[39;00mmethod\u001b[39m}\u001b[39;00m\u001b[39m in class \u001b[39m\u001b[39m{\u001b[39;00m\u001b[39mself\u001b[39m\u001b[39m}\u001b[39;00m\u001b[39m'\u001b[39m)\n",
      "File \u001b[0;32m/usr/local/lib/python3.10/dist-packages/synbio_morpher/utils/circuit/agnostic_circuits/circuit_manager.py:101\u001b[0m, in \u001b[0;36mCircuitModeller.compute_interactions\u001b[0;34m(self, circuit)\u001b[0m\n\u001b[1;32m     98\u001b[0m input_species \u001b[39m=\u001b[39m circuit\u001b[39m.\u001b[39mget_input_species()\n\u001b[1;32m     99\u001b[0m reactant_species \u001b[39m=\u001b[39m [\n\u001b[1;32m    100\u001b[0m     r\u001b[39m.\u001b[39mspecies \u001b[39mfor\u001b[39;00m r \u001b[39min\u001b[39;00m circuit\u001b[39m.\u001b[39mqreactions\u001b[39m.\u001b[39mreactants]\n\u001b[0;32m--> 101\u001b[0m interactions \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mrun_interaction_simulator(\n\u001b[1;32m    102\u001b[0m     species\u001b[39m=\u001b[39;49minput_species,\n\u001b[1;32m    103\u001b[0m     filename\u001b[39m=\u001b[39;49mfilename,\n\u001b[1;32m    104\u001b[0m     quantities\u001b[39m=\u001b[39;49m[circuit\u001b[39m.\u001b[39;49mqreactions\u001b[39m.\u001b[39;49mreactants[reactant_species\u001b[39m.\u001b[39;49mindex(s)]\u001b[39m.\u001b[39;49mquantity \u001b[39mfor\u001b[39;49;00m s \u001b[39min\u001b[39;49;00m input_species])\n\u001b[1;32m    105\u001b[0m circuit\u001b[39m.\u001b[39minteractions \u001b[39m=\u001b[39m interactions\n\u001b[1;32m    106\u001b[0m circuit\u001b[39m.\u001b[39minteractions_state \u001b[39m=\u001b[39m \u001b[39m'\u001b[39m\u001b[39mcomputed\u001b[39m\u001b[39m'\u001b[39m\n",
      "File \u001b[0;32m/usr/local/lib/python3.10/dist-packages/synbio_morpher/utils/circuit/agnostic_circuits/circuit_manager.py:137\u001b[0m, in \u001b[0;36mCircuitModeller.run_interaction_simulator\u001b[0;34m(self, species, quantities, filename)\u001b[0m\n\u001b[1;32m    133\u001b[0m data \u001b[39m=\u001b[39m {s: s\u001b[39m.\u001b[39mphysical_data \u001b[39mfor\u001b[39;00m s \u001b[39min\u001b[39;00m species}\n\u001b[1;32m    134\u001b[0m \u001b[39m# if filename is not None:\u001b[39;00m\n\u001b[1;32m    135\u001b[0m \u001b[39m#     return self.interaction_simulator.run((filename, data), compute_by_filename=True)\u001b[39;00m\n\u001b[1;32m    136\u001b[0m \u001b[39m# else:\u001b[39;00m\n\u001b[0;32m--> 137\u001b[0m \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49minteraction_simulator\u001b[39m.\u001b[39;49mrun(data, quantities\u001b[39m=\u001b[39;49mquantities, compute_by_filename\u001b[39m=\u001b[39;49m\u001b[39mFalse\u001b[39;49;00m)\n",
      "File \u001b[0;32m/usr/local/lib/python3.10/dist-packages/synbio_morpher/srv/parameter_prediction/interactions.py:221\u001b[0m, in \u001b[0;36mInteractionSimulator.run\u001b[0;34m(self, input, quantities, compute_by_filename)\u001b[0m\n\u001b[1;32m    217\u001b[0m \u001b[39m\u001b[39m\u001b[39m\"\"\" Makes nested dictionary for querying interactions as \u001b[39;00m\n\u001b[1;32m    218\u001b[0m \u001b[39m{sample1: {sample2: interaction}} \"\"\"\u001b[39;00m\n\u001b[1;32m    219\u001b[0m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mdata_handler \u001b[39m=\u001b[39m InteractionDataHandler(\n\u001b[1;32m    220\u001b[0m     simulation_handler\u001b[39m=\u001b[39m\u001b[39mself\u001b[39m\u001b[39m.\u001b[39msimulation_handler, quantities\u001b[39m=\u001b[39mquantities)\n\u001b[0;32m--> 221\u001b[0m data \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49msimulator(\u001b[39minput\u001b[39;49m, compute_by_filename\u001b[39m=\u001b[39;49mcompute_by_filename)\n\u001b[1;32m    222\u001b[0m data \u001b[39m=\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mdata_handler\u001b[39m.\u001b[39minit_data(data)\n\u001b[1;32m    223\u001b[0m \u001b[39mreturn\u001b[39;00m data\n",
      "File \u001b[0;32m/usr/local/lib/python3.10/dist-packages/synbio_morpher/srv/parameter_prediction/simulator.py:142\u001b[0m, in \u001b[0;36msimulate_IntaRNA\u001b[0;34m(input, compute_by_filename, allow_self_interaction, sim_kwargs, simulator)\u001b[0m\n\u001b[1;32m    140\u001b[0m \u001b[39melse\u001b[39;00m:\n\u001b[1;32m    141\u001b[0m     f \u001b[39m=\u001b[39m simulate_IntaRNA_data\n\u001b[0;32m--> 142\u001b[0m \u001b[39mreturn\u001b[39;00m f(\u001b[39minput\u001b[39;49m, allow_self_interaction\u001b[39m=\u001b[39;49mallow_self_interaction, sim_kwargs\u001b[39m=\u001b[39;49msim_kwargs, simulator\u001b[39m=\u001b[39;49msimulator)\n",
      "File \u001b[0;32m/usr/local/lib/python3.10/dist-packages/synbio_morpher/srv/parameter_prediction/simulator.py:162\u001b[0m, in \u001b[0;36msimulate_IntaRNA_data\u001b[0;34m(batch, allow_self_interaction, sim_kwargs, simulator)\u001b[0m\n\u001b[1;32m    160\u001b[0m                 sim_kwargs[\u001b[39m\"\u001b[39m\u001b[39mquery\u001b[39m\u001b[39m\"\u001b[39m] \u001b[39m=\u001b[39m sample_i\n\u001b[1;32m    161\u001b[0m                 sim_kwargs[\u001b[39m\"\u001b[39m\u001b[39mtarget\u001b[39m\u001b[39m\"\u001b[39m] \u001b[39m=\u001b[39m sample_j\n\u001b[0;32m--> 162\u001b[0m                 current_pair[label_j] \u001b[39m=\u001b[39m simulator\u001b[39m.\u001b[39;49mrun(\u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49msim_kwargs)\n\u001b[1;32m    163\u001b[0m         data[label_i] \u001b[39m=\u001b[39m current_pair\n\u001b[1;32m    164\u001b[0m \u001b[39melse\u001b[39;00m:\n",
      "File \u001b[0;32m/usr/local/lib/python3.10/dist-packages/synbio_morpher/srv/parameter_prediction/IntaRNA/bin/copomus/IntaRNA.py:61\u001b[0m, in \u001b[0;36mIntaRNA.run\u001b[0;34m(self, query, target, qidxpos0, tidxpos0, outcsvcols, threads, n, param_file, extra_params, raw_stdout)\u001b[0m\n\u001b[1;32m     52\u001b[0m param_file \u001b[39m=\u001b[39m \u001b[39mf\u001b[39m\u001b[39m'\u001b[39m\u001b[39m--parameterFile=\u001b[39m\u001b[39m{\u001b[39;00mparam_file\u001b[39m}\u001b[39;00m\u001b[39m'\u001b[39m \u001b[39mif\u001b[39;00m param_file \u001b[39melse\u001b[39;00m \u001b[39m'\u001b[39m\u001b[39m'\u001b[39m\n\u001b[1;32m     53\u001b[0m p \u001b[39m=\u001b[39m Popen([\u001b[39m'\u001b[39m\u001b[39mIntaRNA\u001b[39m\u001b[39m'\u001b[39m, \u001b[39m'\u001b[39m\u001b[39m-q\u001b[39m\u001b[39m'\u001b[39m, query, \u001b[39m'\u001b[39m\u001b[39m-t\u001b[39m\u001b[39m'\u001b[39m, target,\n\u001b[1;32m     54\u001b[0m            \u001b[39m'\u001b[39m\u001b[39m--outMode=C\u001b[39m\u001b[39m'\u001b[39m, \u001b[39mf\u001b[39m\u001b[39m'\u001b[39m\u001b[39m--outcsvcols=\u001b[39m\u001b[39m{\u001b[39;00moutcsvcols\u001b[39m}\u001b[39;00m\u001b[39m'\u001b[39m,\n\u001b[1;32m     55\u001b[0m            \u001b[39mf\u001b[39m\u001b[39m'\u001b[39m\u001b[39m--qIdxPos0=\u001b[39m\u001b[39m{\u001b[39;00mqidxpos0\u001b[39m}\u001b[39;00m\u001b[39m'\u001b[39m,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m     58\u001b[0m            \u001b[39mf\u001b[39m\u001b[39m'\u001b[39m\u001b[39m--threads=\u001b[39m\u001b[39m{\u001b[39;00mthreads\u001b[39m}\u001b[39;00m\u001b[39m'\u001b[39m, param_file]\n\u001b[1;32m     59\u001b[0m           \u001b[39m+\u001b[39m extra_params, stdout\u001b[39m=\u001b[39mPIPE, stderr\u001b[39m=\u001b[39mPIPE, universal_newlines\u001b[39m=\u001b[39m\u001b[39mTrue\u001b[39;00m)\n\u001b[0;32m---> 61\u001b[0m stdout, stderr \u001b[39m=\u001b[39m p\u001b[39m.\u001b[39;49mcommunicate()\n\u001b[1;32m     62\u001b[0m \u001b[39mif\u001b[39;00m p\u001b[39m.\u001b[39mreturncode:\n\u001b[1;32m     63\u001b[0m     sys\u001b[39m.\u001b[39mstderr\u001b[39m.\u001b[39mwrite(\u001b[39mf\u001b[39m\u001b[39m'\u001b[39m\u001b[39mIntaRNA Error: \u001b[39m\u001b[39m{\u001b[39;00mstdout\u001b[39m}\u001b[39;00m\u001b[39m \u001b[39m\u001b[39m{\u001b[39;00mstderr\u001b[39m}\u001b[39;00m\u001b[39m\\n\u001b[39;00m\u001b[39m'\u001b[39m)\n",
      "File \u001b[0;32m/usr/lib/python3.10/subprocess.py:1154\u001b[0m, in \u001b[0;36mPopen.communicate\u001b[0;34m(self, input, timeout)\u001b[0m\n\u001b[1;32m   1151\u001b[0m     endtime \u001b[39m=\u001b[39m \u001b[39mNone\u001b[39;00m\n\u001b[1;32m   1153\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[0;32m-> 1154\u001b[0m     stdout, stderr \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_communicate(\u001b[39minput\u001b[39;49m, endtime, timeout)\n\u001b[1;32m   1155\u001b[0m \u001b[39mexcept\u001b[39;00m \u001b[39mKeyboardInterrupt\u001b[39;00m:\n\u001b[1;32m   1156\u001b[0m     \u001b[39m# https://bugs.python.org/issue25942\u001b[39;00m\n\u001b[1;32m   1157\u001b[0m     \u001b[39m# See the detailed comment in .wait().\u001b[39;00m\n\u001b[1;32m   1158\u001b[0m     \u001b[39mif\u001b[39;00m timeout \u001b[39mis\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39mNone\u001b[39;00m:\n",
      "File \u001b[0;32m/usr/lib/python3.10/subprocess.py:2021\u001b[0m, in \u001b[0;36mPopen._communicate\u001b[0;34m(self, input, endtime, orig_timeout)\u001b[0m\n\u001b[1;32m   2014\u001b[0m     \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_check_timeout(endtime, orig_timeout,\n\u001b[1;32m   2015\u001b[0m                         stdout, stderr,\n\u001b[1;32m   2016\u001b[0m                         skip_check_and_raise\u001b[39m=\u001b[39m\u001b[39mTrue\u001b[39;00m)\n\u001b[1;32m   2017\u001b[0m     \u001b[39mraise\u001b[39;00m \u001b[39mRuntimeError\u001b[39;00m(  \u001b[39m# Impossible :)\u001b[39;00m\n\u001b[1;32m   2018\u001b[0m         \u001b[39m'\u001b[39m\u001b[39m_check_timeout(..., skip_check_and_raise=True) \u001b[39m\u001b[39m'\u001b[39m\n\u001b[1;32m   2019\u001b[0m         \u001b[39m'\u001b[39m\u001b[39mfailed to raise TimeoutExpired.\u001b[39m\u001b[39m'\u001b[39m)\n\u001b[0;32m-> 2021\u001b[0m ready \u001b[39m=\u001b[39m selector\u001b[39m.\u001b[39;49mselect(timeout)\n\u001b[1;32m   2022\u001b[0m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_check_timeout(endtime, orig_timeout, stdout, stderr)\n\u001b[1;32m   2024\u001b[0m \u001b[39m# XXX Rewrite these to use non-blocking I/O on the file\u001b[39;00m\n\u001b[1;32m   2025\u001b[0m \u001b[39m# objects; they are no longer using C stdio!\u001b[39;00m\n",
      "File \u001b[0;32m/usr/lib/python3.10/selectors.py:416\u001b[0m, in \u001b[0;36m_PollLikeSelector.select\u001b[0;34m(self, timeout)\u001b[0m\n\u001b[1;32m    414\u001b[0m ready \u001b[39m=\u001b[39m []\n\u001b[1;32m    415\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[0;32m--> 416\u001b[0m     fd_event_list \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_selector\u001b[39m.\u001b[39;49mpoll(timeout)\n\u001b[1;32m    417\u001b[0m \u001b[39mexcept\u001b[39;00m \u001b[39mInterruptedError\u001b[39;00m:\n\u001b[1;32m    418\u001b[0m     \u001b[39mreturn\u001b[39;00m ready\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "# modeller = CircuitModeller(result_writer=data_writer, config=config_file)\n",
    "# circuits = modeller.batch_circuits(\n",
    "#     circuits=circuits,\n",
    "#     write_to_subsystem=True, batch_size=config_file['simulation'].get('batch_size', 100),\n",
    "#     methods={\n",
    "#         \"compute_interactions\": {},\n",
    "#         # \"init_circuits\": {'batch': True},\n",
    "#         # \"simulate_signal_batch\": {'ref_circuit': None,\n",
    "#         #                           'batch': True},\n",
    "#         \"write_results\": {'no_visualisations': config_file['experiment'].get('no_visualisations', True),\n",
    "#                           'no_numerical': config_file['experiment'].get('no_numerical', False)}\n",
    "#     })"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Save"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>circuit_name</th>\n",
       "      <th>mutation_name</th>\n",
       "      <th>mutation_num</th>\n",
       "      <th>mutation_type</th>\n",
       "      <th>mutation_positions</th>\n",
       "      <th>path_to_template_circuit</th>\n",
       "      <th>index</th>\n",
       "      <th>name</th>\n",
       "      <th>interacting</th>\n",
       "      <th>self_interacting</th>\n",
       "      <th>...</th>\n",
       "      <th>energies_1-0</th>\n",
       "      <th>energies_1-1</th>\n",
       "      <th>eqconstants_0-0</th>\n",
       "      <th>eqconstants_0-1</th>\n",
       "      <th>eqconstants_1-0</th>\n",
       "      <th>eqconstants_1-1</th>\n",
       "      <th>num_interacting_diff_to_base_circuit</th>\n",
       "      <th>num_self_interacting_diff_to_base_circuit</th>\n",
       "      <th>num_interacting_ratio_from_mutation_to_base</th>\n",
       "      <th>num_self_interacting_ratio_from_mutation_to_base</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>merged</td>\n",
       "      <td>ref_circuit</td>\n",
       "      <td>0</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>data/merged.fasta</td>\n",
       "      <td>0.0</td>\n",
       "      <td>merged</td>\n",
       "      <td>[]</td>\n",
       "      <td>[[0 0]]</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.170795</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>merged</td>\n",
       "      <td>6S_RNA_m1-0</td>\n",
       "      <td>1</td>\n",
       "      <td>[2]</td>\n",
       "      <td>[0]</td>\n",
       "      <td>data/merged.fasta</td>\n",
       "      <td>0.0</td>\n",
       "      <td>merged</td>\n",
       "      <td>[]</td>\n",
       "      <td>[[0 0]]</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.170795</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>merged</td>\n",
       "      <td>6S_RNA_m1-1</td>\n",
       "      <td>1</td>\n",
       "      <td>[1]</td>\n",
       "      <td>[0]</td>\n",
       "      <td>data/merged.fasta</td>\n",
       "      <td>1.0</td>\n",
       "      <td>merged</td>\n",
       "      <td>[]</td>\n",
       "      <td>[[0 0]]</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.124887</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>merged</td>\n",
       "      <td>6S_RNA_m1-10</td>\n",
       "      <td>1</td>\n",
       "      <td>[9]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>data/merged.fasta</td>\n",
       "      <td>10.0</td>\n",
       "      <td>merged</td>\n",
       "      <td>[]</td>\n",
       "      <td>[[0 0]]</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.322931</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>merged</td>\n",
       "      <td>6S_RNA_m1-100</td>\n",
       "      <td>1</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[33]</td>\n",
       "      <td>data/merged.fasta</td>\n",
       "      <td>100.0</td>\n",
       "      <td>merged</td>\n",
       "      <td>[]</td>\n",
       "      <td>[[0 0]]</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.174415</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>737</th>\n",
       "      <td>merged</td>\n",
       "      <td>antisense_regulatory_RNA_RdlD_m1-95</td>\n",
       "      <td>1</td>\n",
       "      <td>[8]</td>\n",
       "      <td>[31]</td>\n",
       "      <td>data/merged.fasta</td>\n",
       "      <td>644.0</td>\n",
       "      <td>merged</td>\n",
       "      <td>[[0 1]]</td>\n",
       "      <td>[[0 0]]</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.16</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.170795</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>738</th>\n",
       "      <td>merged</td>\n",
       "      <td>antisense_regulatory_RNA_RdlD_m1-96</td>\n",
       "      <td>1</td>\n",
       "      <td>[11]</td>\n",
       "      <td>[32]</td>\n",
       "      <td>data/merged.fasta</td>\n",
       "      <td>645.0</td>\n",
       "      <td>merged</td>\n",
       "      <td>[]</td>\n",
       "      <td>[[0 0]]</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.170795</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>739</th>\n",
       "      <td>merged</td>\n",
       "      <td>antisense_regulatory_RNA_RdlD_m1-97</td>\n",
       "      <td>1</td>\n",
       "      <td>[9]</td>\n",
       "      <td>[32]</td>\n",
       "      <td>data/merged.fasta</td>\n",
       "      <td>646.0</td>\n",
       "      <td>merged</td>\n",
       "      <td>[]</td>\n",
       "      <td>[[0 0]]</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.170795</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>740</th>\n",
       "      <td>merged</td>\n",
       "      <td>antisense_regulatory_RNA_RdlD_m1-98</td>\n",
       "      <td>1</td>\n",
       "      <td>[10]</td>\n",
       "      <td>[32]</td>\n",
       "      <td>data/merged.fasta</td>\n",
       "      <td>647.0</td>\n",
       "      <td>merged</td>\n",
       "      <td>[]</td>\n",
       "      <td>[[0 0]]</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.170795</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>741</th>\n",
       "      <td>merged</td>\n",
       "      <td>antisense_regulatory_RNA_RdlD_m1-99</td>\n",
       "      <td>1</td>\n",
       "      <td>[3]</td>\n",
       "      <td>[33]</td>\n",
       "      <td>data/merged.fasta</td>\n",
       "      <td>648.0</td>\n",
       "      <td>merged</td>\n",
       "      <td>[]</td>\n",
       "      <td>[[0 0]]</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.170795</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>742 rows × 32 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    circuit_name                        mutation_name  mutation_num  \\\n",
       "0         merged                          ref_circuit             0   \n",
       "1         merged                          6S_RNA_m1-0             1   \n",
       "2         merged                          6S_RNA_m1-1             1   \n",
       "3         merged                         6S_RNA_m1-10             1   \n",
       "4         merged                        6S_RNA_m1-100             1   \n",
       "..           ...                                  ...           ...   \n",
       "737       merged  antisense_regulatory_RNA_RdlD_m1-95             1   \n",
       "738       merged  antisense_regulatory_RNA_RdlD_m1-96             1   \n",
       "739       merged  antisense_regulatory_RNA_RdlD_m1-97             1   \n",
       "740       merged  antisense_regulatory_RNA_RdlD_m1-98             1   \n",
       "741       merged  antisense_regulatory_RNA_RdlD_m1-99             1   \n",
       "\n",
       "    mutation_type mutation_positions path_to_template_circuit  index    name  \\\n",
       "0              []                 []        data/merged.fasta    0.0  merged   \n",
       "1             [2]                [0]        data/merged.fasta    0.0  merged   \n",
       "2             [1]                [0]        data/merged.fasta    1.0  merged   \n",
       "3             [9]                [3]        data/merged.fasta   10.0  merged   \n",
       "4             [3]               [33]        data/merged.fasta  100.0  merged   \n",
       "..            ...                ...                      ...    ...     ...   \n",
       "737           [8]               [31]        data/merged.fasta  644.0  merged   \n",
       "738          [11]               [32]        data/merged.fasta  645.0  merged   \n",
       "739           [9]               [32]        data/merged.fasta  646.0  merged   \n",
       "740          [10]               [32]        data/merged.fasta  647.0  merged   \n",
       "741           [3]               [33]        data/merged.fasta  648.0  merged   \n",
       "\n",
       "    interacting self_interacting  ...  energies_1-0  energies_1-1  \\\n",
       "0            []          [[0 0]]  ...          0.00           0.0   \n",
       "1            []          [[0 0]]  ...          0.00           0.0   \n",
       "2            []          [[0 0]]  ...          0.00           0.0   \n",
       "3            []          [[0 0]]  ...          0.00           0.0   \n",
       "4            []          [[0 0]]  ...          0.00           0.0   \n",
       "..          ...              ...  ...           ...           ...   \n",
       "737     [[0 1]]          [[0 0]]  ...         -0.16           0.0   \n",
       "738          []          [[0 0]]  ...          0.00           0.0   \n",
       "739          []          [[0 0]]  ...          0.00           0.0   \n",
       "740          []          [[0 0]]  ...          0.00           0.0   \n",
       "741          []          [[0 0]]  ...          0.00           0.0   \n",
       "\n",
       "    eqconstants_0-0 eqconstants_0-1 eqconstants_1-0 eqconstants_1-1  \\\n",
       "0          1.170795        0.000002        0.000002        0.000002   \n",
       "1          1.170795        0.000002        0.000002        0.000002   \n",
       "2          1.124887        0.000002        0.000002        0.000002   \n",
       "3          0.322931        0.000002        0.000002        0.000002   \n",
       "4          0.174415        0.000002        0.000002        0.000002   \n",
       "..              ...             ...             ...             ...   \n",
       "737        1.170795        0.000002        0.000002        0.000002   \n",
       "738        1.170795        0.000002        0.000002        0.000002   \n",
       "739        1.170795        0.000002        0.000002        0.000002   \n",
       "740        1.170795        0.000002        0.000002        0.000002   \n",
       "741        1.170795        0.000002        0.000002        0.000002   \n",
       "\n",
       "     num_interacting_diff_to_base_circuit  \\\n",
       "0                                       0   \n",
       "1                                       0   \n",
       "2                                       0   \n",
       "3                                       0   \n",
       "4                                       0   \n",
       "..                                    ...   \n",
       "737                                    -1   \n",
       "738                                     0   \n",
       "739                                     0   \n",
       "740                                     0   \n",
       "741                                     0   \n",
       "\n",
       "     num_self_interacting_diff_to_base_circuit  \\\n",
       "0                                            0   \n",
       "1                                            0   \n",
       "2                                            0   \n",
       "3                                            0   \n",
       "4                                            0   \n",
       "..                                         ...   \n",
       "737                                          0   \n",
       "738                                          0   \n",
       "739                                          0   \n",
       "740                                          0   \n",
       "741                                          0   \n",
       "\n",
       "     num_interacting_ratio_from_mutation_to_base  \\\n",
       "0                                            0.0   \n",
       "1                                            0.0   \n",
       "2                                            0.0   \n",
       "3                                            0.0   \n",
       "4                                            0.0   \n",
       "..                                           ...   \n",
       "737                                          0.0   \n",
       "738                                          0.0   \n",
       "739                                          0.0   \n",
       "740                                          0.0   \n",
       "741                                          0.0   \n",
       "\n",
       "     num_self_interacting_ratio_from_mutation_to_base  \n",
       "0                                                 1.0  \n",
       "1                                                 1.0  \n",
       "2                                                 1.0  \n",
       "3                                                 1.0  \n",
       "4                                                 1.0  \n",
       "..                                                ...  \n",
       "737                                               1.0  \n",
       "738                                               1.0  \n",
       "739                                               1.0  \n",
       "740                                               1.0  \n",
       "741                                               1.0  \n",
       "\n",
       "[742 rows x 32 columns]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# source_dirs = [data_writer.top_write_dir]\n",
    "# source_dirs = ['data/tests/2023_11_03_104039/']\n",
    "# tables = load_multiple_as_list(inputs_list=source_dirs, load_func=b_tabulate_mutation_info, \n",
    "#             data_writer=data_writer)\n",
    "# saves = pd.concat(tables, axis=0)\n",
    "# saves"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Compare"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number simulated: 3363\n",
      "Total number of mutations: 28821\n"
     ]
    }
   ],
   "source": [
    "ddir = '/home/wadh6511/Kode/EvoScaper/notebooks_sRNA/data/04_distribution_mutational_robustness/2023_11_10_191219/mutations'\n",
    "mutations = pd.read_csv('/home/wadh6511/Kode/EvoScaper/notebooks_sRNA/data/04_distribution_mutational_robustness/2023_11_10_191219/mutations.csv')\n",
    "\n",
    "print(f'Number simulated: {len(os.listdir(ddir))}')\n",
    "print(f'Percent simulated: {len(os.listdir(ddir)) / len(mutations) * 100} %')\n",
    "print(f'Total number of mutations: {len(mutations)}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "env_evo",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
