{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Comparison of binding sequences for documented and predicted sRNA-mRNA interactions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import re\n",
    "\n",
    "from synbio_morpher.utils.misc.type_handling import flatten_listlike\n",
    "from synbio_morpher.utils.data.data_format_tools.common import load_json_as_dict\n",
    "from synbio_morpher.srv.io.manage.script_manager import script_preamble\n",
    "\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import jax\n",
    "import subprocess\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "fn_RNAInter = os.path.join('..', 'data', 'sRNA', 'RNAInter', 'Download_data_RR.csv')\n",
    "fn_merged = os.path.join('..', 'data', 'sRNA', 'merged_EcoCyc_RNAInter.csv')\n",
    "data = pd.read_csv(fn_RNAInter)\n",
    "merged = pd.read_csv(fn_merged)\n",
    "try:\n",
    "    data = data.drop(columns=['Unnamed: 0'])\n",
    "    merged = merged.drop(columns=['Unnamed: 0'])\n",
    "except:\n",
    "    pass\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'ACACCGTCGCTTAAAGTGACGGCATAATAATAAAAAAATGAAATTCCTCTTTGACGGGCCAATAGCGATATTGGCCATTTTTTT'"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged[merged['Name'] == 'chiX']['Sequence'].iloc[1]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Predicted interactions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['id1', 'id2', 'E', 'E_norm', 'bpList', 'hybridDPfull']"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fn = './data/05_orthogonality_mRNA_official/2023_11_21_095238/inter_data_raw.json'\n",
    "\n",
    "sim_data_l = load_json_as_dict(fn)\n",
    "\n",
    "k1 = list(sim_data_l.keys())[0]\n",
    "k2 = list(sim_data_l[k1].keys())[0]\n",
    "header = list(sim_data_l[k1][k2].keys())\n",
    "default_vals = {\n",
    "    'id1': '', 'id2': '', 'E': 0.0, 'E_norm': 0.0, 'bpList': '', 'hybridDPfull': ''\n",
    "}\n",
    "for s, v in sim_data_l.items():\n",
    "    if len(v.values()) < len(sim_data_l[k1]):\n",
    "        diffs = set(sim_data_l[k1].keys()) - set(v.keys())\n",
    "        for d in diffs:\n",
    "            sim_data_l[s][d] = default_vals\n",
    "        \n",
    "header"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Get interaction subsequences "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "      <th>ID</th>\n",
       "      <th>Category</th>\n",
       "      <th>Symbol</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Database</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>acnA</td>\n",
       "      <td>NCBI:946724</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>acnA</td>\n",
       "      <td>ATTCGGAACGAGGCCTGAAGCAGTGTCGCCGTCCCTCTGCCTTGCA...</td>\n",
       "      <td>RNAInter</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>acrZ</td>\n",
       "      <td>NCBI:945365</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>acrZ</td>\n",
       "      <td>ACTTACTACTGTCTTCGGGGGGTCCGAGGTTTCTGGGGGGTCGTAC...</td>\n",
       "      <td>RNAInter</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>arcZ</td>\n",
       "      <td>NCBI:2847690</td>\n",
       "      <td>sRNA</td>\n",
       "      <td>arcZ</td>\n",
       "      <td>CTCATGTTGACCGCTTGTTTAGCAGCTTCAAGGAAGCTGAAGGGCA...</td>\n",
       "      <td>RNAInter</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>argR</td>\n",
       "      <td>NCBI:947861</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>argR</td>\n",
       "      <td>GACAATGGCGATAGTATGGCGGTTGTTTCTTTCCCATCTCTACTCA...</td>\n",
       "      <td>RNAInter</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>arrS</td>\n",
       "      <td>NCBI:11115378</td>\n",
       "      <td>sRNA</td>\n",
       "      <td>arrS</td>\n",
       "      <td>CTCTCCCTCTCTTTCTCTCTTCTCCGCGGCGATACGATCCGAGATG...</td>\n",
       "      <td>RNAInter</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>596</th>\n",
       "      <td>msrA</td>\n",
       "      <td>chromosome:NC_000913.3, Gene ID:948734</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>msrA</td>\n",
       "      <td>ATGAGTTTATTTGATAAAAAGCATCTGGTTTCCCCCGCCGATGCCC...</td>\n",
       "      <td>sRNATarBase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>597</th>\n",
       "      <td>hdeA</td>\n",
       "      <td>chromosome:NC_000913.3, Gene ID:948025</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>hdeA</td>\n",
       "      <td>ATGAAAAAAGTATTAGGCGTTATTCTTGGTGGTCTGCTTCTTCTGC...</td>\n",
       "      <td>sRNATarBase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>598</th>\n",
       "      <td>hdeB</td>\n",
       "      <td>chromosome:NC_000913.3, Gene ID:948026</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>hdeB</td>\n",
       "      <td>ATGAATATTTCATCTCTCCGTAAAGCGTTTATTTTTATGGGCGCTG...</td>\n",
       "      <td>sRNATarBase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>599</th>\n",
       "      <td>fliC</td>\n",
       "      <td>chromosome:NC_002695.1, Gene ID:</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>fliC</td>\n",
       "      <td>ATGGCACAAGTCATTAATACCAACAGCCTCTCGCTGATCACTCAAA...</td>\n",
       "      <td>sRNATarBase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>600</th>\n",
       "      <td>mreB</td>\n",
       "      <td>chromosome:NC_000913.3, Gene ID:948588</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>mreB</td>\n",
       "      <td>ATGTTGAAAAAATTTCGTGGCATGTTTTCCAATGACTTGTCCATTG...</td>\n",
       "      <td>sRNATarBase</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>601 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Name                                      ID Category Symbol  \\\n",
       "0    acnA                             NCBI:946724     mRNA   acnA   \n",
       "1    acrZ                             NCBI:945365     mRNA   acrZ   \n",
       "2    arcZ                            NCBI:2847690     sRNA   arcZ   \n",
       "3    argR                             NCBI:947861     mRNA   argR   \n",
       "4    arrS                           NCBI:11115378     sRNA   arrS   \n",
       "..    ...                                     ...      ...    ...   \n",
       "596  msrA  chromosome:NC_000913.3, Gene ID:948734     mRNA   msrA   \n",
       "597  hdeA  chromosome:NC_000913.3, Gene ID:948025     mRNA   hdeA   \n",
       "598  hdeB  chromosome:NC_000913.3, Gene ID:948026     mRNA   hdeB   \n",
       "599  fliC        chromosome:NC_002695.1, Gene ID:     mRNA   fliC   \n",
       "600  mreB  chromosome:NC_000913.3, Gene ID:948588     mRNA   mreB   \n",
       "\n",
       "                                              Sequence     Database  \n",
       "0    ATTCGGAACGAGGCCTGAAGCAGTGTCGCCGTCCCTCTGCCTTGCA...     RNAInter  \n",
       "1    ACTTACTACTGTCTTCGGGGGGTCCGAGGTTTCTGGGGGGTCGTAC...     RNAInter  \n",
       "2    CTCATGTTGACCGCTTGTTTAGCAGCTTCAAGGAAGCTGAAGGGCA...     RNAInter  \n",
       "3    GACAATGGCGATAGTATGGCGGTTGTTTCTTTCCCATCTCTACTCA...     RNAInter  \n",
       "4    CTCTCCCTCTCTTTCTCTCTTCTCCGCGGCGATACGATCCGAGATG...     RNAInter  \n",
       "..                                                 ...          ...  \n",
       "596  ATGAGTTTATTTGATAAAAAGCATCTGGTTTCCCCCGCCGATGCCC...  sRNATarBase  \n",
       "597  ATGAAAAAAGTATTAGGCGTTATTCTTGGTGGTCTGCTTCTTCTGC...  sRNATarBase  \n",
       "598  ATGAATATTTCATCTCTCCGTAAAGCGTTTATTTTTATGGGCGCTG...  sRNATarBase  \n",
       "599  ATGGCACAAGTCATTAATACCAACAGCCTCTCGCTGATCACTCAAA...  sRNATarBase  \n",
       "600  ATGTTGAAAAAATTTCGTGGCATGTTTTCCAATGACTTGTCCATTG...  sRNATarBase  \n",
       "\n",
       "[601 rows x 6 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def process_bplist(bplist, idx):\n",
    "    return [int(i.replace(')', ',').split(',')[idx]) for i in bplist.split('(')[1:]]\n",
    "\n",
    "\n",
    "def longest_common_substring(x, y):\n",
    "    \"\"\" Bard \"\"\"\n",
    "    import difflib\n",
    "    # Create a sequence matcher object\n",
    "    matcher = difflib.SequenceMatcher(None, x, y)\n",
    "\n",
    "    # Find the longest common substring match\n",
    "    match = matcher.find_longest_match()\n",
    "\n",
    "    # Extract the longest common substring\n",
    "    if match.size != 0:\n",
    "        substring = x[match.a:match.a + match.size]\n",
    "        return substring\n",
    "    else:\n",
    "        return \"\"\n",
    "    \n",
    "\n",
    "def create_groups(x, y):\n",
    "    \"\"\" CGPT \"\"\"\n",
    "    groups = []\n",
    "    group = []\n",
    "    for i, num in enumerate(x):\n",
    "        if i in y:\n",
    "            if group:\n",
    "                groups.append(group)\n",
    "                group = []\n",
    "            group.append(num)\n",
    "        else:\n",
    "            group.append(num)\n",
    "    groups.append(group)\n",
    "    return groups"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "for k1 in sim_data_l:\n",
    "    for k2, v in sim_data_l[k1].items():\n",
    "        if v['id1'] == '' or v['id2'] == '':\n",
    "            sim_data_l[k1][k2]['subsequence_groups'] = []\n",
    "            continue\n",
    "        seq1, cat1 = merged[(merged['Name'] == v['id1']) | (merged['Symbol'] == v['id1'])][['Sequence', 'Category']].iloc[0].to_list()\n",
    "        seq2, cat2 = merged[(merged['Name'] == v['id2']) | (merged['Symbol'] == v['id2'])][['Sequence', 'Category']].iloc[0].to_list()\n",
    "        \n",
    "        if (cat1 == 'mRNA') and (cat2 == 'mRNA'):\n",
    "            print(k1, k2)\n",
    "            break\n",
    "        \n",
    "        subsequence, i = (seq1, 0) if cat1 == 'mRNA' else (seq2, 1)\n",
    "        idxs = np.array(sorted(process_bplist(v['bpList'], i)))\n",
    "        breakup = np.where(idxs[1:] - idxs[:-1] > 1)[0] + 1\n",
    "        subsequence_groups = []\n",
    "        for g in create_groups(idxs, breakup):\n",
    "            subsequence_groups.append(''.join(np.array(list(subsequence))[np.array(g) - 1]))\n",
    "        \n",
    "        sim_data_l[k1][k2]['subsequence_groups'] = subsequence_groups\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['CTCA', 'C', 'CACCT', 'GTGGG', 'CCAGCACACT', 'TCGGAGGCCA', 'GC']"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sim_data_l['arcZ']['acnA']['subsequence_groups']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Run bpRNA on all sequences\n",
    "\n",
    "Takes ca. 10 minutes on 37 sRNA x 232 mRNA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "config = {\n",
    "    \"experiment\": {\n",
    "        \"purpose\": \"tests\",\n",
    "        \"no_visualisations\": False,\n",
    "        \"no_numerical\": False,\n",
    "        \"debug_mode\": False\n",
    "    }\n",
    "}\n",
    "data_writer = None\n",
    "config, data_writer = script_preamble(config, data_writer)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "arrS ilvI\n"
     ]
    }
   ],
   "source": [
    "def write_dbn(outname, outdir, id_name, seq, db):\n",
    "    fn = os.path.join(outdir, outname + '.dbn')\n",
    "    with open(fn, 'w') as f:\n",
    "        f.writelines('>' + id_name + '\\n' +\n",
    "                     seq + '\\n' +\n",
    "                     db + '\\n')\n",
    "    return fn\n",
    "\n",
    "\n",
    "# def make_db(bplist, seq_len):\n",
    "#     db = np.array(list('.' * seq_len))\n",
    "#     db[process_bplist(bplist, idx=0)] = '('\n",
    "#     db[process_bplist(bplist, idx=1)] = ')'\n",
    "#     return ''.join(db)\n",
    "\n",
    "\n",
    "def execute_perl_script(*args):\n",
    "    \"\"\" Bard \"\"\"\n",
    "    script_path = './bpRNA.pl'\n",
    "    try:\n",
    "        subprocess.run([\"perl\", script_path, *args])\n",
    "    except Exception as e:\n",
    "        print(f\"Error executing Perl script: {e}\")\n",
    "\n",
    "\n",
    "for k1 in sim_data_l:\n",
    "    data_writer.subdivide_writing('st')\n",
    "    data_writer.subdivide_writing(k1, safe_dir_change=False)\n",
    "    data_writer.unsubdivide()\n",
    "    data_writer.subdivide_writing('dbn')\n",
    "    data_writer.subdivide_writing(k1, safe_dir_change=False)\n",
    "    \n",
    "    for k2 in sim_data_l[k1]:\n",
    "        bplist = sim_data_l[k1][k2]['bpList']\n",
    "        # make_db(bplist, seq_len=len(db))\n",
    "        db = sim_data_l[k1][k2]['hybridDPfull'].replace('&', '')\n",
    "        seq = merged[merged['Symbol'] == k1]['Sequence'].iloc[0] + \\\n",
    "            merged[merged['Symbol'] == k2]['Sequence'].iloc[0]\n",
    "        fn = write_dbn(k1 + '_' + k2, data_writer.write_dir, id_name='arcZ', seq=seq, db=db)\n",
    "        try:\n",
    "            execute_perl_script(fn, fn.replace('.dbn', '').replace('dbn', 'st'))\n",
    "        except:\n",
    "            print(k1, k2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Examine bpRNA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [],
   "source": [
    "BPRNA_STRUCTURE_TYPES = [\"S\", \"H\", \"B\", \"I\", \"M\", \"X\", \"E\", \"PK\", \"PKBP\", \"NCBP\", \"segment\"]\n",
    "\n",
    "def process_st(fn_st):\n",
    "    \n",
    "    # Match all digits in the string and replace them with an empty string\n",
    "    pattern1 = r'[0-9]'\n",
    "    pattern2 = \"[^0-9]\"\n",
    "    \n",
    "    with open(fn_st, 'r') as f:\n",
    "        reads = [l.split(' ')[:2] for l in f.readlines()[7:]]\n",
    "        motifs = [l[0] for l in reads]\n",
    "        motifs = list(map(lambda x: [re.sub(pattern1, '', x.replace('.', '')), [int(re.sub(pattern2, '', xx)) for xx in x.split('.')]], motifs))\n",
    "        \n",
    "        motif_lengths = [l[1].split('..') for l in reads]\n",
    "        motif_lengths = list(map(lambda x: int(x[1]) - int(x[0]) if len(x) > 1 else 1, motif_lengths))\n",
    "        \n",
    "    d = {}\n",
    "    for (a, b), ml in zip(motifs, motif_lengths):\n",
    "        d.setdefault(a, {}).setdefault('motif_groups', []).append(b)\n",
    "        d.setdefault(a, {}).setdefault('motif_length', []).append(ml)\n",
    "        \n",
    "    return d\n",
    "    \n",
    "    \n",
    "def aggregate_motifs(sim_data):\n",
    "    structures = pd.DataFrame()\n",
    "    dir_st = './data/08_comparison_sequences/2023_11_22_225107/st'\n",
    "    for id1 in sim_data:\n",
    "        for id2 in sim_data[id1]:\n",
    "            fn_st = os.path.join(dir_st, id1, id1 + '_' + id2 + '.st')\n",
    "            if not os.path.isfile(fn_st):\n",
    "                continue\n",
    "            structures_d = {}\n",
    "            d = process_st(fn_st)\n",
    "            structures_d[('sRNA', '')] = id1\n",
    "            structures_d[('Target', '')] = id2\n",
    "            for s in BPRNA_STRUCTURE_TYPES:\n",
    "                if d.get(s):\n",
    "                    structures_d[('Mean Length', s)] = np.mean(d[s]['motif_length'])\n",
    "                    structures_d[('Num in seq', s)] = len(d[s]['motif_groups'])\n",
    "                else:\n",
    "                    structures_d[('Mean Length', s)] = [0]\n",
    "                    structures_d[('Num in seq', s)] = [0]\n",
    "            structures = pd.concat([structures, pd.DataFrame.from_dict(structures_d)])\n",
    "            \n",
    "    structures = structures.reset_index().drop(columns='index')\n",
    "    return structures[structures.columns[:2].to_list() + sorted(structures.columns[2:])]\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [],
   "source": [
    "fn_st = './data/08_comparison_sequences/2023_11_22_225107/st/arcZ/arcZ_acnA.st'\n",
    "\n",
    "d = process_st(fn_st)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_493764/3331885936.py:46: PerformanceWarning: dropping on a non-lexsorted multi-index without a level parameter may impact performance.\n",
      "  structures = structures.reset_index().drop(columns='index')\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th>sRNA</th>\n",
       "      <th>Target</th>\n",
       "      <th colspan=\"8\" halign=\"left\">Mean Length</th>\n",
       "      <th>...</th>\n",
       "      <th colspan=\"10\" halign=\"left\">Num in seq</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>B</th>\n",
       "      <th>E</th>\n",
       "      <th>H</th>\n",
       "      <th>I</th>\n",
       "      <th>M</th>\n",
       "      <th>NCBP</th>\n",
       "      <th>PK</th>\n",
       "      <th>PKBP</th>\n",
       "      <th>...</th>\n",
       "      <th>E</th>\n",
       "      <th>H</th>\n",
       "      <th>I</th>\n",
       "      <th>M</th>\n",
       "      <th>NCBP</th>\n",
       "      <th>PK</th>\n",
       "      <th>PKBP</th>\n",
       "      <th>S</th>\n",
       "      <th>X</th>\n",
       "      <th>segment</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>arcZ</td>\n",
       "      <td>chiP</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>10.0</td>\n",
       "      <td>371.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>arcZ</td>\n",
       "      <td>acrZ</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>16.0</td>\n",
       "      <td>336.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>22</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>arcZ</td>\n",
       "      <td>asr</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>102.0</td>\n",
       "      <td>311.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>17</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>arcZ</td>\n",
       "      <td>cheZ</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>194.5</td>\n",
       "      <td>116.0</td>\n",
       "      <td>3.333333</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>19</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>arcZ</td>\n",
       "      <td>acnA</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>35.5</td>\n",
       "      <td>481.0</td>\n",
       "      <td>1.600000</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "      <td>34</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>459</th>\n",
       "      <td>arrS</td>\n",
       "      <td>yhjJ</td>\n",
       "      <td>0.750000</td>\n",
       "      <td>800.0</td>\n",
       "      <td>67.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>26</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>460</th>\n",
       "      <td>arrS</td>\n",
       "      <td>pmbA</td>\n",
       "      <td>1.500000</td>\n",
       "      <td>362.5</td>\n",
       "      <td>632.0</td>\n",
       "      <td>1.136364</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>22</td>\n",
       "      <td>0</td>\n",
       "      <td>72</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>20</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>461</th>\n",
       "      <td>arrS</td>\n",
       "      <td>yphD</td>\n",
       "      <td>2.250000</td>\n",
       "      <td>105.0</td>\n",
       "      <td>938.0</td>\n",
       "      <td>1.166667</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>21</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>462</th>\n",
       "      <td>arrS</td>\n",
       "      <td>tolC</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>52.5</td>\n",
       "      <td>1528.0</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>12</td>\n",
       "      <td>0</td>\n",
       "      <td>27</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>463</th>\n",
       "      <td>arrS</td>\n",
       "      <td>uhpT</td>\n",
       "      <td>3.500000</td>\n",
       "      <td>268.0</td>\n",
       "      <td>1048.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>14</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>464 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     sRNA Target Mean Length                                           ...  \\\n",
       "                           B      E       H         I  M NCBP PK PKBP  ...   \n",
       "0    arcZ   chiP    1.000000   10.0   371.0  0.000000  0  1.0  0    0  ...   \n",
       "1    arcZ   acrZ    2.000000   16.0   336.0  0.000000  0  1.0  0    0  ...   \n",
       "2    arcZ    asr    1.000000  102.0   311.0  0.000000  0  1.0  0    0  ...   \n",
       "3    arcZ   cheZ    0.000000  194.5   116.0  3.333333  0  1.0  0    0  ...   \n",
       "4    arcZ   acnA    0.333333   35.5   481.0  1.600000  0  1.0  0    0  ...   \n",
       "..    ...    ...         ...    ...     ...       ... ..  ... ..  ...  ...   \n",
       "459  arrS   yhjJ    0.750000  800.0    67.0  1.000000  0  1.0  0    0  ...   \n",
       "460  arrS   pmbA    1.500000  362.5   632.0  1.136364  0  1.0  0    0  ...   \n",
       "461  arrS   yphD    2.250000  105.0   938.0  1.166667  0  1.0  0    0  ...   \n",
       "462  arrS   tolC    0.000000   52.5  1528.0  0.666667  0  1.0  0    0  ...   \n",
       "463  arrS   uhpT    3.500000  268.0  1048.0  0.000000  0  1.0  0    0  ...   \n",
       "\n",
       "    Num in seq                                        \n",
       "             E  H   I  M NCBP PK PKBP   S  X segment  \n",
       "0            1  1   0  0    9  0    0   2  0       1  \n",
       "1            2  1   2  0   22  0    0   4  0       1  \n",
       "2            2  1   0  0   17  0    0   3  0       1  \n",
       "3            2  1   6  0   19  0    0   4  0       1  \n",
       "4            2  1  10  0   34  0    0   9  0       1  \n",
       "..         ... ..  .. ..  ... ..  ...  .. ..     ...  \n",
       "459          2  1   2  0   26  0    0   6  0       1  \n",
       "460          2  1  22  0   72  0    0  20  0       1  \n",
       "461          2  1   6  0   21  0    0   8  0       1  \n",
       "462          2  1  12  0   27  0    0   7  0       1  \n",
       "463          2  1   0  0   14  0    0   3  0       1  \n",
       "\n",
       "[464 rows x 24 columns]"
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "motifs = aggregate_motifs({'arcZ': sim_data_l['arcZ'], 'arrS': sim_data_l['arrS']})\n",
    "motifs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.barplot(motifs, x=)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "env_evo",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
