{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import re\n",
    "\n",
    "import numpy as np\n",
    "import jax\n",
    "jax.config.update('jax_platform_name', 'gpu')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "fn_EcoCyc = '../data/sRNA/EcoCyc/EcoCyc_regulatory_RNAs.csv'\n",
    "fn_RNAInter = os.path.join('..', 'data', 'sRNA', 'RNAInter', 'Download_data_RR.csv')\n",
    "fn_srnatarbase = os.path.join('..', 'data', 'sRNA', 'sRNATarBase', 'sRNATarBase.csv')\n",
    "ecocyc = pd.read_csv(fn_EcoCyc)\n",
    "rnainter = pd.read_csv(fn_RNAInter)\n",
    "srnatarbase = pd.read_csv(fn_srnatarbase)\n",
    "# data = pd.concat([pd.read_csv(fn_RNAInter), pd.read_csv(fn_EcoCyc)])\n",
    "try:\n",
    "    rnainter = rnainter.drop(columns=['Unnamed: 0'])\n",
    "    srnatarbase = srnatarbase.drop(columns=['Unnamed: 0'])\n",
    "except:\n",
    "    pass\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Start with RNAInter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.sum(rnainter['Sequence1'].isna()) + np.sum(rnainter['Sequence2'].isna())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>level_0</th>\n",
       "      <th>index</th>\n",
       "      <th>RNAInterID</th>\n",
       "      <th>Interactor1.Symbol</th>\n",
       "      <th>Category1</th>\n",
       "      <th>Species1</th>\n",
       "      <th>Interactor2.Symbol</th>\n",
       "      <th>Category2</th>\n",
       "      <th>Species2</th>\n",
       "      <th>Raw_ID1</th>\n",
       "      <th>Raw_ID2</th>\n",
       "      <th>score</th>\n",
       "      <th>strong</th>\n",
       "      <th>weak</th>\n",
       "      <th>predict</th>\n",
       "      <th>Sequence1</th>\n",
       "      <th>Sequence2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>1029724</td>\n",
       "      <td>RR05384747</td>\n",
       "      <td>acnA</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>Escherichia coli str. K-12 substr. MG1655</td>\n",
       "      <td>ryhB</td>\n",
       "      <td>sRNA</td>\n",
       "      <td>Escherichia coli str. K-12 substr. MG1655</td>\n",
       "      <td>NCBI:946724</td>\n",
       "      <td>NCBI:2847761</td>\n",
       "      <td>0.2292</td>\n",
       "      <td>Northern blot//Reporter assay</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ATTCGGAACGAGGCCTGAAGCAGTGTCGCCGTCCCTCTGCCTTGCA...</td>\n",
       "      <td>TTTGAGAGCGTCTCTGTCCCTCGTTTTGCGGTTAAGCCGCATCCAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>1029725</td>\n",
       "      <td>RR05384845</td>\n",
       "      <td>acrZ</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>Escherichia coli str. K-12 substr. MG1655</td>\n",
       "      <td>omrB</td>\n",
       "      <td>sRNA</td>\n",
       "      <td>Escherichia coli str. K-12 substr. MG1655</td>\n",
       "      <td>NCBI:945365</td>\n",
       "      <td>NCBI:2847747</td>\n",
       "      <td>0.1778</td>\n",
       "      <td>Northern blot</td>\n",
       "      <td>Microarray</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ACTTACTACTGTCTTCGGGGGGTCCGAGGTTTCTGGGGGGTCGTAC...</td>\n",
       "      <td>TGTTCTATACTTGGGTTCGACTTGGGTTAGACTTGTCTTTACTGTC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>1029732</td>\n",
       "      <td>RR05387056</td>\n",
       "      <td>arcZ</td>\n",
       "      <td>sRNA</td>\n",
       "      <td>Escherichia coli str. K-12 substr. MG1655</td>\n",
       "      <td>flhD</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>Escherichia coli str. K-12 substr. MG1655</td>\n",
       "      <td>NCBI:2847690</td>\n",
       "      <td>NCBI:945442</td>\n",
       "      <td>0.2314</td>\n",
       "      <td>Northern blot//Reporter assay</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CTCATGTTGACCGCTTGTTTAGCAGCTTCAAGGAAGCTGAAGGGCA...</td>\n",
       "      <td>CTCGTGCCGAATTCGGCACGAGCGATATTTCATCAGTTATCGGTAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>1029733</td>\n",
       "      <td>RR05387057</td>\n",
       "      <td>arcZ</td>\n",
       "      <td>sRNA</td>\n",
       "      <td>Escherichia coli str. K-12 substr. MG1655</td>\n",
       "      <td>rpoS</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>Escherichia coli str. K-12 substr. MG1655</td>\n",
       "      <td>NCBI:2847690</td>\n",
       "      <td>NCBI:947210</td>\n",
       "      <td>0.1996</td>\n",
       "      <td>Northern blot//RACE//RT-PCR//Beta-galactosidas...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CTCATGTTGACCGCTTGTTTAGCAGCTTCAAGGAAGCTGAAGGGCA...</td>\n",
       "      <td>ATCCTCGGGTCTTGCAGGCCACACAGGACACCCTGAACCGTCATGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>1029734</td>\n",
       "      <td>RR05387126</td>\n",
       "      <td>argR</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>Escherichia coli str. K-12 substr. MG1655</td>\n",
       "      <td>dsrA</td>\n",
       "      <td>sRNA</td>\n",
       "      <td>Escherichia coli str. K-12 substr. MG1655</td>\n",
       "      <td>NCBI:947861</td>\n",
       "      <td>NCBI:946470</td>\n",
       "      <td>0.2139</td>\n",
       "      <td>Primer extension assay</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GACAATGGCGATAGTATGGCGGTTGTTTCTTTCCCATCTCTACTCA...</td>\n",
       "      <td>TGAGCAACTTTATTCACATAATTTCTACACCAAGAACTCGAGGTTA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>349</th>\n",
       "      <td>349</td>\n",
       "      <td>8031771</td>\n",
       "      <td>RR05460721</td>\n",
       "      <td>ryhB</td>\n",
       "      <td>sRNA</td>\n",
       "      <td>Escherichia coli str. K-12 substr. MG1655</td>\n",
       "      <td>ynfF</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>Escherichia coli str. K-12 substr. MG1655</td>\n",
       "      <td>NCBI:2847761</td>\n",
       "      <td>NCBI:945268</td>\n",
       "      <td>0.1778</td>\n",
       "      <td>Northern blot</td>\n",
       "      <td>Microarray</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TTTGAGAGCGTCTCTGTCCCTCGTTTTGCGGTTAAGCCGCATCCAT...</td>\n",
       "      <td>ACTTACTACTGTGCCTTTAGGGTGGGGGGATACACCGTTTCCGCCG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>350</th>\n",
       "      <td>350</td>\n",
       "      <td>8031772</td>\n",
       "      <td>RR05460722</td>\n",
       "      <td>sdsR</td>\n",
       "      <td>sRNA</td>\n",
       "      <td>Escherichia coli str. K-12 substr. MG1655</td>\n",
       "      <td>tolC</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>Escherichia coli str. K-12 substr. MG1655</td>\n",
       "      <td>NCBI:2847772</td>\n",
       "      <td>NCBI:947521</td>\n",
       "      <td>0.3292</td>\n",
       "      <td>Northern blot//Western blot</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CTACCTCAACACCCAACAATCGTTGATTGAAAACATCTTCTATCGC...</td>\n",
       "      <td>ATGAAGAAATTGCTCCCCATTCTTATCGGCCTGAGCCTTTCTGGGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>351</th>\n",
       "      <td>351</td>\n",
       "      <td>8031773</td>\n",
       "      <td>RR05460723</td>\n",
       "      <td>sgrS</td>\n",
       "      <td>sRNA</td>\n",
       "      <td>Escherichia coli str. K-12 substr. MG1655</td>\n",
       "      <td>sodB</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>Escherichia coli str. K-12 substr. MG1655</td>\n",
       "      <td>NCBI:4056038</td>\n",
       "      <td>NCBI:944953</td>\n",
       "      <td>0.1983</td>\n",
       "      <td>Northern blot</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CCAGAACATTGGCTTTCAAAAGTTTTTTAATTGATGGAAACTCTTT...</td>\n",
       "      <td>GAAGAGAAAAGAAAAAAAGACCTTATTTGCTTATCCTGCATCTCTC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>352</th>\n",
       "      <td>352</td>\n",
       "      <td>8031774</td>\n",
       "      <td>RR05460724</td>\n",
       "      <td>sgrS</td>\n",
       "      <td>sRNA</td>\n",
       "      <td>Escherichia coli str. K-12 substr. MG1655</td>\n",
       "      <td>yigL</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>Escherichia coli str. K-12 substr. MG1655</td>\n",
       "      <td>NCBI:4056038</td>\n",
       "      <td>NCBI:2847768</td>\n",
       "      <td>0.3273</td>\n",
       "      <td>Northern blot//Reporter assay</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CCAGAACATTGGCTTTCAAAAGTTTTTTAATTGATGGAAACTCTTT...</td>\n",
       "      <td>TGCCGCTGACGAGGGAGAGGATGAGGATTGCGAGCCAGAGCATCCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>353</th>\n",
       "      <td>353</td>\n",
       "      <td>8051381</td>\n",
       "      <td>RR05460725</td>\n",
       "      <td>symE</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>Escherichia coli str. K-12 substr. MG1655</td>\n",
       "      <td>symR</td>\n",
       "      <td>sRNA</td>\n",
       "      <td>Escherichia coli str. K-12 substr. MG1655</td>\n",
       "      <td>NCBI:949088</td>\n",
       "      <td>NCBI:5061533</td>\n",
       "      <td>0.3330</td>\n",
       "      <td>RT-PCR//Reporter assay//Western blot</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ATGACTGACACGCATTCTATTGCACAACCGTTCGAAGCAGAAGTCT...</td>\n",
       "      <td>AGAAAAGATCTGATCGTTGAATTGATGACNCTCNANANCACCANNN...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>354 rows × 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     level_0    index  RNAInterID Interactor1.Symbol Category1  \\\n",
       "0          0  1029724  RR05384747               acnA      mRNA   \n",
       "1          1  1029725  RR05384845               acrZ      mRNA   \n",
       "2          2  1029732  RR05387056               arcZ      sRNA   \n",
       "3          3  1029733  RR05387057               arcZ      sRNA   \n",
       "4          4  1029734  RR05387126               argR      mRNA   \n",
       "..       ...      ...         ...                ...       ...   \n",
       "349      349  8031771  RR05460721               ryhB      sRNA   \n",
       "350      350  8031772  RR05460722               sdsR      sRNA   \n",
       "351      351  8031773  RR05460723               sgrS      sRNA   \n",
       "352      352  8031774  RR05460724               sgrS      sRNA   \n",
       "353      353  8051381  RR05460725               symE      mRNA   \n",
       "\n",
       "                                      Species1 Interactor2.Symbol Category2  \\\n",
       "0    Escherichia coli str. K-12 substr. MG1655               ryhB      sRNA   \n",
       "1    Escherichia coli str. K-12 substr. MG1655               omrB      sRNA   \n",
       "2    Escherichia coli str. K-12 substr. MG1655               flhD      mRNA   \n",
       "3    Escherichia coli str. K-12 substr. MG1655               rpoS      mRNA   \n",
       "4    Escherichia coli str. K-12 substr. MG1655               dsrA      sRNA   \n",
       "..                                         ...                ...       ...   \n",
       "349  Escherichia coli str. K-12 substr. MG1655               ynfF      mRNA   \n",
       "350  Escherichia coli str. K-12 substr. MG1655               tolC      mRNA   \n",
       "351  Escherichia coli str. K-12 substr. MG1655               sodB      mRNA   \n",
       "352  Escherichia coli str. K-12 substr. MG1655               yigL      mRNA   \n",
       "353  Escherichia coli str. K-12 substr. MG1655               symR      sRNA   \n",
       "\n",
       "                                      Species2       Raw_ID1       Raw_ID2  \\\n",
       "0    Escherichia coli str. K-12 substr. MG1655   NCBI:946724  NCBI:2847761   \n",
       "1    Escherichia coli str. K-12 substr. MG1655   NCBI:945365  NCBI:2847747   \n",
       "2    Escherichia coli str. K-12 substr. MG1655  NCBI:2847690   NCBI:945442   \n",
       "3    Escherichia coli str. K-12 substr. MG1655  NCBI:2847690   NCBI:947210   \n",
       "4    Escherichia coli str. K-12 substr. MG1655   NCBI:947861   NCBI:946470   \n",
       "..                                         ...           ...           ...   \n",
       "349  Escherichia coli str. K-12 substr. MG1655  NCBI:2847761   NCBI:945268   \n",
       "350  Escherichia coli str. K-12 substr. MG1655  NCBI:2847772   NCBI:947521   \n",
       "351  Escherichia coli str. K-12 substr. MG1655  NCBI:4056038   NCBI:944953   \n",
       "352  Escherichia coli str. K-12 substr. MG1655  NCBI:4056038  NCBI:2847768   \n",
       "353  Escherichia coli str. K-12 substr. MG1655   NCBI:949088  NCBI:5061533   \n",
       "\n",
       "      score                                             strong        weak  \\\n",
       "0    0.2292                      Northern blot//Reporter assay         NaN   \n",
       "1    0.1778                                      Northern blot  Microarray   \n",
       "2    0.2314                      Northern blot//Reporter assay         NaN   \n",
       "3    0.1996  Northern blot//RACE//RT-PCR//Beta-galactosidas...         NaN   \n",
       "4    0.2139                             Primer extension assay         NaN   \n",
       "..      ...                                                ...         ...   \n",
       "349  0.1778                                      Northern blot  Microarray   \n",
       "350  0.3292                        Northern blot//Western blot         NaN   \n",
       "351  0.1983                                      Northern blot         NaN   \n",
       "352  0.3273                      Northern blot//Reporter assay         NaN   \n",
       "353  0.3330               RT-PCR//Reporter assay//Western blot         NaN   \n",
       "\n",
       "     predict                                          Sequence1  \\\n",
       "0        NaN  ATTCGGAACGAGGCCTGAAGCAGTGTCGCCGTCCCTCTGCCTTGCA...   \n",
       "1        NaN  ACTTACTACTGTCTTCGGGGGGTCCGAGGTTTCTGGGGGGTCGTAC...   \n",
       "2        NaN  CTCATGTTGACCGCTTGTTTAGCAGCTTCAAGGAAGCTGAAGGGCA...   \n",
       "3        NaN  CTCATGTTGACCGCTTGTTTAGCAGCTTCAAGGAAGCTGAAGGGCA...   \n",
       "4        NaN  GACAATGGCGATAGTATGGCGGTTGTTTCTTTCCCATCTCTACTCA...   \n",
       "..       ...                                                ...   \n",
       "349      NaN  TTTGAGAGCGTCTCTGTCCCTCGTTTTGCGGTTAAGCCGCATCCAT...   \n",
       "350      NaN  CTACCTCAACACCCAACAATCGTTGATTGAAAACATCTTCTATCGC...   \n",
       "351      NaN  CCAGAACATTGGCTTTCAAAAGTTTTTTAATTGATGGAAACTCTTT...   \n",
       "352      NaN  CCAGAACATTGGCTTTCAAAAGTTTTTTAATTGATGGAAACTCTTT...   \n",
       "353      NaN  ATGACTGACACGCATTCTATTGCACAACCGTTCGAAGCAGAAGTCT...   \n",
       "\n",
       "                                             Sequence2  \n",
       "0    TTTGAGAGCGTCTCTGTCCCTCGTTTTGCGGTTAAGCCGCATCCAT...  \n",
       "1    TGTTCTATACTTGGGTTCGACTTGGGTTAGACTTGTCTTTACTGTC...  \n",
       "2    CTCGTGCCGAATTCGGCACGAGCGATATTTCATCAGTTATCGGTAA...  \n",
       "3    ATCCTCGGGTCTTGCAGGCCACACAGGACACCCTGAACCGTCATGG...  \n",
       "4    TGAGCAACTTTATTCACATAATTTCTACACCAAGAACTCGAGGTTA...  \n",
       "..                                                 ...  \n",
       "349  ACTTACTACTGTGCCTTTAGGGTGGGGGGATACACCGTTTCCGCCG...  \n",
       "350  ATGAAGAAATTGCTCCCCATTCTTATCGGCCTGAGCCTTTCTGGGT...  \n",
       "351  GAAGAGAAAAGAAAAAAAGACCTTATTTGCTTATCCTGCATCTCTC...  \n",
       "352  TGCCGCTGACGAGGGAGAGGATGAGGATTGCGAGCCAGAGCATCCA...  \n",
       "353  AGAAAAGATCTGATCGTTGAATTGATGACNCTCNANANCACCANNN...  \n",
       "\n",
       "[354 rows x 17 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rnainter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "data = pd.DataFrame()\n",
    "rnainter['Database'] = 'RNAInter'\n",
    "for n, o in zip(['Name', 'ID', 'Category', 'Symbol', 'Sequence', 'Database'], ['Interactor1.Symbol', 'Raw_ID1', 'Category1', 'Interactor1.Symbol', 'Sequence1', 'Database']):\n",
    "    data[n] = rnainter.drop_duplicates(subset=['Interactor1.Symbol'])[o]\n",
    "data2 = pd.DataFrame()\n",
    "for n, o in zip(['Name', 'ID', 'Category', 'Symbol', 'Sequence', 'Database'], ['Interactor2.Symbol', 'Raw_ID2', 'Category2', 'Interactor2.Symbol', 'Sequence2', 'Database']):\n",
    "    data2[n] = rnainter.drop_duplicates(subset=['Interactor2.Symbol'])[o]\n",
    "\n",
    "# data = pd.concat([data, rnainter.drop_duplicates(subset=['Interactor2.Symbol']).reset_index()[['Interactor2.Symbol', 'Raw_ID2', 'Category2', 'Sequence2']].rename(\n",
    "#     columns={j: i for i, j in zip(['Name', 'ID', 'Category', 'Symbol', 'Sequence'], ['Interactor2.Symbol', 'Raw_ID2', 'Category2', 'Interactor2.Symbol', 'Sequence2'])})], axis=0)\n",
    "data = pd.concat([data, data2])\n",
    "data['Alias'] = ''\n",
    "\n",
    "data = data[~data['Name'].isna()]\n",
    "data = data.drop_duplicates(subset=['Name'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Symbol\n",
       "acnA    1\n",
       "phoP    1\n",
       "sibC    1\n",
       "sibB    1\n",
       "sibA    1\n",
       "       ..\n",
       "luxS    1\n",
       "malK    1\n",
       "manX    1\n",
       "manY    1\n",
       "symR    1\n",
       "Name: count, Length: 276, dtype: int64"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data['Symbol'].value_counts()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Now EcoCyc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>Ontology -  all instances of class from My Favorites</th>\n",
       "      <th>Matches</th>\n",
       "      <th>Common-Name</th>\n",
       "      <th>GO-Terms</th>\n",
       "      <th>Names</th>\n",
       "      <th>Object ID</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Sequence - DNA sequence</th>\n",
       "      <th>Right-End-Position</th>\n",
       "      <th>Left-End-Position</th>\n",
       "      <th>Product</th>\n",
       "      <th>Accession-1</th>\n",
       "      <th>Accession-2</th>\n",
       "      <th>Direction</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>6S-RNA</td>\n",
       "      <td>Regulatory-RNAs</td>\n",
       "      <td>6S RNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6S RNA // SsrS // Ssr</td>\n",
       "      <td>6S-RNA</td>\n",
       "      <td>EG30099</td>\n",
       "      <td>ATTTCTCTGAGATGTTCGCAAGCGGGCCAGTCCCCTGAGCCGATAT...</td>\n",
       "      <td>3056165</td>\n",
       "      <td>3055983</td>\n",
       "      <td>6S-RNA</td>\n",
       "      <td>b2911</td>\n",
       "      <td>ECK2906</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>RDLD-RNA</td>\n",
       "      <td>Regulatory-RNAs</td>\n",
       "      <td>antisense regulatory RNA RdlD</td>\n",
       "      <td>GO:2000766 // GO:0030371</td>\n",
       "      <td>antisense regulatory RNA RdlD // RdlD // RdlD ...</td>\n",
       "      <td>RDLD-RNA</td>\n",
       "      <td>G0-9042</td>\n",
       "      <td>GTCTAGAGTCAAGATTAGCCCCCGTGGTGTTGTCAGGTGCATACCT...</td>\n",
       "      <td>3700199</td>\n",
       "      <td>3700136</td>\n",
       "      <td>RDLD-RNA</td>\n",
       "      <td>b4454</td>\n",
       "      <td>ECK3525</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>RNA0-441</td>\n",
       "      <td>Regulatory-RNAs</td>\n",
       "      <td>arfA-derived small regulatory RNA</td>\n",
       "      <td>GO:0006970 // GO:0003729</td>\n",
       "      <td>&lt;i&gt;arfA&lt;/i&gt;-derived small regulatory RNA</td>\n",
       "      <td>RNA0-441</td>\n",
       "      <td>G0-17130</td>\n",
       "      <td>AAGAAAGTAAATCACTTTTTTACCACTGGTCTTCTGCTTTCAGGCG...</td>\n",
       "      <td>3438484</td>\n",
       "      <td>3438431</td>\n",
       "      <td>RNA0-441</td>\n",
       "      <td>b4846</td>\n",
       "      <td>ECK4703</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>RNA0-409</td>\n",
       "      <td>Regulatory-RNAs</td>\n",
       "      <td>e14 prophage; small regulatory RNA C0293</td>\n",
       "      <td>GO:0035194 // GO:0040033 // GO:0003729</td>\n",
       "      <td>e14 prophage; small regulatory RNA C0293 // C0293</td>\n",
       "      <td>RNA0-409</td>\n",
       "      <td>G0-8897</td>\n",
       "      <td>TACTCACCCGGGACTCGCCAGGGGACAGCCAACAGGCATTGGGTGC...</td>\n",
       "      <td>1196782</td>\n",
       "      <td>1196711</td>\n",
       "      <td>RNA0-409</td>\n",
       "      <td>b4806</td>\n",
       "      <td>ECK4663</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>RNA0-162</td>\n",
       "      <td>Regulatory-RNAs</td>\n",
       "      <td>putative antisense regulatory RNA RdlB</td>\n",
       "      <td>GO:1902201</td>\n",
       "      <td>putative antisense regulatory RNA RdlB // RdlB</td>\n",
       "      <td>RNA0-162</td>\n",
       "      <td>G0-9606</td>\n",
       "      <td>GTCTGGTTTCAAGATTAGCCCCCGTTCTGTTGTCAGGTTTTACCTC...</td>\n",
       "      <td>1269923</td>\n",
       "      <td>1269858</td>\n",
       "      <td>RNA0-162</td>\n",
       "      <td>b4422</td>\n",
       "      <td>ECK1207</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Unnamed: 0 Ontology -  all instances of class from My Favorites  \\\n",
       "0           0                                             6S-RNA     \n",
       "1           1                                           RDLD-RNA     \n",
       "2           2                                           RNA0-441     \n",
       "3           3                                           RNA0-409     \n",
       "4           4                                           RNA0-162     \n",
       "\n",
       "           Matches                               Common-Name  \\\n",
       "0  Regulatory-RNAs                                    6S RNA   \n",
       "1  Regulatory-RNAs             antisense regulatory RNA RdlD   \n",
       "2  Regulatory-RNAs         arfA-derived small regulatory RNA   \n",
       "3  Regulatory-RNAs  e14 prophage; small regulatory RNA C0293   \n",
       "4  Regulatory-RNAs    putative antisense regulatory RNA RdlB   \n",
       "\n",
       "                                 GO-Terms  \\\n",
       "0                                     NaN   \n",
       "1                GO:2000766 // GO:0030371   \n",
       "2                GO:0006970 // GO:0003729   \n",
       "3  GO:0035194 // GO:0040033 // GO:0003729   \n",
       "4                              GO:1902201   \n",
       "\n",
       "                                               Names Object ID      Gene  \\\n",
       "0                              6S RNA // SsrS // Ssr    6S-RNA   EG30099   \n",
       "1  antisense regulatory RNA RdlD // RdlD // RdlD ...  RDLD-RNA   G0-9042   \n",
       "2           <i>arfA</i>-derived small regulatory RNA  RNA0-441  G0-17130   \n",
       "3  e14 prophage; small regulatory RNA C0293 // C0293  RNA0-409   G0-8897   \n",
       "4     putative antisense regulatory RNA RdlB // RdlB  RNA0-162   G0-9606   \n",
       "\n",
       "                             Sequence - DNA sequence  Right-End-Position  \\\n",
       "0  ATTTCTCTGAGATGTTCGCAAGCGGGCCAGTCCCCTGAGCCGATAT...             3056165   \n",
       "1  GTCTAGAGTCAAGATTAGCCCCCGTGGTGTTGTCAGGTGCATACCT...             3700199   \n",
       "2  AAGAAAGTAAATCACTTTTTTACCACTGGTCTTCTGCTTTCAGGCG...             3438484   \n",
       "3  TACTCACCCGGGACTCGCCAGGGGACAGCCAACAGGCATTGGGTGC...             1196782   \n",
       "4  GTCTGGTTTCAAGATTAGCCCCCGTTCTGTTGTCAGGTTTTACCTC...             1269923   \n",
       "\n",
       "   Left-End-Position   Product Accession-1 Accession-2 Direction  \n",
       "0            3055983    6S-RNA       b2911     ECK2906         +  \n",
       "1            3700136  RDLD-RNA       b4454     ECK3525         +  \n",
       "2            3438431  RNA0-441       b4846     ECK4703         -  \n",
       "3            1196711  RNA0-409       b4806     ECK4663         +  \n",
       "4            1269858  RNA0-162       b4422     ECK1207         +  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ecocyc.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "ecocyc['Category'] = 'sRNA'\n",
    "ecocyc['Database'] = 'EcoCyc'\n",
    "ecocyc = ecocyc.drop_duplicates(subset=['Common-Name'])\n",
    "\n",
    "ecocyc['Symbol'] = ecocyc['Common-Name']\n",
    "ecocyc['ID'] = 'chromosome:NC_000913.3, Gene ID:' # E. coli NCBI accession\n",
    "ecocyc['Genomic position'] = list(map(lambda l, r: '..'.join([l, r]), ecocyc['Left-End-Position'].apply(str), ecocyc['Right-End-Position'].apply(str)))\n",
    "ecocyc['Alias'] = list(map(lambda x: ';'.join('//'.join([str(i) for i in x]).split('//')).replace(' ; ', ';'), ecocyc[['Names', 'Accession-1', 'Accession-2']].to_numpy()))\n",
    "data = pd.concat([data, ecocyc.rename(columns={j: i for i, j in zip(['Name', 'Sequence', 'Strand'], ['Common-Name', 'Sequence - DNA sequence', 'Direction'])})[\n",
    "    ['Name', 'ID', 'Category', 'Symbol', 'Alias', 'Sequence', 'Database', 'Genomic position', 'Strand']]], axis=0)\n",
    "strand = {'+': 'forward', '-': 'reverse'}\n",
    "data.loc[~data['Strand'].isna(), 'Strand'] = data[~data['Strand'].isna()]['Strand'].apply(lambda x: strand[x] if x else x)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Finally sRNATarBase"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['trans-encoded antisense RNA', 'protein binding small RNA', 'sRNA'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "srnatarbase.loc[srnatarbase['sRNA Type'].isna(), 'sRNA Type'] = 'sRNA'\n",
    "srnatarbase.loc[srnatarbase['Target Type'].isna(), 'Target Type'] = 'mRNA'\n",
    "srnatarbase['sRNA Type'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "      <th>ID</th>\n",
       "      <th>Alias</th>\n",
       "      <th>Genomic position</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Category</th>\n",
       "      <th>Symbol</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Database</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>dsrA</td>\n",
       "      <td>chromosome:NC_000913.3, Gene ID:946470</td>\n",
       "      <td>ECK1952; IS095; JWR0036</td>\n",
       "      <td>2025227..2025313</td>\n",
       "      <td>reverse</td>\n",
       "      <td>sRNA</td>\n",
       "      <td>dsrA</td>\n",
       "      <td>AACACATCAGATTTCCTGGTGTAACGAATTTTTTAAGTGCTTCTTG...</td>\n",
       "      <td>sRNATarBase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>micF</td>\n",
       "      <td>chromosome:NC_000913.3, Gene ID:2847742</td>\n",
       "      <td>ECK2208; IS113; JWR0043; stc</td>\n",
       "      <td>2313084..2313176</td>\n",
       "      <td>forward</td>\n",
       "      <td>sRNA</td>\n",
       "      <td>micF</td>\n",
       "      <td>GCTATCATCATTAACTTTATTTATTACCGTCATTCATTTCTGAATG...</td>\n",
       "      <td>sRNATarBase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>micC</td>\n",
       "      <td>chromosome:NC_000913.3, Gene ID:2847713</td>\n",
       "      <td>ECK1373; IS063; JWR0234; tke8</td>\n",
       "      <td>1437121..1437229</td>\n",
       "      <td>forward</td>\n",
       "      <td>sRNA</td>\n",
       "      <td>micC</td>\n",
       "      <td>GTTATATGCCTTTATTGTCACAGATTTTATTTTCTGTTGGGCCATT...</td>\n",
       "      <td>sRNATarBase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>oxyS</td>\n",
       "      <td>chromosome:NC_000913.3, Gene ID:2847701</td>\n",
       "      <td>ECK3952; JWR0098</td>\n",
       "      <td>4158285..4158394</td>\n",
       "      <td>reverse</td>\n",
       "      <td>sRNA</td>\n",
       "      <td>oxyS</td>\n",
       "      <td>GAAACGGAGCGGCACCTCTTTTAACCCTTGAAGTCACTGCCCGTTT...</td>\n",
       "      <td>sRNATarBase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>istR</td>\n",
       "      <td>chromosome:NC_000913.3, Gene ID:5061525</td>\n",
       "      <td>ECK4425; psr19; psrA19</td>\n",
       "      <td>3853118..3853257</td>\n",
       "      <td>reverse</td>\n",
       "      <td>sRNA</td>\n",
       "      <td>istR</td>\n",
       "      <td>GCACTAAATACGTCAAAATTCGTGCCGAAATTGCGCGTTCTGCGCG...</td>\n",
       "      <td>sRNATarBase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>319</th>\n",
       "      <td>msrA</td>\n",
       "      <td>chromosome:NC_000913.3, Gene ID:948734</td>\n",
       "      <td>ECK4215; JW4178; pms; pmsR</td>\n",
       "      <td>4441538..4442176</td>\n",
       "      <td>reverse</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>msrA</td>\n",
       "      <td>ATGAGTTTATTTGATAAAAAGCATCTGGTTTCCCCCGCCGATGCCC...</td>\n",
       "      <td>sRNATarBase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>320</th>\n",
       "      <td>hdeA</td>\n",
       "      <td>chromosome:NC_000913.3, Gene ID:948025</td>\n",
       "      <td>ECK3494; JW3478; yhhC; yhiB</td>\n",
       "      <td>3656408..3656740</td>\n",
       "      <td>reverse</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>hdeA</td>\n",
       "      <td>ATGAAAAAAGTATTAGGCGTTATTCTTGGTGGTCTGCTTCTTCTGC...</td>\n",
       "      <td>sRNATarBase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>321</th>\n",
       "      <td>hdeB</td>\n",
       "      <td>chromosome:NC_000913.3, Gene ID:948026</td>\n",
       "      <td>ECK3493; JW5669; yhhD; yhiC</td>\n",
       "      <td>3655966..3656292</td>\n",
       "      <td>reverse</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>hdeB</td>\n",
       "      <td>ATGAATATTTCATCTCTCCGTAAAGCGTTTATTTTTATGGGCGCTG...</td>\n",
       "      <td>sRNATarBase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>322</th>\n",
       "      <td>fliC</td>\n",
       "      <td>chromosome:NC_002695.1, Gene ID:</td>\n",
       "      <td>ECs2662</td>\n",
       "      <td>2624379..2626136</td>\n",
       "      <td>reverse</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>fliC</td>\n",
       "      <td>ATGGCACAAGTCATTAATACCAACAGCCTCTCGCTGATCACTCAAA...</td>\n",
       "      <td>sRNATarBase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>323</th>\n",
       "      <td>mreB</td>\n",
       "      <td>chromosome:NC_000913.3, Gene ID:948588</td>\n",
       "      <td>ECK3239; envB; JW3220; mon; rodY</td>\n",
       "      <td>3400044..3401087</td>\n",
       "      <td>reverse</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>mreB</td>\n",
       "      <td>ATGTTGAAAAAATTTCGTGGCATGTTTTCCAATGACTTGTCCATTG...</td>\n",
       "      <td>sRNATarBase</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>324 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Name                                       ID  \\\n",
       "0    dsrA   chromosome:NC_000913.3, Gene ID:946470   \n",
       "1    micF  chromosome:NC_000913.3, Gene ID:2847742   \n",
       "2    micC  chromosome:NC_000913.3, Gene ID:2847713   \n",
       "3    oxyS  chromosome:NC_000913.3, Gene ID:2847701   \n",
       "4    istR  chromosome:NC_000913.3, Gene ID:5061525   \n",
       "..    ...                                      ...   \n",
       "319  msrA   chromosome:NC_000913.3, Gene ID:948734   \n",
       "320  hdeA   chromosome:NC_000913.3, Gene ID:948025   \n",
       "321  hdeB   chromosome:NC_000913.3, Gene ID:948026   \n",
       "322  fliC         chromosome:NC_002695.1, Gene ID:   \n",
       "323  mreB   chromosome:NC_000913.3, Gene ID:948588   \n",
       "\n",
       "                                Alias  Genomic position   Strand Category  \\\n",
       "0             ECK1952; IS095; JWR0036  2025227..2025313  reverse     sRNA   \n",
       "1        ECK2208; IS113; JWR0043; stc  2313084..2313176  forward     sRNA   \n",
       "2       ECK1373; IS063; JWR0234; tke8  1437121..1437229  forward     sRNA   \n",
       "3                    ECK3952; JWR0098  4158285..4158394  reverse     sRNA   \n",
       "4              ECK4425; psr19; psrA19  3853118..3853257  reverse     sRNA   \n",
       "..                                ...               ...      ...      ...   \n",
       "319        ECK4215; JW4178; pms; pmsR  4441538..4442176  reverse     mRNA   \n",
       "320       ECK3494; JW3478; yhhC; yhiB  3656408..3656740  reverse     mRNA   \n",
       "321       ECK3493; JW5669; yhhD; yhiC  3655966..3656292  reverse     mRNA   \n",
       "322                           ECs2662  2624379..2626136  reverse     mRNA   \n",
       "323  ECK3239; envB; JW3220; mon; rodY  3400044..3401087  reverse     mRNA   \n",
       "\n",
       "    Symbol                                           Sequence     Database  \n",
       "0     dsrA  AACACATCAGATTTCCTGGTGTAACGAATTTTTTAAGTGCTTCTTG...  sRNATarBase  \n",
       "1     micF  GCTATCATCATTAACTTTATTTATTACCGTCATTCATTTCTGAATG...  sRNATarBase  \n",
       "2     micC  GTTATATGCCTTTATTGTCACAGATTTTATTTTCTGTTGGGCCATT...  sRNATarBase  \n",
       "3     oxyS  GAAACGGAGCGGCACCTCTTTTAACCCTTGAAGTCACTGCCCGTTT...  sRNATarBase  \n",
       "4     istR  GCACTAAATACGTCAAAATTCGTGCCGAAATTGCGCGTTCTGCGCG...  sRNATarBase  \n",
       "..     ...                                                ...          ...  \n",
       "319   msrA  ATGAGTTTATTTGATAAAAAGCATCTGGTTTCCCCCGCCGATGCCC...  sRNATarBase  \n",
       "320   hdeA  ATGAAAAAAGTATTAGGCGTTATTCTTGGTGGTCTGCTTCTTCTGC...  sRNATarBase  \n",
       "321   hdeB  ATGAATATTTCATCTCTCCGTAAAGCGTTTATTTTTATGGGCGCTG...  sRNATarBase  \n",
       "322   fliC  ATGGCACAAGTCATTAATACCAACAGCCTCTCGCTGATCACTCAAA...  sRNATarBase  \n",
       "323   mreB  ATGTTGAAAAAATTTCGTGGCATGTTTTCCAATGACTTGTCCATTG...  sRNATarBase  \n",
       "\n",
       "[324 rows x 9 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "srnatarbase_merged = pd.DataFrame()\n",
    "srnatarbase_merged['Name'] = np.concatenate(\n",
    "    [srnatarbase['sRNA'].unique(), srnatarbase['Target'].unique()], axis=0)\n",
    "srnatarbase_merged['ID'] = srnatarbase_merged['Name'].apply(\n",
    "    lambda x: srnatarbase[srnatarbase['sRNA'] == x]['sRNA ID'].iloc[0] if x in srnatarbase['sRNA'].tolist()\n",
    "    else srnatarbase[srnatarbase['Target'] == x]['Target ID'].iloc[0])\n",
    "srnatarbase_merged['Alias'] = srnatarbase_merged['Name'].apply(\n",
    "    lambda x: srnatarbase[srnatarbase['sRNA'] == x]['sRNA Alias'].iloc[0] if x in srnatarbase['sRNA'].tolist()\n",
    "    else srnatarbase[srnatarbase['Target'] == x]['Target Alias'].iloc[0])\n",
    "srnatarbase_merged['Genomic position'] = srnatarbase_merged['Name'].apply(\n",
    "    lambda x: srnatarbase[srnatarbase['sRNA'] == x]['sRNA Genome Position'].iloc[0] if x in srnatarbase['sRNA'].tolist()\n",
    "    else srnatarbase[srnatarbase['Target'] == x]['Target Genome Position'].iloc[0])\n",
    "srnatarbase_merged['Strand'] = srnatarbase_merged['Name'].apply(\n",
    "    lambda x: srnatarbase[srnatarbase['sRNA'] == x]['sRNA Strand'].iloc[0] if x in srnatarbase['sRNA'].tolist()\n",
    "    else srnatarbase[srnatarbase['Target'] == x]['Target Strand'].iloc[0])\n",
    "srnatarbase_merged['Category'] = srnatarbase_merged['Name'].apply(\n",
    "    lambda x: 'sRNA' if x in srnatarbase['sRNA'].tolist()\n",
    "    else srnatarbase[srnatarbase['Target'] == x]['Target Type'].iloc[0])\n",
    "srnatarbase_merged['Symbol'] = srnatarbase_merged['Name']\n",
    "srnatarbase_merged['Sequence'] = srnatarbase_merged['Name'].apply(\n",
    "    lambda x: srnatarbase[srnatarbase['sRNA'] == x]['sRNA Sequence'].iloc[0] if x in srnatarbase['sRNA'].tolist()\n",
    "    else srnatarbase[srnatarbase['Target'] == x]['Target Sequence'].iloc[0])\n",
    "srnatarbase_merged['Database'] = 'sRNATarBase'\n",
    "srnatarbase_merged"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "data = pd.concat([data, srnatarbase_merged], axis=0)\n",
    "data = data.reset_index().drop(columns='index')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Delete duplicates responsibly"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "data['Name overlaps'] = list(map(lambda n: [n2 for n2, a in zip(data['Name'], data['Alias']) if (n.lower() in str(a).lower()) or (n.lower() in n2.lower())], data['Name']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Removing small regulatory RNA GlmY\n",
      "Removing small regulatory RNA CsrB\n",
      "Removing 6S RNA\n"
     ]
    }
   ],
   "source": [
    "doubles = []\n",
    "for i, r in data.iterrows():\n",
    "    if len(r['Name overlaps']) > 1:\n",
    "        if len(data[data['Name'].isin(r['Name overlaps'])]['Sequence'].unique()) == 1:\n",
    "            doubles.append(r['Name overlaps'])\n",
    "            \n",
    "for d in doubles:\n",
    "    shortest = np.argmin(list(map(len, d)))\n",
    "    for drop in [i for i in d if i != d[shortest]]:\n",
    "        print('Removing', drop)\n",
    "        data = data.drop(index=data[data['Name'] == drop].index[0])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "278"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# doubles = data['Symbol'].value_counts()[data['Symbol'].value_counts() > 1].index.to_list()\n",
    "doubles = set(data[data['Name overlaps'].apply(len) > 1]['Name'])\n",
    "len(doubles)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "      <th>ID</th>\n",
       "      <th>Category</th>\n",
       "      <th>Symbol</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Database</th>\n",
       "      <th>Alias</th>\n",
       "      <th>Genomic position</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Name overlaps</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>108</th>\n",
       "      <td>micA</td>\n",
       "      <td>NCBI:2847697</td>\n",
       "      <td>sRNA</td>\n",
       "      <td>micA</td>\n",
       "      <td>CAAAAATGGTGTCGCTCGTATTATTGTCGAGAAACCTTTCGGCAAG...</td>\n",
       "      <td>RNAInter</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[micA, small regulatory RNA MicA, MicA]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Name            ID Category Symbol  \\\n",
       "108  micA  NCBI:2847697     sRNA   micA   \n",
       "\n",
       "                                              Sequence  Database Alias  \\\n",
       "108  CAAAAATGGTGTCGCTCGTATTATTGTCGAGAAACCTTTCGGCAAG...  RNAInter         \n",
       "\n",
       "    Genomic position Strand                            Name overlaps  \n",
       "108              NaN    NaN  [micA, small regulatory RNA MicA, MicA]  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data[data['Name'] == 'micA']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Did not remove yfeK\n",
      "Did not remove rrrD\n",
      "Removing duplicate of  malK ['malK']\n",
      "Removing duplicate of  cysE ['cysE']\n",
      "Did not remove yccS\n",
      "Did not remove cydD\n",
      "Did not remove rdlC\n",
      "Did not remove shiA\n",
      "Did not remove ldrB\n",
      "Did not remove gmhB\n",
      "Did not remove tolB\n",
      "Removing duplicate of  MicA ['MicA']\n",
      "Did not remove acrZ\n",
      "Removing duplicate of  pmbA ['pmbA']\n",
      "Removing duplicate of  hyuA ['hyuA']\n",
      "Did not remove folP\n",
      "Removing duplicate of  csgB ['csgB']\n",
      "Did not remove mipA\n",
      "Did not remove yeeE\n",
      "Removing duplicate of  uhpT ['uhpT']\n",
      "Did not remove sokB\n",
      "Did not remove moaD\n",
      "Did not remove sodB\n",
      "Did not remove fdoI\n",
      "Removing duplicate of  rydC ['rydC']\n",
      "Did not remove rdlA\n",
      "Did not remove ompF\n",
      "Did not remove clpB\n",
      "Did not remove xylH\n",
      "Removing duplicate of  ycfL ['ycfL']\n",
      "Removing duplicate of  ygiM ['ygiM']\n",
      "Did not remove msrB\n",
      "Removing duplicate of  marA ['marA']\n",
      "Did not remove citA\n",
      "Removing duplicate of  narK ['narK']\n",
      "Did not remove symR\n",
      "Did not remove arcZ\n",
      "Removing duplicate of  mraZ ['mraZ']\n",
      "Did not remove erpA\n",
      "Did not remove yadD\n",
      "Removing duplicate of  yccE ['yccE']\n",
      "Did not remove ryhB\n",
      "Did not remove icd\n",
      "Did not remove asr\n",
      "Removing duplicate of  micA ['micA']\n",
      "Did not remove nagZ\n",
      "Removing duplicate of  hinT ['hinT']\n",
      "Removing duplicate of  mdh ['mdh']\n",
      "Did not remove sdhA\n",
      "Did not remove sgrS\n",
      "Removing duplicate of  ompT ['ompT']\n",
      "Removing duplicate of  folE ['folE']\n",
      "Did not remove ompN\n",
      "Did not remove lpp\n",
      "Removing duplicate of  MicL ['MicL']\n",
      "Removing duplicate of  hdeB ['hdeB']\n",
      "Did not remove tolR\n",
      "Did not remove yneM\n",
      "Did not remove oxyS\n",
      "Did not remove ybgF\n",
      "Did not remove gtrB\n",
      "Removing duplicate of  csrC ['csrC']\n",
      "Did not remove chiX\n",
      "Did not remove dksA\n",
      "Did not remove yhiM\n",
      "Removing duplicate of  yhjJ ['yhjJ']\n",
      "Did not remove glcD\n",
      "Did not remove fecA\n",
      "Removing duplicate of  gadX ['gadX']\n",
      "Did not remove ybjE\n",
      "Did not remove ybjG\n",
      "Did not remove arrS\n",
      "Did not remove tisB\n",
      "Did not remove manZ\n",
      "Did not remove fadL\n",
      "Removing duplicate of  fimA ['fimA']\n",
      "Did not remove gspD\n",
      "Did not remove gltA\n",
      "Removing duplicate of  symE ['symE']\n",
      "Removing duplicate of  phoE ['phoE']\n",
      "Removing duplicate of  glmZ ['glmZ']\n",
      "Did not remove ydaN\n",
      "Removing duplicate of  ompW ['ompW']\n",
      "Removing duplicate of  yhbE ['yhbE']\n",
      "Did not remove sufD\n",
      "Did not remove iscR\n",
      "Removing duplicate of  rbsD ['rbsD']\n",
      "Did not remove chiP\n",
      "Did not remove deoR\n",
      "Did not remove gfcB\n",
      "Did not remove mnmH\n",
      "Did not remove narH\n",
      "Removing duplicate of  fimB ['fimB']\n",
      "Removing duplicate of  bfr ['bfr']\n",
      "Did not remove tsx\n",
      "Did not remove osmB\n",
      "Did not remove ldtC\n",
      "Did not remove cpxR\n",
      "Removing duplicate of  sugE ['sugE']\n",
      "Removing duplicate of  eptB ['eptB']\n",
      "Did not remove pqqL\n",
      "Did not remove sucD\n",
      "Did not remove uup\n",
      "Did not remove dicF\n",
      "Did not remove fabB\n",
      "Did not remove cheZ\n",
      "Did not remove gadY\n",
      "Removing duplicate of  gadE ['gadE']\n",
      "Did not remove luxS\n",
      "Did not remove cyaR\n",
      "Did not remove ompA\n",
      "Did not remove fryA\n",
      "Did not remove rybB\n",
      "Did not remove rdlD\n",
      "Removing duplicate of  rbsB ['rbsB']\n",
      "Did not remove paoA\n",
      "Did not remove pinH\n",
      "Removing duplicate of  ykgE ['ykgE']\n",
      "Did not remove istR\n",
      "Did not remove ibsC\n",
      "Removing duplicate of  fepA ['fepA']\n",
      "Did not remove rluD\n",
      "Removing duplicate of  pal ['pal']\n",
      "Did not remove kdpA\n",
      "Removing duplicate of  rbsK ['rbsK']\n",
      "Removing duplicate of  cycA ['cycA']\n",
      "Removing duplicate of  fimF ['fimF']\n",
      "Did not remove ygaY\n",
      "Removing duplicate of  ilvI ['ilvI']\n",
      "Removing duplicate of  dps ['dps']\n",
      "Did not remove rlmD\n",
      "Did not remove ldrC\n",
      "Did not remove maeA\n",
      "Did not remove fumA\n",
      "Did not remove yfgO\n",
      "Removing duplicate of  csrA ['csrA']\n",
      "Removing duplicate of  btuB ['btuB']\n",
      "Did not remove yfbT\n",
      "Removing duplicate of  yadS ['yadS']\n",
      "Removing duplicate of  yobA ['yobA']\n",
      "Did not remove nmpC\n",
      "Did not remove mokC\n",
      "Removing duplicate of  metE ['metE']\n",
      "Removing duplicate of  gntP ['gntP']\n",
      "Removing duplicate of  lrp ['lrp']\n",
      "Did not remove trxC\n",
      "Removing duplicate of  csgD ['csgD']\n",
      "Removing duplicate of  RyhB ['RyhB']\n",
      "Did not remove micC\n",
      "Did not remove ftsZ\n",
      "Did not remove rdlB\n",
      "Removing duplicate of  yhcN ['yhcN']\n",
      "Did not remove yagJ\n",
      "Did not remove yiaM\n",
      "Did not remove mgrR\n",
      "Did not remove ydbC\n",
      "Removing duplicate of  sstT ['sstT']\n",
      "Did not remove sibA\n",
      "Did not remove sdhD\n",
      "Removing duplicate of  hdeA ['hdeA']\n",
      "Did not remove sdsR\n",
      "Removing duplicate of  csrB ['csrB']\n",
      "Did not remove folA\n",
      "Did not remove sibE\n",
      "Did not remove ydaM\n",
      "Did not remove galK\n",
      "Removing duplicate of  yjhI ['yjhI']\n",
      "Removing duplicate of  mreB ['mreB']\n",
      "Did not remove mokB\n",
      "Did not remove fnrS\n",
      "Did not remove yegK\n",
      "Did not remove sdhC\n",
      "Removing duplicate of  glmY ['glmY']\n",
      "Removing duplicate of  nirB ['nirB']\n",
      "Did not remove tusD\n",
      "Did not remove fecD\n",
      "Did not remove hokC\n",
      "Did not remove lpxT\n",
      "Did not remove nadE\n",
      "Did not remove ssuC\n",
      "Did not remove ibsA\n",
      "Did not remove yzgL\n",
      "Did not remove argR\n",
      "Removing duplicate of  hokD ['hokD']\n",
      "Did not remove ptsG\n",
      "Removing duplicate of  sodA ['sodA']\n",
      "Did not remove rseX\n",
      "Removing duplicate of  lamB ['lamB']\n",
      "Did not remove ibsE\n",
      "Did not remove flhD\n",
      "Did not remove sibC\n",
      "Did not remove chbC\n",
      "Did not remove fldA\n",
      "Did not remove yeaK\n",
      "Did not remove csiE\n",
      "Removing duplicate of  metH ['metH']\n",
      "Removing duplicate of  msrA ['msrA']\n",
      "Did not remove hofN\n",
      "Removing duplicate of  ssrS ['ssrS']\n",
      "Did not remove yigL\n",
      "Did not remove ykfI\n",
      "Did not remove sokC\n",
      "Removing duplicate of  yhdN ['yhdN']\n",
      "Removing duplicate of  dppA ['dppA']\n",
      "Did not remove rprA\n",
      "Did not remove acnA\n",
      "Did not remove ldrD\n",
      "Did not remove proA\n",
      "Did not remove yqaE\n",
      "Did not remove ohsC\n",
      "Did not remove rcsA\n",
      "Did not remove ydhT\n",
      "Did not remove manX\n",
      "Did not remove yecD\n",
      "Did not remove ompR\n",
      "Removing duplicate of  yphD ['yphD']\n",
      "Did not remove ydeN\n",
      "Did not remove ygiQ\n",
      "Did not remove tsaB\n",
      "Did not remove ecnB\n",
      "Did not remove fur\n",
      "Did not remove folX\n",
      "Did not remove ompX\n",
      "Did not remove gpmA\n",
      "Did not remove yaeP\n",
      "Did not remove yaeH\n",
      "Did not remove rpmG\n",
      "Did not remove fumC\n",
      "Did not remove fiu\n",
      "Did not remove sibB\n",
      "Removing duplicate of  hns ['hns']\n",
      "Did not remove metI\n",
      "Did not remove ompC\n",
      "Removing duplicate of  oppA ['oppA']\n",
      "Did not remove ibsD\n",
      "Did not remove ibsB\n",
      "Removing duplicate of  rraB ['rraB']\n",
      "Did not remove ybeT\n",
      "Did not remove dppD\n",
      "Did not remove yadL\n",
      "Did not remove phoP\n",
      "Did not remove wrbA\n",
      "Did not remove mutM\n",
      "Did not remove yeaC\n",
      "Did not remove manY\n",
      "Did not remove gcvB\n",
      "Did not remove hokB\n",
      "Did not remove omrA\n",
      "Did not remove sucC\n",
      "Did not remove micF\n",
      "Did not remove sibD\n",
      "Did not remove motA\n",
      "Removing duplicate of  srlB ['srlB']\n",
      "Did not remove omrB\n",
      "Did not remove gloA\n",
      "Did not remove iscS\n",
      "Did not remove mcaS\n",
      "Did not remove nadA\n",
      "Did not remove yobF\n",
      "Removing duplicate of  yaiY ['yaiY']\n",
      "Removing duplicate of  mqsA ['mqsA']\n",
      "Removing duplicate of  lit ['lit']\n",
      "Did not remove dsrA\n",
      "Did not remove ygdQ\n",
      "Did not remove citG\n",
      "Did not remove shoB\n",
      "Did not remove ynfF\n",
      "Removing duplicate of  glmM ['glmM']\n",
      "Removing duplicate of  cirA ['cirA']\n",
      "Did not remove gadB\n",
      "Did not remove ykgJ\n",
      "Did not remove ftnA\n",
      "Did not remove higB\n",
      "Did not remove ldrA\n",
      "Did not remove hisM\n",
      "Did not remove tolA\n",
      "Did not remove rpoS\n",
      "Did not remove sucB\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "      <th>ID</th>\n",
       "      <th>Category</th>\n",
       "      <th>Symbol</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Database</th>\n",
       "      <th>Alias</th>\n",
       "      <th>Genomic position</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Name overlaps</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>acnA</td>\n",
       "      <td>NCBI:946724</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>acnA</td>\n",
       "      <td>ATTCGGAACGAGGCCTGAAGCAGTGTCGCCGTCCCTCTGCCTTGCA...</td>\n",
       "      <td>RNAInter</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[acnA, acnA]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>acrZ</td>\n",
       "      <td>NCBI:945365</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>acrZ</td>\n",
       "      <td>ACTTACTACTGTCTTCGGGGGGTCCGAGGTTTCTGGGGGGTCGTAC...</td>\n",
       "      <td>RNAInter</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[acrZ, acrZ]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>arcZ</td>\n",
       "      <td>NCBI:2847690</td>\n",
       "      <td>sRNA</td>\n",
       "      <td>arcZ</td>\n",
       "      <td>CTCATGTTGACCGCTTGTTTAGCAGCTTCAAGGAAGCTGAAGGGCA...</td>\n",
       "      <td>RNAInter</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[arcZ, small regulatory RNA ArcZ, arcZ]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>argR</td>\n",
       "      <td>NCBI:947861</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>argR</td>\n",
       "      <td>GACAATGGCGATAGTATGGCGGTTGTTTCTTTCCCATCTCTACTCA...</td>\n",
       "      <td>RNAInter</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[argR, argR]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>arrS</td>\n",
       "      <td>NCBI:11115378</td>\n",
       "      <td>sRNA</td>\n",
       "      <td>arrS</td>\n",
       "      <td>CTCTCCCTCTCTTTCTCTCTTCTCCGCGGCGATACGATCCGAGATG...</td>\n",
       "      <td>RNAInter</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[arrS, small regulatory RNA ArrS, arrS]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>670</th>\n",
       "      <td>msrA</td>\n",
       "      <td>chromosome:NC_000913.3, Gene ID:948734</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>msrA</td>\n",
       "      <td>ATGAGTTTATTTGATAAAAAGCATCTGGTTTCCCCCGCCGATGCCC...</td>\n",
       "      <td>sRNATarBase</td>\n",
       "      <td>ECK4215; JW4178; pms; pmsR</td>\n",
       "      <td>4441538..4442176</td>\n",
       "      <td>reverse</td>\n",
       "      <td>[msrA, msrA]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>671</th>\n",
       "      <td>hdeA</td>\n",
       "      <td>chromosome:NC_000913.3, Gene ID:948025</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>hdeA</td>\n",
       "      <td>ATGAAAAAAGTATTAGGCGTTATTCTTGGTGGTCTGCTTCTTCTGC...</td>\n",
       "      <td>sRNATarBase</td>\n",
       "      <td>ECK3494; JW3478; yhhC; yhiB</td>\n",
       "      <td>3656408..3656740</td>\n",
       "      <td>reverse</td>\n",
       "      <td>[hdeA, hdeA]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>672</th>\n",
       "      <td>hdeB</td>\n",
       "      <td>chromosome:NC_000913.3, Gene ID:948026</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>hdeB</td>\n",
       "      <td>ATGAATATTTCATCTCTCCGTAAAGCGTTTATTTTTATGGGCGCTG...</td>\n",
       "      <td>sRNATarBase</td>\n",
       "      <td>ECK3493; JW5669; yhhD; yhiC</td>\n",
       "      <td>3655966..3656292</td>\n",
       "      <td>reverse</td>\n",
       "      <td>[hdeB, hdeB]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>673</th>\n",
       "      <td>fliC</td>\n",
       "      <td>chromosome:NC_002695.1, Gene ID:</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>fliC</td>\n",
       "      <td>ATGGCACAAGTCATTAATACCAACAGCCTCTCGCTGATCACTCAAA...</td>\n",
       "      <td>sRNATarBase</td>\n",
       "      <td>ECs2662</td>\n",
       "      <td>2624379..2626136</td>\n",
       "      <td>reverse</td>\n",
       "      <td>[fliC]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>674</th>\n",
       "      <td>mreB</td>\n",
       "      <td>chromosome:NC_000913.3, Gene ID:948588</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>mreB</td>\n",
       "      <td>ATGTTGAAAAAATTTCGTGGCATGTTTTCCAATGACTTGTCCATTG...</td>\n",
       "      <td>sRNATarBase</td>\n",
       "      <td>ECK3239; envB; JW3220; mon; rodY</td>\n",
       "      <td>3400044..3401087</td>\n",
       "      <td>reverse</td>\n",
       "      <td>[mreB, mreB]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>590 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Name                                      ID Category Symbol  \\\n",
       "0    acnA                             NCBI:946724     mRNA   acnA   \n",
       "1    acrZ                             NCBI:945365     mRNA   acrZ   \n",
       "2    arcZ                            NCBI:2847690     sRNA   arcZ   \n",
       "3    argR                             NCBI:947861     mRNA   argR   \n",
       "4    arrS                           NCBI:11115378     sRNA   arrS   \n",
       "..    ...                                     ...      ...    ...   \n",
       "670  msrA  chromosome:NC_000913.3, Gene ID:948734     mRNA   msrA   \n",
       "671  hdeA  chromosome:NC_000913.3, Gene ID:948025     mRNA   hdeA   \n",
       "672  hdeB  chromosome:NC_000913.3, Gene ID:948026     mRNA   hdeB   \n",
       "673  fliC        chromosome:NC_002695.1, Gene ID:     mRNA   fliC   \n",
       "674  mreB  chromosome:NC_000913.3, Gene ID:948588     mRNA   mreB   \n",
       "\n",
       "                                              Sequence     Database  \\\n",
       "0    ATTCGGAACGAGGCCTGAAGCAGTGTCGCCGTCCCTCTGCCTTGCA...     RNAInter   \n",
       "1    ACTTACTACTGTCTTCGGGGGGTCCGAGGTTTCTGGGGGGTCGTAC...     RNAInter   \n",
       "2    CTCATGTTGACCGCTTGTTTAGCAGCTTCAAGGAAGCTGAAGGGCA...     RNAInter   \n",
       "3    GACAATGGCGATAGTATGGCGGTTGTTTCTTTCCCATCTCTACTCA...     RNAInter   \n",
       "4    CTCTCCCTCTCTTTCTCTCTTCTCCGCGGCGATACGATCCGAGATG...     RNAInter   \n",
       "..                                                 ...          ...   \n",
       "670  ATGAGTTTATTTGATAAAAAGCATCTGGTTTCCCCCGCCGATGCCC...  sRNATarBase   \n",
       "671  ATGAAAAAAGTATTAGGCGTTATTCTTGGTGGTCTGCTTCTTCTGC...  sRNATarBase   \n",
       "672  ATGAATATTTCATCTCTCCGTAAAGCGTTTATTTTTATGGGCGCTG...  sRNATarBase   \n",
       "673  ATGGCACAAGTCATTAATACCAACAGCCTCTCGCTGATCACTCAAA...  sRNATarBase   \n",
       "674  ATGTTGAAAAAATTTCGTGGCATGTTTTCCAATGACTTGTCCATTG...  sRNATarBase   \n",
       "\n",
       "                                Alias  Genomic position   Strand  \\\n",
       "0                                                   NaN      NaN   \n",
       "1                                                   NaN      NaN   \n",
       "2                                                   NaN      NaN   \n",
       "3                                                   NaN      NaN   \n",
       "4                                                   NaN      NaN   \n",
       "..                                ...               ...      ...   \n",
       "670        ECK4215; JW4178; pms; pmsR  4441538..4442176  reverse   \n",
       "671       ECK3494; JW3478; yhhC; yhiB  3656408..3656740  reverse   \n",
       "672       ECK3493; JW5669; yhhD; yhiC  3655966..3656292  reverse   \n",
       "673                           ECs2662  2624379..2626136  reverse   \n",
       "674  ECK3239; envB; JW3220; mon; rodY  3400044..3401087  reverse   \n",
       "\n",
       "                               Name overlaps  \n",
       "0                               [acnA, acnA]  \n",
       "1                               [acrZ, acrZ]  \n",
       "2    [arcZ, small regulatory RNA ArcZ, arcZ]  \n",
       "3                               [argR, argR]  \n",
       "4    [arrS, small regulatory RNA ArrS, arrS]  \n",
       "..                                       ...  \n",
       "670                             [msrA, msrA]  \n",
       "671                             [hdeA, hdeA]  \n",
       "672                             [hdeB, hdeB]  \n",
       "673                                   [fliC]  \n",
       "674                             [mreB, mreB]  \n",
       "\n",
       "[590 rows x 10 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "for d in doubles:\n",
    "    \n",
    "    if len(data[data['Symbol'] == d]['Sequence'].unique()) == 1:\n",
    "        print('Removing duplicate of ', d, data[data['Symbol'] == d]['Name'].unique())\n",
    "        data = data.drop(index=data[(data['Symbol'] == d) & (data['Database'] != 'EcoCyc')].index[0])\n",
    "    else:\n",
    "        print('Did not remove', d)\n",
    "        # print(data[data['Symbol'] == d]['Sequence'].values)\n",
    "    \n",
    "data = data[~data['Sequence'].isna()]\n",
    "data\n",
    "# data[data['Symbol'] == d].index[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Surrounding sequences\n",
    "\n",
    "Since some of the interactions between sRNAs and mRNAs are upstream of the mRNA transcription or translation start site."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import Entrez\n",
    "from urllib.error import HTTPError\n",
    "\n",
    "\n",
    "Entrez.email = \"olivia.gallup@gmail.com\"  # Add your email address\n",
    "Entrez.api_key = \"8167226cf01abaa985a1c23d9b5283d22208\"  # Add your NCBI API key\n",
    "\n",
    "\n",
    "def is_complement(gene_id):\n",
    "    handle = Entrez.efetch(db=\"gene\", id=gene_id, rettype=\"txt\", retmode=\"gb\")\n",
    "    record = handle.read()\n",
    "    if 'complement' in record:\n",
    "        return True\n",
    "    return False\n",
    "\n",
    "\n",
    "def get_seq_from_genome(gene_id: int, thresh: int, before: bool, dna_start: int = None, dna_end: int = None, nuc_id: str = None, strand: str = None):\n",
    "\n",
    "    try:\n",
    "        if not strand:\n",
    "            is_comp = is_complement(gene_id)\n",
    "        else:\n",
    "            is_comp = True if strand == 'reverse' else False\n",
    "        if not dna_start and (not dna_end):\n",
    "            handle = Entrez.efetch(db=\"gene\", id=gene_id,\n",
    "                                rettype=\"fasta\", retmode=\"xml\")\n",
    "            record = Entrez.read(handle)\n",
    "            if not nuc_id:\n",
    "                nuc_id = record[0]['Entrezgene_locus'][0]['Gene-commentary_accession'] + \\\n",
    "                    '.' + record[0]['Entrezgene_locus'][0]['Gene-commentary_version']\n",
    "\n",
    "            if (before and not(is_comp)) or (is_comp and not(before)):\n",
    "                dna_start = int(record[0]['Entrezgene_locus'][0]['Gene-commentary_seqs']\n",
    "                                [0]['Seq-loc_int']['Seq-interval']['Seq-interval_from']) - thresh\n",
    "                dna_end = int(record[0]['Entrezgene_locus'][0]['Gene-commentary_seqs']\n",
    "                            [0]['Seq-loc_int']['Seq-interval']['Seq-interval_from'])\n",
    "            if (is_comp and before) or (not(is_comp) and not(before)):\n",
    "                dna_start = int(record[0]['Entrezgene_locus'][0]['Gene-commentary_seqs']\n",
    "                                [0]['Seq-loc_int']['Seq-interval']['Seq-interval_to'])\n",
    "                dna_end = int(record[0]['Entrezgene_locus'][0]['Gene-commentary_seqs']\n",
    "                            [0]['Seq-loc_int']['Seq-interval']['Seq-interval_to']) + thresh\n",
    "\n",
    "        search_handle = Entrez.efetch(db=\"nuccore\", id=nuc_id, retmode=\"text\", rettype='gb',\n",
    "                                      strand=\"complement\" if is_comp else '', seq_start=int(dna_start), seq_stop=int(dna_end))\n",
    "        # dna_sequence = SeqIO.read(search_handle, \"fasta\").seq\n",
    "        r = search_handle.read()\n",
    "        if r:\n",
    "            if len(r.split('>')) == 1 or (len(r.split('>')[1].split('\\n')) <= 1):\n",
    "                dna_sequence = ''\n",
    "            # dna_sequence = ''.join(r.split('>')[1].split('\\n')[1:])\n",
    "            dna_sequence = re.sub('[^atcgACTG]', '', (r.split('ORIGIN')[-1]))\n",
    "        else:\n",
    "            dna_sequence = ''\n",
    "        return dna_sequence\n",
    "    except HTTPError:\n",
    "        return ''\n",
    "\n",
    "\n",
    "def get_seq_info(gene_id: int, thresh: int = 200, before: bool = True, dna_start: int = None, dna_end: int = None, nuc_id: str = None, strand: str = None):\n",
    "    try:\n",
    "        handle = Entrez.efetch(db=\"nucleotide\", id=gene_id,\n",
    "                               rettype=\"fasta\", retmode=\"text\")\n",
    "        fast = handle.read()\n",
    "        seq = ''.join(fast.split('\\n')[1:])\n",
    "        if not seq or ('N' in seq):\n",
    "            seq = get_seq_from_genome(gene_id, thresh, before, dna_start, dna_end, nuc_id, strand)\n",
    "        return seq\n",
    "    except HTTPError:\n",
    "        seq = get_seq_from_genome(gene_id, thresh, before, dna_start, dna_end, nuc_id, strand)\n",
    "        return seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "data['Raw ID'] = data['ID'].str.split(':').apply(lambda x: x[-1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total: 87 \n",
      "\n"
     ]
    }
   ],
   "source": [
    "from Bio import Entrez\n",
    "\n",
    "def find_gene_accession(species_number, gene_name):\n",
    "    # Provide your email to NCBI\n",
    "    Entrez.email = \"your.email@example.com\"\n",
    "\n",
    "    try:\n",
    "        # Search for gene information related to the specified species number\n",
    "        search_term = f\"{species_number}[Accession] AND {gene_name}\"\n",
    "        handle = Entrez.esearch(db=\"nucleotide\", term=search_term)\n",
    "        record = Entrez.read(handle)\n",
    "        handle.close()\n",
    "\n",
    "        # Retrieve the first accession number found\n",
    "        if record[\"Count\"] != \"0\":\n",
    "            accession_number = record[\"IdList\"][0]\n",
    "            return int(accession_number)\n",
    "        else:\n",
    "            return None\n",
    "\n",
    "    except Exception as e:\n",
    "        return f\"An error occurred: {e}\"\n",
    "\n",
    "\n",
    "print(f'Total:', len(data[(data['Raw ID'] == '') | (data['Raw ID'].isna())]), '\\n')\n",
    "for i, interactors in data[(data['Raw ID'] == '') | (data['Raw ID'].isna())].iterrows():\n",
    "    gene_name = interactors['Name']\n",
    "    species_number = interactors['ID'].split(',')[0].split(':')[-1]\n",
    "    accession_number = find_gene_accession(species_number, gene_name)\n",
    "    if type(accession_number) != int:\n",
    "        alias = data[data['Name'] == gene_name]['Alias'].iloc[0]\n",
    "        if type(alias) == str:\n",
    "            gene_names = [ii.strip(' ') for ii in alias.split(';') if ii.strip(' ')]\n",
    "            cand_accession_numbers = []\n",
    "            for g in gene_names:\n",
    "                accession_number = find_gene_accession(species_number, gene_name)\n",
    "                if type(accession_number) != int:\n",
    "                    cand_accession_numbers.append(accession_number)\n",
    "            if not all(element == cand_accession_numbers[0] for element in cand_accession_numbers) or (type(cand_accession_numbers[0]) == str):\n",
    "                print(f'Warning: the accession number for {gene_name} found through aliases {gene_names} did not match: {cand_accession_numbers}')\n",
    "            accession_number = cand_accession_numbers[0]\n",
    "    \n",
    "    data['Raw ID'].loc[i] = accession_number\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "      <th>ID</th>\n",
       "      <th>Category</th>\n",
       "      <th>Symbol</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Database</th>\n",
       "      <th>Alias</th>\n",
       "      <th>Genomic position</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Name overlaps</th>\n",
       "      <th>Raw ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>285</th>\n",
       "      <td>small antisense RNA aMEF</td>\n",
       "      <td>chromosome:NC_000913.3, Gene ID:</td>\n",
       "      <td>sRNA</td>\n",
       "      <td>small antisense RNA aMEF</td>\n",
       "      <td>CAAATCAGATCGCCCATATCGGGTACGTATCGGCTTACCATTACCA...</td>\n",
       "      <td>EcoCyc</td>\n",
       "      <td>small antisense RNA aMEF;aMEF;b4844;ECK4701</td>\n",
       "      <td>2911051..2911450</td>\n",
       "      <td>forward</td>\n",
       "      <td>[small antisense RNA aMEF]</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>286</th>\n",
       "      <td>small antisense RNA AsCRP</td>\n",
       "      <td>chromosome:NC_000913.3, Gene ID:</td>\n",
       "      <td>sRNA</td>\n",
       "      <td>small antisense RNA AsCRP</td>\n",
       "      <td>CCGCTTTTTCACCCTGGTGAATAAGCGTGCTCTTGGATGGGTACTT...</td>\n",
       "      <td>EcoCyc</td>\n",
       "      <td>small antisense RNA AsCRP;b4841;ECK4698</td>\n",
       "      <td>3485953..3486231</td>\n",
       "      <td>reverse</td>\n",
       "      <td>[small antisense RNA AsCRP]</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>287</th>\n",
       "      <td>small antisense RNA AsflhD</td>\n",
       "      <td>chromosome:NC_000913.3, Gene ID:</td>\n",
       "      <td>sRNA</td>\n",
       "      <td>small antisense RNA AsflhD</td>\n",
       "      <td>TCAGCAACTCGGAGGTATGCATTATTCCCACCCAGAATAACCAACT...</td>\n",
       "      <td>EcoCyc</td>\n",
       "      <td>small antisense RNA AsflhD;AsflhD;b4839;ECK4696</td>\n",
       "      <td>1978176..1978395</td>\n",
       "      <td>forward</td>\n",
       "      <td>[small antisense RNA AsflhD]</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>288</th>\n",
       "      <td>small antisense RNA AsphoP</td>\n",
       "      <td>chromosome:NC_000913.3, Gene ID:</td>\n",
       "      <td>sRNA</td>\n",
       "      <td>small antisense RNA AsphoP</td>\n",
       "      <td>CACCGGCACTTAATACTTCGACTTTGTCCTGCCAGCTTTCACGGGC...</td>\n",
       "      <td>EcoCyc</td>\n",
       "      <td>small antisense RNA AsphoP;b4843;ECK4700</td>\n",
       "      <td>1190165..1190508</td>\n",
       "      <td>forward</td>\n",
       "      <td>[small antisense RNA AsphoP]</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>289</th>\n",
       "      <td>small antisense RNA StfZ</td>\n",
       "      <td>chromosome:NC_000913.3, Gene ID:</td>\n",
       "      <td>sRNA</td>\n",
       "      <td>small antisense RNA StfZ</td>\n",
       "      <td>GGTGATACCGCTACCGATTTGAATCGTCTGTCCAACCGCTGTTTTA...</td>\n",
       "      <td>EcoCyc</td>\n",
       "      <td>small antisense RNA StfZ;StfZ;b4840;ECK4697</td>\n",
       "      <td>104948..105499</td>\n",
       "      <td>reverse</td>\n",
       "      <td>[small antisense RNA StfZ]</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>298</th>\n",
       "      <td>small regulatory RNA AzuR</td>\n",
       "      <td>chromosome:NC_000913.3, Gene ID:</td>\n",
       "      <td>sRNA</td>\n",
       "      <td>small regulatory RNA AzuR</td>\n",
       "      <td>AAAGACAATAACACCTGTATAACAAATGGTCGGAGTGCCGCGATGA...</td>\n",
       "      <td>EcoCyc</td>\n",
       "      <td>small regulatory RNA AzuR;AzuR;b4434;ECK1902</td>\n",
       "      <td>1987843..1988001</td>\n",
       "      <td>reverse</td>\n",
       "      <td>[small regulatory RNA AzuR]</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>314</th>\n",
       "      <td>small regulatory RNA GlnZ</td>\n",
       "      <td>chromosome:NC_000913.3, Gene ID:</td>\n",
       "      <td>sRNA</td>\n",
       "      <td>small regulatory RNA GlnZ</td>\n",
       "      <td>TTTTAGTTGCCGTGGAAACTTTTCGCCTGTCTCTGGCAGGCCTGGG...</td>\n",
       "      <td>EcoCyc</td>\n",
       "      <td>small regulatory RNA GlnZ;glnA 3&lt;sup&gt;A&lt;/sup&gt; U...</td>\n",
       "      <td>4056428..4056621</td>\n",
       "      <td>reverse</td>\n",
       "      <td>[small regulatory RNA GlnZ]</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>324</th>\n",
       "      <td>small regulatory RNA Och5</td>\n",
       "      <td>chromosome:NC_000913.3, Gene ID:</td>\n",
       "      <td>sRNA</td>\n",
       "      <td>small regulatory RNA Och5</td>\n",
       "      <td>ATCCTTGTCCGCTTAACCGTTTCAGTCAGCCTATCCTTGATGAAAC...</td>\n",
       "      <td>EcoCyc</td>\n",
       "      <td>small regulatory RNA Och5;Och5;nan;nan</td>\n",
       "      <td>3067187..3067344</td>\n",
       "      <td>forward</td>\n",
       "      <td>[small regulatory RNA Och5]</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>363</th>\n",
       "      <td>Spot_42</td>\n",
       "      <td>chromosome:NC_000913.3, Gene ID:</td>\n",
       "      <td>sRNA</td>\n",
       "      <td>Spot_42</td>\n",
       "      <td>GTAGGGTACAGAGGTAAGATGTTCTATCTTTCAGACCTTTTACTTC...</td>\n",
       "      <td>sRNATarBase</td>\n",
       "      <td>spf ; ECK3856; IS197; JWR0097;</td>\n",
       "      <td>4049899..4050007</td>\n",
       "      <td>forward</td>\n",
       "      <td>[Spot_42]</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>390</th>\n",
       "      <td>OrzO</td>\n",
       "      <td>chromosome:NC_002695.1, Gene ID:</td>\n",
       "      <td>sRNA</td>\n",
       "      <td>OrzO</td>\n",
       "      <td>GTTGGTACGAAATGTTGCACAATCAGCAAGTTAGCGTCACTGTAGT...</td>\n",
       "      <td>sRNATarBase</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2883467..2883547</td>\n",
       "      <td>forward</td>\n",
       "      <td>[OrzO]</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>395</th>\n",
       "      <td>IpeX</td>\n",
       "      <td>chromosome:NC_000913.3, Gene ID:</td>\n",
       "      <td>sRNA</td>\n",
       "      <td>IpeX</td>\n",
       "      <td>TAATCTGATTACGAAAAAGATATGTTGCGGGAGGCGTTGCCTCCCC...</td>\n",
       "      <td>sRNATarBase</td>\n",
       "      <td>NaN</td>\n",
       "      <td>573588..573754</td>\n",
       "      <td>reverse</td>\n",
       "      <td>[IpeX]</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>401</th>\n",
       "      <td>AfaR</td>\n",
       "      <td>chromosome:AF072900.3, Gene ID:</td>\n",
       "      <td>sRNA</td>\n",
       "      <td>AfaR</td>\n",
       "      <td>TAGCTACATATGTTCCATAACAGACGTCCACTTTTTGATGAAAAGG...</td>\n",
       "      <td>sRNATarBase</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5301..5534</td>\n",
       "      <td>reverse</td>\n",
       "      <td>[AfaR]</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>402</th>\n",
       "      <td>Esr41</td>\n",
       "      <td>chromosome:NC_002695.1, Gene ID:</td>\n",
       "      <td>sRNA</td>\n",
       "      <td>Esr41</td>\n",
       "      <td>GATGCTCTAGGCATCACATTTTCTCCATGGGGTATTCCCTCCGCCG...</td>\n",
       "      <td>sRNATarBase</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1422406..1422479</td>\n",
       "      <td>forward</td>\n",
       "      <td>[Esr41]</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>615</th>\n",
       "      <td>zorO</td>\n",
       "      <td>chromosome:NC_002695.1, Gene ID:</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>zorO</td>\n",
       "      <td>GTTGGGAGGTTGTGCCGGATCGGACCAATTTTAAGTCCTGGCTGCC...</td>\n",
       "      <td>sRNATarBase</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2883139..2883402</td>\n",
       "      <td>reverse</td>\n",
       "      <td>[zorO]</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>656</th>\n",
       "      <td>sdhCDAB</td>\n",
       "      <td>chromosome:NC_000913.3, Gene ID:</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>sdhCDAB</td>\n",
       "      <td>ATGATAAGAAATGTGAAAAAACAAAGACCTGTTAATCTGGACCTAC...</td>\n",
       "      <td>sRNATarBase</td>\n",
       "      <td>NaN</td>\n",
       "      <td>755177..758405</td>\n",
       "      <td>forward</td>\n",
       "      <td>[sdhCDAB]</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>658</th>\n",
       "      <td>yliL</td>\n",
       "      <td>chromosome:NC_000913.3, Gene ID:</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>yliL</td>\n",
       "      <td>GTGGAAAGGGGTCCGCATCCGCGAGCCGCAATATACGCAGGGCTGC...</td>\n",
       "      <td>sRNATarBase</td>\n",
       "      <td>ECK0805; JW5969</td>\n",
       "      <td>851894..852163</td>\n",
       "      <td>forward</td>\n",
       "      <td>[yliL]</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>673</th>\n",
       "      <td>fliC</td>\n",
       "      <td>chromosome:NC_002695.1, Gene ID:</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>fliC</td>\n",
       "      <td>ATGGCACAAGTCATTAATACCAACAGCCTCTCGCTGATCACTCAAA...</td>\n",
       "      <td>sRNATarBase</td>\n",
       "      <td>ECs2662</td>\n",
       "      <td>2624379..2626136</td>\n",
       "      <td>reverse</td>\n",
       "      <td>[fliC]</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           Name                                ID Category  \\\n",
       "285    small antisense RNA aMEF  chromosome:NC_000913.3, Gene ID:     sRNA   \n",
       "286   small antisense RNA AsCRP  chromosome:NC_000913.3, Gene ID:     sRNA   \n",
       "287  small antisense RNA AsflhD  chromosome:NC_000913.3, Gene ID:     sRNA   \n",
       "288  small antisense RNA AsphoP  chromosome:NC_000913.3, Gene ID:     sRNA   \n",
       "289    small antisense RNA StfZ  chromosome:NC_000913.3, Gene ID:     sRNA   \n",
       "298   small regulatory RNA AzuR  chromosome:NC_000913.3, Gene ID:     sRNA   \n",
       "314   small regulatory RNA GlnZ  chromosome:NC_000913.3, Gene ID:     sRNA   \n",
       "324   small regulatory RNA Och5  chromosome:NC_000913.3, Gene ID:     sRNA   \n",
       "363                     Spot_42  chromosome:NC_000913.3, Gene ID:     sRNA   \n",
       "390                        OrzO  chromosome:NC_002695.1, Gene ID:     sRNA   \n",
       "395                        IpeX  chromosome:NC_000913.3, Gene ID:     sRNA   \n",
       "401                        AfaR   chromosome:AF072900.3, Gene ID:     sRNA   \n",
       "402                       Esr41  chromosome:NC_002695.1, Gene ID:     sRNA   \n",
       "615                        zorO  chromosome:NC_002695.1, Gene ID:     mRNA   \n",
       "656                     sdhCDAB  chromosome:NC_000913.3, Gene ID:     mRNA   \n",
       "658                        yliL  chromosome:NC_000913.3, Gene ID:     mRNA   \n",
       "673                        fliC  chromosome:NC_002695.1, Gene ID:     mRNA   \n",
       "\n",
       "                         Symbol  \\\n",
       "285    small antisense RNA aMEF   \n",
       "286   small antisense RNA AsCRP   \n",
       "287  small antisense RNA AsflhD   \n",
       "288  small antisense RNA AsphoP   \n",
       "289    small antisense RNA StfZ   \n",
       "298   small regulatory RNA AzuR   \n",
       "314   small regulatory RNA GlnZ   \n",
       "324   small regulatory RNA Och5   \n",
       "363                     Spot_42   \n",
       "390                        OrzO   \n",
       "395                        IpeX   \n",
       "401                        AfaR   \n",
       "402                       Esr41   \n",
       "615                        zorO   \n",
       "656                     sdhCDAB   \n",
       "658                        yliL   \n",
       "673                        fliC   \n",
       "\n",
       "                                              Sequence     Database  \\\n",
       "285  CAAATCAGATCGCCCATATCGGGTACGTATCGGCTTACCATTACCA...       EcoCyc   \n",
       "286  CCGCTTTTTCACCCTGGTGAATAAGCGTGCTCTTGGATGGGTACTT...       EcoCyc   \n",
       "287  TCAGCAACTCGGAGGTATGCATTATTCCCACCCAGAATAACCAACT...       EcoCyc   \n",
       "288  CACCGGCACTTAATACTTCGACTTTGTCCTGCCAGCTTTCACGGGC...       EcoCyc   \n",
       "289  GGTGATACCGCTACCGATTTGAATCGTCTGTCCAACCGCTGTTTTA...       EcoCyc   \n",
       "298  AAAGACAATAACACCTGTATAACAAATGGTCGGAGTGCCGCGATGA...       EcoCyc   \n",
       "314  TTTTAGTTGCCGTGGAAACTTTTCGCCTGTCTCTGGCAGGCCTGGG...       EcoCyc   \n",
       "324  ATCCTTGTCCGCTTAACCGTTTCAGTCAGCCTATCCTTGATGAAAC...       EcoCyc   \n",
       "363  GTAGGGTACAGAGGTAAGATGTTCTATCTTTCAGACCTTTTACTTC...  sRNATarBase   \n",
       "390  GTTGGTACGAAATGTTGCACAATCAGCAAGTTAGCGTCACTGTAGT...  sRNATarBase   \n",
       "395  TAATCTGATTACGAAAAAGATATGTTGCGGGAGGCGTTGCCTCCCC...  sRNATarBase   \n",
       "401  TAGCTACATATGTTCCATAACAGACGTCCACTTTTTGATGAAAAGG...  sRNATarBase   \n",
       "402  GATGCTCTAGGCATCACATTTTCTCCATGGGGTATTCCCTCCGCCG...  sRNATarBase   \n",
       "615  GTTGGGAGGTTGTGCCGGATCGGACCAATTTTAAGTCCTGGCTGCC...  sRNATarBase   \n",
       "656  ATGATAAGAAATGTGAAAAAACAAAGACCTGTTAATCTGGACCTAC...  sRNATarBase   \n",
       "658  GTGGAAAGGGGTCCGCATCCGCGAGCCGCAATATACGCAGGGCTGC...  sRNATarBase   \n",
       "673  ATGGCACAAGTCATTAATACCAACAGCCTCTCGCTGATCACTCAAA...  sRNATarBase   \n",
       "\n",
       "                                                 Alias  Genomic position  \\\n",
       "285        small antisense RNA aMEF;aMEF;b4844;ECK4701  2911051..2911450   \n",
       "286            small antisense RNA AsCRP;b4841;ECK4698  3485953..3486231   \n",
       "287    small antisense RNA AsflhD;AsflhD;b4839;ECK4696  1978176..1978395   \n",
       "288           small antisense RNA AsphoP;b4843;ECK4700  1190165..1190508   \n",
       "289        small antisense RNA StfZ;StfZ;b4840;ECK4697    104948..105499   \n",
       "298       small regulatory RNA AzuR;AzuR;b4434;ECK1902  1987843..1988001   \n",
       "314  small regulatory RNA GlnZ;glnA 3<sup>A</sup> U...  4056428..4056621   \n",
       "324             small regulatory RNA Och5;Och5;nan;nan  3067187..3067344   \n",
       "363                    spf ; ECK3856; IS197; JWR0097;   4049899..4050007   \n",
       "390                                                NaN  2883467..2883547   \n",
       "395                                                NaN    573588..573754   \n",
       "401                                                NaN        5301..5534   \n",
       "402                                                NaN  1422406..1422479   \n",
       "615                                                NaN  2883139..2883402   \n",
       "656                                                NaN    755177..758405   \n",
       "658                                    ECK0805; JW5969    851894..852163   \n",
       "673                                            ECs2662  2624379..2626136   \n",
       "\n",
       "      Strand                 Name overlaps Raw ID  \n",
       "285  forward    [small antisense RNA aMEF]   None  \n",
       "286  reverse   [small antisense RNA AsCRP]   None  \n",
       "287  forward  [small antisense RNA AsflhD]   None  \n",
       "288  forward  [small antisense RNA AsphoP]   None  \n",
       "289  reverse    [small antisense RNA StfZ]   None  \n",
       "298  reverse   [small regulatory RNA AzuR]   None  \n",
       "314  reverse   [small regulatory RNA GlnZ]   None  \n",
       "324  forward   [small regulatory RNA Och5]   None  \n",
       "363  forward                     [Spot_42]   None  \n",
       "390  forward                        [OrzO]   None  \n",
       "395  reverse                        [IpeX]   None  \n",
       "401  reverse                        [AfaR]   None  \n",
       "402  forward                       [Esr41]   None  \n",
       "615  reverse                        [zorO]   None  \n",
       "656  forward                     [sdhCDAB]   None  \n",
       "658  forward                        [yliL]   None  \n",
       "673  reverse                        [fliC]   None  "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data[(data['Raw ID'] == '') | (data['Raw ID'].isna())]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "seqbefore = {n: get_seq_info(None, thresh=200, before=True, dna_start=p.split('..')[0], dna_end=p.split('..')[1], nuc_id=i.split(':')[1].split(',')[0], strand=s) for n, p, i, s in list(data[data['Raw ID'].isna()].groupby(['Name', 'Genomic position', 'ID', 'Strand']).apply(str).index)}\n",
    "# seqafter = {n: get_seq_info(None, thresh=200, before=False, dna_start=p.split('..')[0], dna_end=p.split('..')[1], nuc_id=i.split(':')[1].split(',')[0], strand=s) for n, p, i, s in list(data[data['Raw ID'].isna()].groupby(['Name', 'Genomic position', 'ID', 'Strand']).apply(str).index)}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "seqbefore.update({n: get_seq_info(str(k), thresh=200, before=True) for n, k in list(data[~data['Raw ID'].isna()].groupby(['Name', 'Raw ID']).apply(str).index)})\n",
    "# seqafter.update({n: get_seq_info(str(k), thresh=200, before=False) for n, k in list(data[~data['Raw ID'].isna()].groupby(['Name', 'Raw ID']).apply(str).index)})\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "data['Sequence before'] = data['Name'].apply(lambda x: seqbefore[x])\n",
    "# data['Sequence after'] = data['Name'].apply(lambda x: seqafter[x])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "i = 3\n",
    "while (len(data[data['Sequence before'] == '']) > 0) and (i > 0):\n",
    "    for i, r in data[data['Sequence before'] == ''].iterrows():\n",
    "        data.loc[i, 'Sequence before'] = get_seq_info(str(r['Raw ID']), thresh=200, before=True)\n",
    "    i -= 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "      <th>ID</th>\n",
       "      <th>Category</th>\n",
       "      <th>Symbol</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Database</th>\n",
       "      <th>Alias</th>\n",
       "      <th>Genomic position</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Name overlaps</th>\n",
       "      <th>Raw ID</th>\n",
       "      <th>Sequence before</th>\n",
       "      <th>Sequence after</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [Name, ID, Category, Symbol, Sequence, Database, Alias, Genomic position, Strand, Name overlaps, Raw ID, Sequence before, Sequence after]\n",
       "Index: []"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data[data['Sequence before'] == '']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'cgaagggctggtgcgaatcaaccgtcagggcgatgacctgcacattgaaacgcaaaacgttgcgccaccggacagccgtattgagctgatttccagcagcgaagctgactggaatgccttgcagagcgcgctgttgaagcttcgtttagcgactaccgcgtaaggttgcctgcgtttttcagtaagataattagagaaaat'"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get_seq_info(str(946693), thresh=200, before=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'cggcggtatccatataagtacggatttgactgaacgtccccaggcggctgtcggcaataataatgccgagcgccaccagcaccgccagaataaggcgaatctgtagcgacgggccacggctaaaaattggcttcataagttatgcgtattctcgtatcagaccaggcagggtaaacagacacttcccctgcctgcatccga'"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get_seq_info(str(948588), thresh=200, before=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Save"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Name', 'ID', 'Category', 'Symbol', 'Sequence', 'Database', 'Alias',\n",
       "       'Genomic position', 'Strand', 'Name overlaps', 'Raw ID',\n",
       "       'Sequence before', 'Sequence after'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "data[['Name', 'ID', 'Category', 'Symbol', 'Sequence', 'Database', 'Alias',\n",
    "      'Genomic position', 'Strand', 'Sequence before']].to_csv('../data/sRNA/merged_EcoCyc_RNAInter_sRNATarBase.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "env_evo",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
